
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Requirements &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Roadmap" href="roadmap.html" />
    <link rel="prev" title="HOW to release a version" href="release.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="requirements">
<h1>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h1>
<p>A list of features to be added to the package. Requirements marked with <code class="docutils literal notranslate"><span class="pre">(*)</span></code>
are not yet assigned a version number.</p>
<dl>
<dt>[R01]</dt><dd><p><code class="docutils literal notranslate"><span class="pre">bam-filter</span></code> covered with tests</p>
</dd>
<dt>[R02]</dt><dd><p><code class="docutils literal notranslate"><span class="pre">samtools</span></code> replaced by <code class="docutils literal notranslate"><span class="pre">pysam</span></code></p>
</dd>
<dt>[R03]</dt><dd><p>Add option <code class="docutils literal notranslate"><span class="pre">--partition</span></code> to <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> to select what fraction of an input
file must be processed</p>
</dd>
<dt>[R04]</dt><dd><p>Implement new <code class="docutils literal notranslate"><span class="pre">csv</span></code> output with methilation state per molecule and circular
consensus DNA sequence</p>
</dd>
<dt>[R05]</dt><dd><p>Option to select model in <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> (simply providing a string, eg <code class="docutils literal notranslate"><span class="pre">P6-C4</span></code>).
Instead of crashing if the expected model dows not agree with the chemistry, the
program looks for different models in the directory and tests them in turn until
a valid one is found.</p>
</dd>
<dt>[R06]</dt><dd><p><em>Legacy code</em> covered with tests: minimal <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> functionality</p>
</dd>
<dt>[R07]</dt><dd><p>Optimization: Reduce the number of temporary files that are created
by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>.</p>
</dd>
<dt>[R08]</dt><dd><p>Decouple from ipdSummary using kineticsTools as a package</p>
</dd>
<dt>[R09]</dt><dd><p>Automatically identify structure of BAM file:</p>
<ul class="simple">
<li><p>where the molecule id is located (column)</p></li>
</ul>
</dd>
<dt>[R10]</dt><dd><p>BAM files created contain info about source file and modifications introduced.</p>
</dd>
<dt>[R11]</dt><dd><p>Miscelaneous improvements:</p>
<ul class="simple">
<li><p>protection against tracebacks (they should not be presented to end user)</p></li>
<li><p>version</p></li>
<li><p>debugging</p></li>
</ul>
</dd>
<dt>[R12]</dt><dd><p>Path to external tools can be provided:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">blasr</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">pbindex</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code></p></li>
</ul>
</dd>
<dt>[R13]</dt><dd><p>Number of processes used by external tools can be chosen.</p>
</dd>
<dt>[R14]</dt><dd><p>Resume interrupted <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> runs.</p>
</dd>
<dt>[R15]</dt><dd><p>Various minor input options:</p>
<ul class="simple">
<li><p>modification types</p></li>
<li><p>keep temporary directory?</p></li>
<li><p>only produce methylation report</p></li>
<li><p>ccs file</p></li>
<li><p>aligned ccs file</p></li>
</ul>
</dd>
<dt>[R16] <code class="docutils literal notranslate"><span class="pre">(*)</span></code></dt><dd><p>Transparent usage of clusters.</p>
</dd>
<dt>[R17]</dt><dd><p>GUI</p>
</dd>
<dt>[R18]</dt><dd><p>Easy installation</p>
</dd>
<dt>[R19]</dt><dd><p>Methylation state conforms to version 2 (see <a class="reference internal" href="../indepth/methylation_reports.html#methylation-report-format"><span class="std std-ref">Methylation Reports</span></a>).</p>
</dd>
<dt>[R20]</dt><dd><p>Documentation (I)</p>
<blockquote>
<div><ul class="simple">
<li><p>Structure documentation</p></li>
<li><p>Add quick start and some more docs for end users</p></li>
<li><p>add docstrings as a starting point for docs for developers</p></li>
<li><p>integrate with sphinx</p></li>
</ul>
</div></blockquote>
</dd>
<dt>[R21]</dt><dd><p>Methylation report format V3</p>
</dd>
<dt>[R22]</dt><dd><p>Summary doc for humans (html?, pdf?) with:</p>
<ul class="simple">
<li><p>plots</p></li>
<li><p>basic statistics about subreads, molecules, methylations, etc.</p></li>
</ul>
</dd>
<dt>[R23]</dt><dd><p>New option that disables the standard filters before
running the Single Molecule Analysis (<code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>)</p>
</dd>
<dt>[R24]</dt><dd><p>Path to <code class="docutils literal notranslate"><span class="pre">ccs</span></code> can be provided as CLO.</p>
</dd>
<dt>[R25]</dt><dd><p>Documentation (II)</p>
<blockquote>
<div><ul class="simple">
<li><p>local distribution and installation of docs</p></li>
<li><p>launch local docs in browser</p></li>
</ul>
</div></blockquote>
</dd>
<dt>[R26]</dt><dd><p>BamFile implementation using pure python library <code class="docutils literal notranslate"><span class="pre">pybam</span></code> (or a
python3 compliant fork of it)</p>
</dd>
<dt>[R27]</dt><dd><p>Organization of files produced by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>:</p>
<blockquote>
<div><ul class="simple">
<li><p>log files go to a dedicated <code class="docutils literal notranslate"><span class="pre">log/</span></code> directory (each molecule
must have its onw log file)</p></li>
<li><p>auxiliary files (<code class="docutils literal notranslate"><span class="pre">ì-shifted...</span></code>, aligned BAMs, CCS, etc)
must go into a dedicated directory (<code class="docutils literal notranslate"><span class="pre">aux/</span></code>?)</p></li>
</ul>
</div></blockquote>
</dd>
<dt>[R28]</dt><dd><p>Browsable HTML document with information about status of individual
GATCs and the neighbours.</p>
</dd>
<dt>[R29]</dt><dd><p>Automatic merge of partition results</p>
</dd>
<dt>[R30]</dt><dd><p>Option to choose the window of the reference to be processed by <code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code>.</p>
</dd>
<dt>[R31]</dt><dd><p>Switching to use the <code class="docutils literal notranslate"><span class="pre">pbmm2</span></code> aligner by default.</p>
</dd>
</dl>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../indepth/index.html">Guides</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Developer’s documentation</a><ul>
      <li>Previous: <a href="release.html" title="previous chapter">HOW to release a version</a></li>
      <li>Next: <a href="roadmap.html" title="next chapter">Roadmap</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/devel/requirements.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>