
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Methylation Reports &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Raw Detections" href="raw_detections.html" />
    <link rel="prev" title="Summary Reports" href="summary_reports.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="methylation-reports">
<span id="methylation-report-format"></span><h1>Methylation Reports<a class="headerlink" href="#methylation-reports" title="Permalink to this headline">¶</a></h1>
<dl class="field-list simple">
<dt class="field-odd">abstract</dt>
<dd class="field-odd"><p>The <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">Single molecule analysis with the sm-analysis program</span></a> pipeline produces a so-called
<em>methylation report</em>. Its contents is described in this
document.</p>
</dd>
</dl>
<section id="format">
<h2>Format<a class="headerlink" href="#format" title="Permalink to this headline">¶</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This section describes the most recent version of
the methylation report’s format.</p>
</div>
<p>The <em>methylation report</em> produced by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> is a <a class="reference internal" href="glossary.html#term-CSV-file"><span class="xref std std-term">CSV file</span></a>
with <code class="docutils literal notranslate"><span class="pre">;</span></code> (semicolon) as separator and <strong>21 columns</strong> with the following
header:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">molecule</span> <span class="nb">id</span><span class="p">;</span><span class="n">sequence</span><span class="p">;</span><span class="n">start</span> <span class="n">of</span> <span class="n">molecule</span><span class="p">;</span><span class="n">end</span> <span class="n">of</span> <span class="n">molecule</span><span class="p">;</span><span class="nb">len</span><span class="p">(</span><span class="n">molecule</span><span class="p">);</span><span class="n">count</span><span class="p">(</span><span class="n">subreads</span><span class="o">+</span><span class="p">);</span><span class="n">count</span><span class="p">(</span><span class="n">subreads</span><span class="o">-</span><span class="p">);</span><span class="n">combined</span> <span class="n">QUAL</span><span class="p">;</span><span class="n">mean</span><span class="p">(</span><span class="n">QUAL</span><span class="p">);</span><span class="n">sim</span> <span class="n">ratio</span><span class="p">;</span><span class="n">count</span><span class="p">(</span><span class="n">GATC</span><span class="p">);</span><span class="n">positions</span> <span class="n">of</span> <span class="n">GATCs</span><span class="p">;</span><span class="n">count</span><span class="p">(</span><span class="n">methylation</span> <span class="n">states</span><span class="p">);</span><span class="n">methylation</span> <span class="n">states</span><span class="p">;</span><span class="n">combined</span> <span class="n">score</span><span class="p">;</span><span class="n">mean</span><span class="p">(</span><span class="n">score</span><span class="p">);</span><span class="nb">min</span><span class="p">(</span><span class="n">IPDRatio</span><span class="p">);</span><span class="n">mean</span><span class="p">(</span><span class="n">IPDRatio</span><span class="p">);</span><span class="n">combined</span> <span class="n">idQV</span><span class="p">;</span><span class="n">mean</span><span class="p">(</span><span class="n">idQV</span><span class="p">);</span><span class="n">mean</span><span class="p">(</span><span class="n">coverage</span><span class="p">)</span>
</pre></div>
</div>
<p>Each column can be itself separated (see e.g. columns 12 and 14). In that case
an <em>internal</em> separator, namely <code class="docutils literal notranslate"><span class="pre">,</span></code>, is used.</p>
<p>The following table summarizes the meaning of each column.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 20%" />
<col style="width: 20%" />
<col style="width: 36%" />
<col style="width: 14%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>col num</p></th>
<th class="head"><p>field name</p></th>
<th class="head"><p>possible values</p></th>
<th class="head"><p>description</p></th>
<th class="head"><p>example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">molecule</span> <span class="pre">id</span></code></p></td>
<td><p>positive int</p></td>
<td><p>value provided by the sequencer</p></td>
<td><p>23480</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code></p></td>
<td><p>[ACGT]*</p></td>
<td><p>DNA sequence corresponding to
the molecule, as reported by CCS</p></td>
<td><p>AGACTTTC…</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">start</span> <span class="pre">of</span></code>
<code class="docutils literal notranslate"><span class="pre">molecule</span></code></p></td>
<td><p>positive int</p></td>
<td><p>1-based start position of the
molecule within the reference;
the values are taken from the
aligner; this value is the
number of bases before the first
base of the sequence plus 1
(the minimum position is 1)</p></td>
<td><p>312</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">end</span> <span class="pre">of</span></code>
<code class="docutils literal notranslate"><span class="pre">molecule</span></code></p></td>
<td><p>positive int</p></td>
<td><p>1-based end position of the
molecule within the reference;
the values are taken from the
aligner. This value is the
number of bases before the last
base of the sequence plus 1</p></td>
<td><p>1509</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">len(molecule)</span></code></p></td>
<td><p>positive int</p></td>
<td><p>length of the DNA sequence
corresponding to the molecule,
according to the aligned CCS
file.</p></td>
<td><p>1198</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">count(sub-</span></code>
<code class="docutils literal notranslate"><span class="pre">reads+)</span></code></p></td>
<td><p>int &gt;= 0</p></td>
<td><p>number of subreads in the +
strand found in the input BAM</p></td>
<td><p>51</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">count(sub-</span></code>
<code class="docutils literal notranslate"><span class="pre">reads-)</span></code></p></td>
<td><p>int &gt;= 0</p></td>
<td><p>number of subreads in the -
strand found in the input BAM</p></td>
<td><p>48</p></td>
</tr>
<tr class="row-odd"><td><p>8</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">combined</span></code>
<code class="docutils literal notranslate"><span class="pre">QUAL</span></code></p></td>
<td><p>positive float</p></td>
<td><p>combined <em>QUAL</em> (asccii of base
quality plus 33). Each QUAL
is the Phred-transformed proba-
bility value that the base is
wrong.</p></td>
<td><p>95.2</p></td>
</tr>
<tr class="row-even"><td><p>9</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean(QUAL)</span></code></p></td>
<td><p>positive float</p></td>
<td><p>mean <em>QUAL</em> (asccii of base
quality plus 33). Each QUAL
is the Phred-transformed proba-
bility value that the base is
wrong.</p></td>
<td><p>101.4</p></td>
</tr>
<tr class="row-odd"><td><p>10</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sim</span> <span class="pre">ratio</span></code></p></td>
<td><p>float between 0
and 1</p></td>
<td><p>ratio of similarity between the
molecule’s sequence and the
corresponding piece in the
reference</p></td>
<td><p>1.0</p></td>
</tr>
<tr class="row-even"><td><p>11</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">count(GATC)</span></code></p></td>
<td><p>positive int</p></td>
<td><p>number of GATCs found in the DNA
sequence.</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-odd"><td><p>12</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">positions</span></code>
<code class="docutils literal notranslate"><span class="pre">of</span> <span class="pre">GATCs</span></code></p></td>
<td><p>comma separated
sequence of int&gt;0</p></td>
<td><p>1-based absolute positions of
the Gs for all the GATCs present
in col 3 (ie, in the current
molecule)</p></td>
<td><p>315,699,902</p></td>
</tr>
<tr class="row-even"><td><p>13</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">count(methy-</span></code>
<code class="docutils literal notranslate"><span class="pre">lation</span></code>
<code class="docutils literal notranslate"><span class="pre">states)</span></code></p></td>
<td><p>positive int</p></td>
<td><p>in how many positions the mole-
cule was detected to have a
methylation (<code class="docutils literal notranslate"><span class="pre">+</span></code>, <code class="docutils literal notranslate"><span class="pre">-</span></code> or
<code class="docutils literal notranslate"><span class="pre">f</span></code> in column 7).</p></td>
<td><p>2</p></td>
</tr>
<tr class="row-odd"><td><p>14</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">methylation</span></code>
<code class="docutils literal notranslate"><span class="pre">states</span></code></p></td>
<td><p>comma separated
sequence of [0-+f]</p></td>
<td><p>each element corresponds to the
methylation state of one GATC in
the sequence as returned by the
<code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code> program</p></td>
<td><p>f,-,0</p></td>
</tr>
<tr class="row-even"><td><p>15</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">combined</span></code>
<code class="docutils literal notranslate"><span class="pre">score</span></code></p></td>
<td><p>positive float</p></td>
<td><p>combined <em>score of the feature</em>
for each detection (each score
is the Phred-transformed proba-
bility value that a kinetic de-
viation exists at a position)</p></td>
<td><p>118</p></td>
</tr>
<tr class="row-odd"><td><p>16</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean(score)</span></code></p></td>
<td><p>positive float</p></td>
<td><p>mean <em>score of the feature</em>
over all detections in the mole-
cule (each score is the Phred-
transformed probability value
that a kinetic deviation exists
at a position)</p></td>
<td><p>150</p></td>
</tr>
<tr class="row-even"><td><p>17</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">min(IPDRatio)</span></code></p></td>
<td><p>positive float</p></td>
<td><p>min of tMean/modelPredictions
(tMean is the capped mean of
normalized IPDs observed at
this position)</p></td>
<td><p>3.4</p></td>
</tr>
<tr class="row-odd"><td><p>18</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean(IPDRatio)</span></code></p></td>
<td><p>positive float</p></td>
<td><p>mean of tMean/modelPredictions
(tMean is the capped mean of
normalized IPDs observed at
this position)</p></td>
<td><p>5.2</p></td>
</tr>
<tr class="row-even"><td><p>19</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">combined</span> <span class="pre">idQV</span></code></p></td>
<td><p>positive float</p></td>
<td><p>combined <code class="docutils literal notranslate"><span class="pre">idQV</span></code> value for all
the detected modifications of
the correct type in the given
molecule</p></td>
<td><p>19.6</p></td>
</tr>
<tr class="row-odd"><td><p>20</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean(idQV)</span></code></p></td>
<td><p>positive float</p></td>
<td><p>mean <code class="docutils literal notranslate"><span class="pre">idQV</span></code> value for all
the detected modifications of
the correct type in the given
molecule</p></td>
<td><p>30</p></td>
</tr>
<tr class="row-even"><td><p>21</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">mean(coverage)</span></code></p></td>
<td><p>positive float</p></td>
<td><p>mean value of the coverage
levels used to assign the modif.
type label</p></td>
<td><p>42</p></td>
</tr>
</tbody>
</table>
<p>Some notes:</p>
<ul class="simple">
<li><p>the number of elements in columns 12 and 14 must be equal to the value in column 11</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">idQV</span></code> is the <a class="reference internal" href="phred_transformed_scores.html#phred-transformed-scores"><span class="std std-ref">Phred-transformed</span></a> QV of having
a modification at a given position</p></li>
<li><p>The meaning of the methylation state symbols:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">0</span></code>:  not methylated</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-</span></code>:  hemi-methylated. Negative strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">+</span></code>:  hemi-methylated. Positive strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">f</span></code>:  full methylated</p></li>
</ul>
</li>
</ul>
</section>
<section id="format-version-2">
<h2>Format (version 2)<a class="headerlink" href="#format-version-2" title="Permalink to this headline">¶</a></h2>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Please, ignore the content of this section if you are working with
a public release of <strong>PacBio Data Processing</strong> (one installed with <code class="docutils literal notranslate"><span class="pre">pip</span></code>, for
instance). It is kept here for reference.</p>
</div>
<p>The <em>methylation report</em> produced by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> is a csv file with <code class="docutils literal notranslate"><span class="pre">;</span></code>
(semicolon) as separator and <strong>7 columns</strong> with the following header:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">molecule</span> <span class="nb">id</span><span class="p">;</span><span class="n">count</span><span class="p">(</span><span class="n">GATC</span><span class="p">);</span><span class="n">sequence</span><span class="p">;</span><span class="n">start</span> <span class="n">of</span> <span class="n">molecule</span><span class="p">;</span><span class="n">end</span> <span class="n">of</span> <span class="n">molecule</span><span class="p">;</span><span class="n">positions</span> <span class="n">of</span> <span class="n">GATCs</span><span class="p">;</span><span class="n">methylation</span> <span class="n">states</span>
</pre></div>
</div>
<p>The following table summarizes the meaning of each column.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 18%" />
<col style="width: 20%" />
<col style="width: 37%" />
<col style="width: 15%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>col num</p></th>
<th class="head"><p>field name</p></th>
<th class="head"><p>possible values</p></th>
<th class="head"><p>description</p></th>
<th class="head"><p>example</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">molecule</span> <span class="pre">id</span></code></p></td>
<td><p>positive int</p></td>
<td><p>value provided by the sequencer</p></td>
<td><p>23480</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">count(GATC)</span></code></p></td>
<td><p>positive int</p></td>
<td><p>number of GATCs found in the DNA
sequence.</p></td>
<td><p>3</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code></p></td>
<td><p>[ACGT]*</p></td>
<td><p>DNA sequence corresponding to
the molecule, as reported by CCS</p></td>
<td><p>AGACTTTC…</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">start</span> <span class="pre">of</span></code>
<code class="docutils literal notranslate"><span class="pre">molecule</span></code></p></td>
<td><p>positive int</p></td>
<td><p>1-based start position of the
molecule within the reference;
the values are taken from the
aligner; this value is the
number of bases before the first
base of the sequence plus 1
(the minimum position is 1)</p></td>
<td><p>312</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">end</span> <span class="pre">of</span></code>
<code class="docutils literal notranslate"><span class="pre">molecule</span></code></p></td>
<td><p>positive int</p></td>
<td><p>1-based end position of the
molecule within the reference;
the values are taken from the
aligner. This value is the
number of bases before the last
base of the sequence plus 1</p></td>
<td><p>1509</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">positions</span></code>
<code class="docutils literal notranslate"><span class="pre">of</span> <span class="pre">GATCs</span></code></p></td>
<td><p>comma separated
sequence of int&gt;0</p></td>
<td><p>1-based absolute positions of
the Gs for all the GATCs present
in col 3 (ie, in the current
molecule)</p></td>
<td><p>315,699,1002</p></td>
</tr>
<tr class="row-even"><td><p>7</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">methylation</span></code>
<code class="docutils literal notranslate"><span class="pre">states</span></code></p></td>
<td><p>comma separated
sequence of [0-+f]</p></td>
<td><p>each element corresponds to the
methylation state of one GATC in
the sequence as returned by the
<code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code> program</p></td>
<td><p>f,-,0</p></td>
</tr>
</tbody>
</table>
<p>Some notes:</p>
<ul class="simple">
<li><p>the number of elements in columns 6 and 7 must be equal to the value in column 2</p></li>
<li><p>The meaning of the methylation state symbols:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">0</span></code>:  not methylated</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-</span></code>:  hemi-methylated. Negative strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">+</span></code>:  hemi-methylated. Positive strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">f</span></code>:  full methylated</p></li>
</ul>
</li>
</ul>
</section>
<section id="format-version-1">
<h2>Format (version 1)<a class="headerlink" href="#format-version-1" title="Permalink to this headline">¶</a></h2>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Please, ignore the content of this section if you are working with
a public release of <strong>PacBio Data Processing</strong> (one installed with <code class="docutils literal notranslate"><span class="pre">pip</span></code>, for
instance). It is kept here for reference.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This version, v1, is an old format no longer used. It was decided to be
replaced by the version 2 (described above) in a work meeting (with DP,
DV and TW) on 18 June 20201.</p>
</div>
<p>The <em>methylation report</em> produced by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> is a csv file with <code class="docutils literal notranslate"><span class="pre">;</span></code>
(semicolon) as separator and <strong>6 columns</strong> with the following header:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">molecule</span> <span class="nb">id</span><span class="p">;</span><span class="n">count</span><span class="p">(</span><span class="n">GATC</span><span class="p">);</span><span class="n">sequence</span><span class="p">;</span><span class="n">start</span><span class="o">-</span><span class="n">end</span> <span class="n">of</span> <span class="n">molecule</span><span class="p">;</span>
<span class="n">positions</span> <span class="n">of</span> <span class="n">GATCs</span><span class="p">;</span><span class="n">methylation</span> <span class="n">states</span>
</pre></div>
</div>
<p>The following table summarizes the meaning of each column.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 10%" />
<col style="width: 33%" />
<col style="width: 21%" />
<col style="width: 37%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>col num</p></th>
<th class="head"><p>field name</p></th>
<th class="head"><p>possible values</p></th>
<th class="head"><p>description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">molecule</span> <span class="pre">id</span></code></p></td>
<td><p>positive int</p></td>
<td><p>value provided by the sequencer</p></td>
</tr>
<tr class="row-odd"><td><p>2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">count(GATC)</span></code></p></td>
<td><p>positive int</p></td>
<td><p>number of GATCs found in the DNA
sequence.</p></td>
</tr>
<tr class="row-even"><td><p>3</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sequence</span></code></p></td>
<td><p>[ACGT]*</p></td>
<td><p>DNA sequence corresponding to
the molecule, as reported by CCS</p></td>
</tr>
<tr class="row-odd"><td><p>4</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">start-end</span> <span class="pre">of</span> <span class="pre">molecule</span></code></p></td>
<td><p>[int&gt;=,int&gt;0]</p></td>
<td><p>inclusive interval corresponding
to the start and end of the
molecule within the reference;
the values are taken from the
aligner but shifted such that
the minimum position is 0 (ie
0-index is used)</p></td>
</tr>
<tr class="row-even"><td><p>5</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">positions</span> <span class="pre">of</span> <span class="pre">GATCs</span></code></p></td>
<td><p>space separated
sequence of int&gt;0</p></td>
<td><p>0-index positions of the A in all
the GATCs present in col 3 and
realtive to that sequence</p></td>
</tr>
<tr class="row-odd"><td><p>6</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">methylation</span> <span class="pre">states</span></code></p></td>
<td><p>space separated
sequence of [0-+f]</p></td>
<td><p>each element corresponds to the
methylation state of one GATC in
the sequence as returned by the
<code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code> program</p></td>
</tr>
</tbody>
</table>
<p>Some notes:</p>
<ul class="simple">
<li><p>the number of elements in columns 5 and 6 must be equal to the value in column 2</p></li>
<li><p>The meaning of the methylation state symbols:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">0</span></code>:  not methylated</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-</span></code>:  hemi-methylated. Negative strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">+</span></code>:  hemi-methylated. Positive strand</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">f</span></code>:  full methylated</p></li>
</ul>
</li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Guides</a><ul>
      <li>Previous: <a href="summary_reports.html" title="previous chapter">Summary Reports</a></li>
      <li>Next: <a href="raw_detections.html" title="next chapter">Raw Detections</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/indepth/methylation_reports.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>