
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>PacBio tools &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Pbindex" href="pbindex.html" />
    <link rel="prev" title="Using PacBio Data Processing on a cluster" href="cluster.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="pacbio-tools">
<span id="id1"></span><h1>PacBio tools<a class="headerlink" href="#pacbio-tools" title="Permalink to this headline">¶</a></h1>
<section id="smrt-link-software-server-tool">
<span id="smrt-link"></span><h2>SMRT-link software server tool<a class="headerlink" href="#smrt-link-software-server-tool" title="Permalink to this headline">¶</a></h2>
<p>The sequencing PacBio data is normally processed by a sofware-server
tool
<a class="reference external" href="https://www.pacb.com/products-and-services/analytical-software/smrt-analysis/">SMRT-link</a>.
Which means that a Software to process the PacBio data should be
intalled in a server and then processed there using the standars
protocols of pacbio.</p>
<p>Here some additional information related to SMART-link:</p>
<ul class="simple">
<li><p><a class="reference external" href="https://www.pacb.com/support/software-downloads/">All SMRT software
versions</a></p></li>
<li><p><a class="reference external" href="http://files.pacb.com/software/smrtanalysis/2.2.0/doc/smrtportal/help/!SSL!/Webhelp/Portal_Main.htm">SMRT® Portal
Help</a></p></li>
<li><p><a class="reference external" href="https://www.youtube.com/user/ZmjhsiehZ/videos">Installing SMRT_youtube
videos</a></p></li>
<li><p><a class="reference external" href="https://www.pacb.com/blog/the-evolution-of-dna-sequencing-tools/">Evolution of sequencing
era</a></p></li>
<li><p><a class="reference external" href="https://www.pacb.com/wp-content/uploads/2015/09/WP_Detecting_DNA_Base_Modifications_Using_SMRT_Sequencing.pdf">The white
paper</a></p></li>
<li><p><a class="reference external" href="http://seqanswers.com/forums/showthread.php?s=f2bc6a3bb12677c9111e6608269733f3&amp;t=93071">SMRT link without
control?</a></p></li>
</ul>
</section>
<section id="pacbio-bioconda">
<h2>PacBio-Bioconda<a class="headerlink" href="#pacbio-bioconda" title="Permalink to this headline">¶</a></h2>
<p>If you are familiar with command lines and programing the PacBio company
provide a collection of tools to manage the sequencing data. This can be
installed in a server or a personal computer using a virtual environment
called Bioconda, specifically Miniconda. The collection of tools are
called the <a class="reference external" href="https://github.com/PacificBiosciences/pbbioconda">PacBio Bioconda
tools</a>.</p>
<section id="installing-miniconda">
<h3>Installing Miniconda<a class="headerlink" href="#installing-miniconda" title="Permalink to this headline">¶</a></h3>
<p>It is possible to work with the PacBio data using differents tools
installed through Bioconda (virtual environment).</p>
<p>Follow the next steps to install Bioconda and its most useful tools.</p>
<ol class="arabic simple">
<li><p><strong>Download miniconda</strong> Choose between miniconca 2 (for using python
2) and miniconca 3 (for using python 3) in the <a class="reference external" href="https://bioconda.github.io/user/install.html#install-conda">miniconda
webpage</a>
depending which version you need to work, choose the operative system
you are using and donwload it.</p></li>
</ol>
<p>To download miniconda 2 in MacOS, you can type the following line in
your terminal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><style type="text/css">
span.prompt1:before {
  content: "$ ";
}
</style><span class="prompt1">curl -O https://repo.anaconda.com/miniconda/Miniconda2-latest-MacOSX-x86_64.sh</span>
</pre></div></div><p>To download miniconda 3 in MacOS, you can type the following line in
your terminal:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">curl -O https://repo.anaconda.com/miniconda/Miniconda3-latest-MacOSX-x86_64.sh</span>
</pre></div></div><ol class="arabic simple" start="2">
<li><p><strong>Install miniconda</strong> Using the terminal in the directory where
miniconda was downloaded, the following line if you want to install
miniconda 2 or 3 (just change the name). In this case miniconda 3 is
going to be installed:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">sh Miniconda3-latest-MacOSX-x86_64.sh</span>
</pre></div></div><ol class="loweralpha simple">
<li><p>Standard directory to install miniconda is in you home directory and
a directory called miniconda3 will be created. Don’t forget the
location of this folder because this is the route to call the virtual
environment.</p></li>
<li><p>The installation process also ask you if you want to specify a
directory. You can create a new one only typing a directory name
after the <code class="docutils literal notranslate"><span class="pre">&gt;&gt;&gt;</span></code>prompt. This is optional and useful when you want
to have more tha one virtual environment related to miniconda.</p></li>
<li><p>Another important question during the installation is if you wish the
installer to initialize Miniconda3, wich means that every time you
start the terminal, the virtual environment will be activated and the
tools installed there. You only need to type ‘yes’ or ‘not’ after the
‘&gt;&gt;&gt;’ prompt.</p></li>
</ol>
<ol class="arabic simple" start="3">
<li><p><strong>Activate miniconda</strong> If you type ‘not’ in the las step, then, when
you open the terminal, go to your home directory where miniconda was
installed and activate it as the following example (activation for
miniconda3):</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1"><span class="nb">source</span> home_directory_route/miniconda3/bin/activate</span>
</pre></div></div><ol class="arabic simple" start="4">
<li><p><strong>Configure miniconda</strong> The first step is to have miniconda
activated. Then, to configure miniconda installation, type the
following lines in this order:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">conda config --add channels defaults</span>
<span class="prompt1">conda config --add channels bioconda</span>
<span class="prompt1">conda config --add channels conda-forge</span>
</pre></div></div><ol class="arabic simple" start="5">
<li><p><strong>Installing tools</strong> The more useful tools can be installed using the
following line:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">conda install -c bioconda blasr pbcore numpy scipy pbbam samtools pbcommand</span>
</pre></div></div><p>If you are interested in other tools, check the collections of tools you
can install in this link –&gt; <a class="reference external" href="https://github.com/PacificBiosciences/pbbioconda">PacBio-Bioconda
tools</a> Is
recomendable to update the tools after installation. To do this type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">conda update tool_name</span>
</pre></div></div><p>for invidual tools or:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">conda update --all</span>
</pre></div></div><p>to update all the tools.</p>
<ol class="arabic simple" start="6">
<li><p><strong>Deactivate miniconda</strong> To deactivate the miniconda you don’t need
to be in the home directory, just type this line and miniconda will
be deactivate:</p></li>
</ol>
<div class="highlight-default notranslate"><div class="highlight"><pre><span class="prompt1">conda deactivate</span>
</pre></div></div><p>Some of the more useful tools to consider are:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p><strong>Tool</strong></p></th>
<th class="head"><p><strong>Description</strong></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>bax2bam</p></td>
<td><p>Allows to transform files from
old chemiestries and formats to
the bam file format.</p></td>
</tr>
<tr class="row-odd"><td><p>blasr</p></td>
<td><p>The official PacBio aligner
adapted for long sequencing
reads. Although other aligners
such as <em>BWA</em>, <em>Segemehl</em> and
<em>pbalign</em> were compared, <em>blasr</em>
had the best mapping along with
pbaling, both aligners found in
the pacbio bioconda tools. It was
decided to take <em>blasr</em> as the
aligner to do the analyses
because it is the only one whose
result includes the ipd columns
needed to be able to detect DNA
modifications.</p></td>
</tr>
<tr class="row-even"><td><p>pbmm2</p></td>
<td><p>New aligner suggested to be a
substitute for blasr. When
evaluated, it turned out to be
faster in the alignment process,
however there is not a big
difference in the total number of
aligned subreads. The output was
not sorted by molecule and has
therefore been discarded for the
time being.</p></td>
</tr>
<tr class="row-odd"><td><p>pbccs</p></td>
<td><p>This produce the Circular
Consensus Sequencing from all the
subreads on each molecule</p></td>
</tr>
</tbody>
</table>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Guides</a><ul>
      <li>Previous: <a href="cluster.html" title="previous chapter">Using <strong>PacBio Data Processing</strong> on a cluster</a></li>
      <li>Next: <a href="pbindex.html" title="next chapter">Pbindex</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/indepth/pacbio_tools.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>