
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>What is the PacBio sequencing technology? &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="BAM format" href="bam.html" />
    <link rel="prev" title="Guides" href="index.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="what-is-the-pacbio-sequencing-technology">
<h1>What is the PacBio sequencing technology?<a class="headerlink" href="#what-is-the-pacbio-sequencing-technology" title="Permalink to this headline">¶</a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>This document, or parts of it, might be obsolete. Be careful.</p>
</div>
<p>PacBio sequencing technology allows the measurement of DNA
polymerization kinetics during the sequencing process. When the DNA
presents some type of modification such as methylation, the PacBio
sequencer records a change in the polymerization kinetics. The
relationship between modified DNA and the change in polymerization time
detected by this technology has allowed the study of methylation in many
organisms.</p>
<section id="pacbio-protocol">
<h2>PacBio protocol<a class="headerlink" href="#pacbio-protocol" title="Permalink to this headline">¶</a></h2>
<ol class="arabic simple">
<li><p>DNA extraction</p></li>
<li><p>DNA processing</p></li>
<li><p>DNA sequencing</p></li>
<li><p>Data output</p></li>
</ol>
</section>
<section id="how-the-pacbio-sequencer-output-is-processed">
<h2>How the PacBio sequencer output is processed?<a class="headerlink" href="#how-the-pacbio-sequencer-output-is-processed" title="Permalink to this headline">¶</a></h2>
<p>The first version of the PacBio sequencer (PacBio RS) produced only one
output file named bas.h5, in the following versions of the sequencer
(PacBio RS II), the output file produced four output files composed of 3
bax.h5 files and one bas.h5 file. Currently, the latest generation of
PacBio sequencers produces a single output file with a .bam extension
(<a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">binary alignment
map</a>) . It is
possible to convert the output formats bas.h5/bax.h5 to a .bam format
using bioinformatic tools that can be installed through
<a class="reference external" href="https://github.com/PacificBiosciences/pbbioconda">PacBio-Bioconda</a>
which offers several tools that will be useful during the primary and
secondary analysis of the data generated by the PacBio sequencer.</p>
<section id="primary-and-secondary-analysis">
<h3>Primary and Secondary analysis<a class="headerlink" href="#primary-and-secondary-analysis" title="Permalink to this headline">¶</a></h3>
<p>After sequencing, the generated PacBio output is processed in two steps.</p>
<ul class="simple">
<li><p>A <strong>Primary analysis</strong> is made by the instrument to obtain information
related to the sequenced DNA like IPD values and sequencing quality.</p></li>
<li><p>A <strong>Secondary analysis</strong> includes some of the steps included in
sm-analysis software like alignment and circular consensus sequence.</p></li>
</ul>
<p>More information can be found <a class="reference external" href="https://www.pacb.com/wp-content/uploads/2015/09/Pacific-Biosciences-Glossary-of-Terms.pdf">here</a>.</p>
</section>
<section id="pacbio-tools">
<h3>PacBio Tools<a class="headerlink" href="#pacbio-tools" title="Permalink to this headline">¶</a></h3>
<p>PacBio provides the following analysis tools:</p>
<ul class="simple">
<li><p><strong>SMRT</strong>: Must be installed on server</p></li>
<li><p><strong>PacBio Bioconda</strong>: Can be installed on server and personal computer</p></li>
</ul>
<p><a class="reference external" href="https://github.com/PacificBiosciences/pbbioconda">PacBio-Bioconda</a>
tools are installed using a virtual environment. This includes several
tools, here we describe some of them:</p>
<ul class="simple">
<li><p><strong>bax2bam</strong> -
<a class="reference external" href="https://github.com/pacificbiosciences/bax2bam/">bax2bam</a> allows
to transform pacbio files from old versions to a bam file manageable
with pacbio tools.</p></li>
<li><p><strong>blasr</strong> - The official PacBio aligner adapted for long sequencing
reads. Although other aligners such as BWA, Segemehl, and pbalign
were compared,
<a class="reference external" href="https://github.com/pacificbiosciences/blasr/">blasr</a> had the best
mapping along with pbaling, both aligners found in the pacbio
bioconda tools. It was decided to take blasr as the aligner to do the
analyses because it is the only one whose result includes the ipd
columns needed to be able to detect DNA modifications.</p></li>
<li><p><strong>pbmm2</strong> - <a class="reference external" href="https://github.com/PacificBiosciences/pbmm2/">pbmm2</a>
is a an aligner suggested to be a substitute for blasr. When
evaluated, it turned out to be faster in the alignment process,
however, there is not a big difference in the total number of aligned
subreads. The output was not sorted by molecule and has therefore
been discarded for the time being.</p></li>
<li><p><strong>CCS</strong> - This tool generate the Circular Consensus Sequence (
<a class="reference external" href="https://ccs.how/">CCS</a> )combining multiple subreads.</p></li>
</ul>
</section>
<section id="modification-detection-tools">
<h3>Modification detection tools<a class="headerlink" href="#modification-detection-tools" title="Permalink to this headline">¶</a></h3>
<p>There are several <a class="reference external" href="https://github.com/PacificBiosciences/Bioinformatics-Training/wiki/Base-Modification-Tools">Base Modification
Tools</a>
related to the detection of modified bases in DNA from files generated
by PacBio sequencing.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 52%" />
<col style="width: 48%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Base Modification Tool</p></th>
<th class="head"><p>Programming language</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>R-kinetics</strong></p></td>
<td><p>R</p></td>
</tr>
<tr class="row-odd"><td><p><strong>kineticsTools</strong></p></td>
<td><p>Python</p></td>
</tr>
<tr class="row-even"><td><p><strong>MotifMaker</strong></p></td>
<td><p>Java</p></td>
</tr>
<tr class="row-odd"><td><p><strong>MotifFinder</strong></p></td>
<td><p>SMRT, R</p></td>
</tr>
</tbody>
</table>
</section>
<section id="kineticstools">
<h3>kineticsTools<a class="headerlink" href="#kineticstools" title="Permalink to this headline">¶</a></h3>
<p>For the analyses we have decided to use
<a class="reference external" href="https://github.com/PacificBiosciences/kineticsTools/blob/master/doc/manual.rst">kineticsTools</a>
for two reasons:</p>
<ol class="arabic simple">
<li><p>It uses python as a programming language just like PacBio bioconda
and it is the programming language used to code sm-analysis.</p></li>
<li><p>It has a tool called ipdSummary that allows us to predict DNA
modifications without the need for a control sample.</p></li>
</ol>
</section>
<section id="ipdsummary">
<h3>ipdSummary<a class="headerlink" href="#ipdsummary" title="Permalink to this headline">¶</a></h3>
<p>ipdSummary allows us to predict sequence modifications using a
computational model (in-silico model). This tool allows us to detect
modifications that occur at the nucleotide level in sequences that
present m5C, m4C or m6A methylation, the last one being the best
detected with this tool.</p>
<p>IpdSummary has its filters:</p>
<ul class="simple">
<li><p><strong>Mapping Quality</strong> - By default the minimum mapping quality required
is 10, which implies that BLASR is 90% confident that the read is
mapped correctly. However, we also find many subreads that have a
mapping in more than one position that is sometimes at a great
distance from each other.</p></li>
<li><p><strong>Number of subreads per molecule</strong> - ipdSummary is effective on
molecules that have at least 20 mapped subreads.</p></li>
<li><p><strong>Length of the subreads</strong> - The PacBio output file can have
subframes of different sizes ranging from less than 50 bases to
thousands of bases.</p></li>
<li><p><strong>Multi-mapping</strong> - Some molecules may sometimes contain subreads
with different mapping positions and this affects the confidence of
the predicted modification in a position. In some cases,
multi-mapping occur in the region comprising the 0 positions of the
reference sequence</p></li>
</ul>
</section>
</section>
<section id="sm-analysis">
<h2>sm-analysis<a class="headerlink" href="#sm-analysis" title="Permalink to this headline">¶</a></h2>
<p>sm-analysis is based on the usage of different tools from
<a class="reference external" href="https://github.com/PacificBiosciences/pbbioconda">PacBio-Bioconda</a>
that allow us the analysis of information coming from PacBio sequencers.
The difference of sm-analysis with the other approach is the possibility
to analyze every PacBio bell (DNA molecule with adapters) separately, to
obtain their methylation status, circular consensus sequence, position
start, end, and GATCs (in case of having it) relative to the reference
sequence. sm-analysis uses
<a class="reference external" href="https://github.com/PacificBiosciences/kineticsTools/blob/master/doc/manual.rst">ipdSummary</a>
which is a pre-trained algorithm able to detect m6A methylations using
the Inter-Pulse Durations (IPD) values.</p>
<p>As we read previously there are different PacBio chemistries that
undergo periodic updates that create conflicts in data compatibility
during the data processing. To solve this, sm-analysis offers to the
user the option to change the chemistry to be used in the analysis.</p>
<p>We have also developed a bam-filter step that includes options to filter
our data. The following sequence of instructions can be used to filter
the aligned subreads.bam file:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 50%" />
<col style="width: 50%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-l 50 -q 254 -m 0 16 -R 0.9</strong></p></td>
<td><p>Minimum subread length of 50
bases Keep only high quality
mapping Molecules with 90% unique
mapping</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-m 0 16</strong></p></td>
<td><p>Keeps only uniquely mapped
subreads</p></td>
</tr>
<tr class="row-even"><td><p><strong>-r 20</strong></p></td>
<td><p>Only molecules with at least 20
subreads</p></td>
</tr>
</tbody>
</table>
<p>*In the table above, the filters are applied to an aligned subreads.bam
file using <a class="reference external" href="https://github.com/pacificbiosciences/blasr/">blasr</a>.
When using a different aligner like
<a class="reference external" href="https://github.com/PacificBiosciences/pbmm2/">pbmm2</a> you should
change the parameters for mapped subreads and mapping quality.</p>
</section>
<section id="pipeline">
<h2>Pipeline<a class="headerlink" href="#pipeline" title="Permalink to this headline">¶</a></h2>
<p>TBD</p>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Input</p></li>
<li><p>Execution</p></li>
<li><p>Output description</p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Guides</a><ul>
      <li>Previous: <a href="index.html" title="previous chapter">Guides</a></li>
      <li>Next: <a href="bam.html" title="next chapter">BAM format</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/indepth/sm-analysisdoc.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>