
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Summary Reports &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Methylation Reports" href="methylation_reports.html" />
    <link rel="prev" title="BAM format" href="bam.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="summary-reports">
<span id="summary-report"></span><h1>Summary Reports<a class="headerlink" href="#summary-reports" title="Permalink to this headline">¶</a></h1>
<dl class="field-list simple">
<dt class="field-odd">abstract</dt>
<dd class="field-odd"><p>The <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">sm-analysis</span></a> program summarizes some
statistics about the analysis with a file aptly called
<em>summary report</em>. This document describes its contents.</p>
</dd>
</dl>
<section id="introduction">
<h2>Introduction<a class="headerlink" href="#introduction" title="Permalink to this headline">¶</a></h2>
<p>Most of the data found in the <a class="reference internal" href="glossary.html#term-summary-report"><span class="xref std std-term">summary report</span></a> are self-explanatory,
but some points deserve special attention. The aim of this document is to
provide additional comments that help to understand the precise meaning
and/or format of the quantities reported in the sumamry report.</p>
</section>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this headline">¶</a></h2>
<p>The remaining subsections of this document map the sections of the
summary report. Not every quantity that appears in a summary report is
described below, though. Only the points that require additional
remarks are documented here.</p>
<section id="overview">
<h3>Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h3>
<p>This section contains general information about the particular run of
<a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">Single molecule analysis with the sm-analysis program</span></a> that produced the summary report in question. The
command given to produce the analysis can be obtained by joining
<strong>Program name</strong> to <strong>Program options</strong>. The date of the analysis
contains the date <em>and</em> time when the <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">sm-analysis</span></a>
program started, and is given in the ISO 8601 format with a precision
of minutes, following this pattern: <code class="docutils literal notranslate"><span class="pre">YYYY-MM-DDTHH:MM</span></code>.</p>
</section>
<section id="result-filenames">
<h3>Result filenames<a class="headerlink" href="#result-filenames" title="Permalink to this headline">¶</a></h3>
<p>The <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">Single molecule analysis with the sm-analysis program</span></a> program produces different output files. Local
links to those files are provided in this section. See
<a class="reference internal" href="methylation_reports.html#methylation-report-format"><span class="std std-ref">Methylation Reports</span></a>, <a class="reference internal" href="raw_detections.html#raw-detections-format"><span class="std std-ref">Raw Detections</span></a> and
<a class="reference external" href="https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md">GFF format</a> for a description of their contents. Notice that actually
the link to the <strong>GFF</strong> file is pointing to a <em>concatenation</em> of all
GFF files produced by <a class="reference internal" href="kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a>.</p>
</section>
<section id="input-files">
<h3>Input files<a class="headerlink" href="#input-files" title="Permalink to this headline">¶</a></h3>
<p>This section contains details about the input files passed in to
the <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">Single molecule analysis with the sm-analysis program</span></a>: the BAM file and the <a class="reference internal" href="glossary.html#term-reference"><span class="xref std std-term">reference</span></a> file.</p>
<section id="bam-file">
<h4>BAM File<a class="headerlink" href="#bam-file" title="Permalink to this headline">¶</a></h4>
<p>Apart from the filename and the size in bytes, the report includes
the <a class="reference internal" href="glossary.html#term-MD5-checksum"><span class="xref std std-term">MD5 checksum</span></a> of the <em>full</em> file, as it was at the time of the
analysis, and the <a class="reference internal" href="glossary.html#term-MD5-checksum"><span class="xref std std-term">MD5 checksum</span></a> of the <em>body</em> of that file. The
reason for adding both is that the header of the BAM file could be altered
by some tool while the body is preserved intact. In a normal case, the
full checksum would be enough. But if, by some reason, the header of the BAM
file was modified after the analysis, having the body checksum at hand
could be helpful.</p>
</section>
<section id="reference">
<h4>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h4>
<p>Apart from the filename, a <a class="reference internal" href="glossary.html#term-FASTA"><span class="xref std std-term">FASTA</span></a> file contains a reference
<em>name</em>, which is given under <em>Reference name</em>. The length of the sequence
itself is also given (in terms of base pairs) and the <a class="reference internal" href="glossary.html#term-MD5-checksum"><span class="xref std std-term">MD5 checksum</span></a>
of the (normalized to upper case) sequence is included as well.</p>
</section>
</section>
<section id="molecules-subreads">
<h3>Molecules/subreads<a class="headerlink" href="#molecules-subreads" title="Permalink to this headline">¶</a></h3>
<p>Basic statistics about the body of the input BAM file. All the quantities
in this section are given for <a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a> and for
<a class="reference internal" href="glossary.html#term-subread"><span class="xref std std-term">subreads</span></a>.</p>
<p><strong>Initial</strong> contains the molecules/subreads counts in the input BAM file.
All percentages in this section refer to these quantities. In particular,
the statistics about <a class="reference internal" href="glossary.html#term-subread"><span class="xref std std-term">subreads</span></a> refer to what is found in
the input BAM file.</p>
<p><strong>Used in aligned CCS BAM</strong> refers to how many molecules/subreads from the
input BAM file are <em>also</em> in <em>any</em> of the
<a class="reference internal" href="glossary.html#term-alignment-variant"><span class="xref std std-term">alignment variants</span></a> of the CCS BAM file.
Each molecule from the input BAM file will be assigned, <em>at most</em>, to one
<a class="reference internal" href="glossary.html#term-variant"><span class="xref std std-term">variant</span></a>, i.e. even if a molecule is found in all the alignment
variants, it (and its subreads) will not be counted more than once.</p>
<p><strong>DNA mismatch discards</strong> gives us the numbers corresponding to molecules
for which the sequence provided by the aligned CCS BAM file
<em>does not match</em> the reference at the position given also by the aligned
CCS BAM file.</p>
<p><strong>Filtered out</strong> contains statistics about the molecules discarded by the
filters applied by <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">Single molecule analysis with the sm-analysis program</span></a> to the input BAM file.</p>
<p><strong>Faulty (with processing error)</strong> are <a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a> whose
corresponding single molecule BAM file had problems when it was processed by
either <a class="reference internal" href="pbindex.html#about-pbindex"><span class="std std-ref">pbindex</span></a> or <a class="reference internal" href="kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a>.
The details about what went wrong are given in the output displayed in the
screen <em>if</em> the <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">sm-analysis</span></a> program was executed in
<em>verbose mode</em>, i.e. <a class="reference internal" href="../usage/sm_analysis.html#cmdoption-sm-analysis-v"><code class="xref std std-option docutils literal notranslate"><span class="pre">sm-analysis</span> <span class="pre">-v</span></code></a>. Faulty molecules
are exceptional. A normal <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">Single molecule analysis with the sm-analysis program</span></a> is expected to have not even
a single faulty molecule.</p>
<p>The <strong>In methylation report…</strong> row contains what fraction of the initial
data ends up in the <a class="reference internal" href="methylation_reports.html#methylation-report-format"><span class="std std-ref">methylation report</span></a>.
These quantities are further splitted into which <a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a>
and <a class="reference internal" href="glossary.html#term-subread"><span class="xref std std-term">subreads</span></a> contain/do not contain GATCs in the rows
<strong>…only with GATCs</strong> and <strong>…only without GATCs</strong>, respectively.
Of course, <strong>…only with GATCs</strong> and <strong>…only without GATCs</strong> add up to
<strong>In methylation report…</strong>.</p>
<p>The quantities in <strong>Used in aligned CCS BAM</strong> should be the sum of the
corresponding numbers found in the following rows: <strong>DNA mismatch discards</strong>,
<strong>Filtered out</strong>, <strong>Faulty (with processing error)</strong> and
<strong>In methylation report…</strong>. The difference between <strong>Initial</strong> and
<strong>Used in aligned CCS BAM</strong> is due to <a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a> that
do not <em>survive</em> the CCS and alignment processes. Since the positioning
of the molecules is essential to determine the location of the detected
methylation, the aligned CCS BAM file is taken as a baseline.</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.0: </span><strong>Faulty (with processing error)</strong> added to the summary report.</p>
</div>
</section>
<section id="sequencing-position-coverage">
<h3>Sequencing Position Coverage<a class="headerlink" href="#sequencing-position-coverage" title="Permalink to this headline">¶</a></h3>
<p><strong>Positions covered by molecules in the BAM file</strong> refer to the position
coverage provided by all the <a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a> in all the
<a class="reference internal" href="glossary.html#term-alignment-variant"><span class="xref std std-term">alignment variants</span></a> of the CCS BAM file.
Obviously the percentages in this section refer to the length of the
<a class="reference internal" href="glossary.html#term-reference"><span class="xref std std-term">reference</span></a>.</p>
</section>
<section id="gatcs">
<h3>GATCs<a class="headerlink" href="#gatcs" title="Permalink to this headline">¶</a></h3>
<p>Again, as in the section <strong>Sequencing Position Coverage</strong>, the
<strong>Number of GATCs identified in the BAM file</strong> include all the
<a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a> in the merged
<a class="reference internal" href="glossary.html#term-alignment-variant"><span class="xref std std-term">alignment variants</span></a> of the CCS BAM file.</p>
</section>
<section id="methylations">
<h3>Methylations<a class="headerlink" href="#methylations" title="Permalink to this headline">¶</a></h3>
<p>As the summary report itself declares, in this section the <em>individual</em>
methylation detections are considered. Any GATC in the <a class="reference internal" href="glossary.html#term-reference"><span class="xref std std-term">reference</span></a>
can be detected multiple times: several <a class="reference internal" href="glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a>
can cover the same GATC, but each molecule will be analyzed independently
by <a class="reference internal" href="../usage/sm_analysis.html#sm-analysis"><span class="std std-ref">sm-analysis</span></a>. That is why
<strong>Total number of GATCs in all the analyzed molecules</strong> does not
(and must not!) agree with the numbers in the <strong>GATCs</strong> section.</p>
</section>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../usage/index.html">Usage</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Guides</a><ul>
      <li>Previous: <a href="bam.html" title="previous chapter">BAM format</a></li>
      <li>Next: <a href="methylation_reports.html" title="next chapter">Methylation Reports</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/indepth/summary_reports.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>