
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pacbio_data_processing.sm_analysis &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css" />
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for pacbio_data_processing.sm_analysis</h1><div class="highlight"><pre>
<span></span><span class="c1">#######################################################################</span>
<span class="c1">#</span>
<span class="c1"># Copyright (C) 2020-2022 David Palao</span>
<span class="c1"># Copyright (C) 2020 David Velázquez</span>
<span class="c1">#</span>
<span class="c1"># This file is part of PacBio data processing.</span>
<span class="c1">#</span>
<span class="c1">#  PacBioDataProcessing is free software: you can redistribute it and/or modify</span>
<span class="c1">#  it under the terms of the GNU General Public License as published by</span>
<span class="c1">#  the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1">#  (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1">#  PacBio data processing is distributed in the hope that it will be useful,</span>
<span class="c1">#  but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1">#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1">#  GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1">#  You should have received a copy of the GNU General Public License</span>
<span class="c1">#  along with PacBioDataProcessing. If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>
<span class="c1">#</span>
<span class="c1">#######################################################################</span>

<span class="sd">&quot;&quot;&quot;This module contains the high level functions necessary to run</span>
<span class="sd">the &#39;Single Molecule Analysis&#39; on an input BAM file.&quot;&quot;&quot;</span>


<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">from</span> <span class="nn">tempfile</span> <span class="kn">import</span> <span class="n">TemporaryDirectory</span>
<span class="kn">import</span> <span class="nn">shutil</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">from</span> <span class="nn">itertools</span> <span class="kn">import</span> <span class="n">tee</span><span class="p">,</span> <span class="n">chain</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span><span class="p">,</span> <span class="n">Counter</span>
<span class="kn">from</span> <span class="nn">collections.abc</span> <span class="kn">import</span> <span class="n">Iterable</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">cached_property</span>

<span class="kn">from</span> <span class="nn">.bam_utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">split_bam_file_in_molecules</span><span class="p">,</span> <span class="n">gen_index_single_molecule_bams</span><span class="p">,</span> <span class="n">join_gffs</span><span class="p">,</span>
    <span class="n">Molecule</span><span class="p">,</span> <span class="n">count_subreads_per_molecule</span><span class="p">,</span> <span class="n">estimate_max_mapping_quality</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">.bam</span> <span class="kn">import</span> <span class="n">BamFile</span>
<span class="kn">from</span> <span class="nn">.logs</span> <span class="kn">import</span> <span class="n">config_logging</span>
<span class="kn">from</span> <span class="nn">.ui.cl</span> <span class="kn">import</span> <span class="n">parse_cl_sm_analysis</span> <span class="k">as</span> <span class="n">parse_cl</span>
<span class="kn">from</span> <span class="nn">.parameters</span> <span class="kn">import</span> <span class="n">SingleMoleculeAnalysisParameters</span>
<span class="kn">from</span> <span class="nn">.ipd</span> <span class="kn">import</span> <span class="n">multi_ipd_summary</span>
<span class="kn">from</span> <span class="nn">.constants</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">PBMM2_PREF</span><span class="p">,</span> <span class="n">BLASR_PREF</span><span class="p">,</span> <span class="n">PI_SHIFTED_PREF</span><span class="p">,</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">,</span>
    <span class="n">STRAIGHT_VARIANT</span><span class="p">,</span> <span class="n">MOLECULES_FILE_SUFFIX</span><span class="p">,</span> <span class="n">MAPQ_MIN_LINES</span><span class="p">,</span> <span class="n">MAPQ_MAX_LINES</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">.errors</span> <span class="kn">import</span> <span class="n">high_level_handler</span><span class="p">,</span> <span class="n">SMAPipelineError</span><span class="p">,</span> <span class="n">SMAMergeError</span>
<span class="kn">from</span> <span class="nn">.external</span> <span class="kn">import</span> <span class="n">Pbmm2</span><span class="p">,</span> <span class="n">Blasr</span><span class="p">,</span> <span class="n">CCS</span>

<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">DNASeq</span><span class="p">,</span> <span class="n">Partition</span><span class="p">,</span> <span class="n">pishift_back_positions_in_gff</span><span class="p">,</span> <span class="n">make_partition_prefix</span><span class="p">,</span>
    <span class="n">try_computations_with_variants_until_done</span><span class="p">,</span> <span class="n">merge_files</span><span class="p">,</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">.filters</span> <span class="kn">import</span> <span class="n">cleanup_molecules</span>
<span class="kn">from</span> <span class="nn">.summary</span> <span class="kn">import</span> <span class="n">SummaryReport</span>
<span class="kn">from</span> <span class="nn">.types</span> <span class="kn">import</span> <span class="n">PathOrStr</span>
<span class="kn">from</span> <span class="nn">.methylation</span> <span class="kn">import</span> <span class="n">MethylationReport</span>


<span class="n">MODIFIED_BASE_STR</span> <span class="o">=</span> <span class="s2">&quot;modified_base&quot;</span>
<span class="n">MISSING_ALIGNED_CCS_MSG</span> <span class="o">=</span> <span class="p">(</span>
    <span class="s2">&quot;The methylation analysis requires aligned CCS files --for all variants--&quot;</span>
    <span class="s2">&quot; to proceed. Trying to get them...&quot;</span>
<span class="p">)</span>
<span class="n">MISSING_CCS_MSG</span> <span class="o">=</span> <span class="p">(</span>
    <span class="s2">&quot;Aligned CCS file cannot be produced without CCS file. &quot;</span>
    <span class="s2">&quot;Trying to produce it...&quot;</span>
<span class="p">)</span>
<span class="n">CCS_FILE_FOUND_AND_COMPUTATION_SKIPPED_MSG</span> <span class="o">=</span> <span class="p">(</span>
    <span class="s2">&quot;CCS file &#39;</span><span class="si">{ccs_filename}</span><span class="s2">&#39; found. Skipping its computation.&quot;</span>
<span class="p">)</span>


<div class="viewcode-block" id="create_raw_detections_file"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.create_raw_detections_file">[docs]</a><span class="k">def</span> <span class="nf">create_raw_detections_file</span><span class="p">(</span>
        <span class="n">gffs</span><span class="p">:</span> <span class="n">Iterable</span><span class="p">[</span><span class="n">PathOrStr</span><span class="p">],</span>
        <span class="n">detections_filename</span><span class="p">:</span> <span class="n">PathOrStr</span><span class="p">,</span>
        <span class="n">modification_types</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="nb">str</span><span class="p">]):</span>
    <span class="sd">&quot;&quot;&quot;Function in charge of creating the *raw detections* file.</span>
<span class="sd">    Starting from a set of .gff files, a csv file (delimiter=&quot;,&quot;),</span>
<span class="sd">    the *raw detections* file, is saved with the following columns:</span>

<span class="sd">      - mol id: taken from each gff filename (e.g. &#39;a.b.c.gff&#39; -&gt;</span>
<span class="sd">        mol id: &#39;b&#39;);</span>
<span class="sd">      - modtype: column number 3 (idx: 2) of the gffs (feature type)</span>
<span class="sd">        (e.g. &#39;m6A&#39;);</span>
<span class="sd">      - GATC position: column number 5 (idx: 4) of each gff which</span>
<span class="sd">        corresponds to the &#39;end coordinate of the feature&#39; in the</span>
<span class="sd">        GFF3 standard;</span>
<span class="sd">      - score of the feature: column number 6 (idx: 5); floating point</span>
<span class="sd">        (:ref:`Phred-transformed pvalue &lt;phred-transformed-scores&gt;`</span>
<span class="sd">        that a kinetic deviation exists at this position)</span>
<span class="sd">      - strand: strand of the feature. It can be +, - with obvious</span>
<span class="sd">        meanings. It can also be ? (meaning unknown) or . (for non</span>
<span class="sd">        stranded features)</span>

<span class="sd">    There are more columns. Although their number is not fixed by</span>
<span class="sd">    this function, in practice they are 4 in the case of a</span>
<span class="sd">    detected modification. In that case these 4 last columns correspond</span>
<span class="sd">    to the values given in the &#39;attributes&#39; column of the gffs</span>
<span class="sd">    (col 9; idx 8). For example, given the following attributes column::</span>

<span class="sd">      coverage=134;context=TCA...;IPDRatio=3.91;identificationQv=228</span>

<span class="sd">    we would get the following 4 &#39;extra&#39; columns in our *raw detections*</span>
<span class="sd">    file::</span>

<span class="sd">      134,TCA...,3.91,228</span>

<span class="sd">    and this is exactly what happens with the m6A modification type.</span>
<span class="sd">    Notice that the value of identificationQV is, again, a</span>
<span class="sd">    :ref:`phred transformed probability &lt;phred-transformed-scores&gt;` of</span>
<span class="sd">    having a detection. See eq. (8) in [1]</span>

<span class="sd">    Parsing: All the lines starting by &#39;#&#39; in the gff files are</span>
<span class="sd">    ignored. The format of the gff file is GFF3:</span>
<span class="sd">    https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md</span>

<span class="sd">    [1]: &quot;Detection and Identification of Base Modifications with Single</span>
<span class="sd">    Molecule Real-Time Sequencing Data&quot;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># wouldn&#39;t it be good if before writing, a dictionary is created</span>
    <span class="c1"># with the data already written to the file?</span>
    <span class="c1"># This would be helpful to avoid double writing lines (see isse #17)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">detections_filename</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">csvfile</span><span class="p">:</span>
        <span class="n">csvwriter</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">writer</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;,&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">gff</span> <span class="ow">in</span> <span class="n">gffs</span><span class="p">:</span>
            <span class="n">molecule</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">2</span><span class="p">]</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gff</span><span class="p">)</span> <span class="k">as</span> <span class="n">gff_file</span><span class="p">:</span>
                <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">gff_file</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                        <span class="k">continue</span>
                    <span class="n">pieces</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">pieces</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="ow">in</span> <span class="n">modification_types</span><span class="p">:</span>
                        <span class="n">extra</span> <span class="o">=</span> <span class="p">[</span><span class="n">_</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;=&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">pieces</span><span class="p">[</span><span class="mi">8</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)]</span>
                        <span class="n">new_line</span> <span class="o">=</span> <span class="p">[</span><span class="n">molecule</span><span class="p">,</span> <span class="n">pieces</span><span class="p">[</span><span class="mi">2</span><span class="p">]]</span> <span class="o">+</span> <span class="n">pieces</span><span class="p">[</span><span class="mi">4</span><span class="p">:</span><span class="mi">7</span><span class="p">]</span> <span class="o">+</span> <span class="n">extra</span>
                        <span class="n">csvwriter</span><span class="o">.</span><span class="n">writerow</span><span class="p">(</span><span class="n">new_line</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Raw detections file &#39;</span><span class="si">{</span><span class="n">detections_filename</span><span class="si">}</span><span class="s2">&#39; created&quot;</span><span class="p">)</span></div>


<div class="viewcode-block" id="restore_old_run"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.restore_old_run">[docs]</a><span class="k">def</span> <span class="nf">restore_old_run</span><span class="p">(</span><span class="n">old_path</span><span class="p">,</span> <span class="n">new_path</span><span class="p">):</span>  <span class="c1">#</span>
    <span class="n">keep_old</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">old_path</span><span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isdir</span><span class="p">(</span><span class="n">new_path</span><span class="p">):</span>
        <span class="k">for</span> <span class="n">fn</span> <span class="ow">in</span> <span class="n">os</span><span class="o">.</span><span class="n">listdir</span><span class="p">(</span><span class="n">old_path</span><span class="p">):</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">shutil</span><span class="o">.</span><span class="n">move</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">old_path</span><span class="o">/</span><span class="n">fn</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="n">new_path</span><span class="p">))</span>
            <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error moving &#39;</span><span class="si">{</span><span class="n">fn</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">keep_old</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">keep_old</span><span class="p">:</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">rmtree</span><span class="p">(</span><span class="n">old_path</span><span class="p">,</span> <span class="n">ignore_errors</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Error removing &#39;</span><span class="si">{</span><span class="n">old_path</span><span class="si">}</span><span class="s2">&#39;: </span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span></div>


<span class="c1"># The next should maybe be absorved by MethylationReport, or in</span>
<span class="c1"># bam_utils and logging factored out. (?)</span>
<div class="viewcode-block" id="map_molecules_with_highest_sim_ratio"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.map_molecules_with_highest_sim_ratio">[docs]</a><span class="k">def</span> <span class="nf">map_molecules_with_highest_sim_ratio</span><span class="p">(</span>
        <span class="n">bam_file_name</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">PathOrStr</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Given the path to a bam file, it returns a dictionary, whose</span>
<span class="sd">    keys are mol ids (ints) and the values are the corresponding</span>
<span class="sd">    Molecules.</span>
<span class="sd">    If multiple lines in the given BAM file share the mol id, only</span>
<span class="sd">    the first line found with the highest similarity ratio (computed from</span>
<span class="sd">    the cigar) is chosen: if multiple lines share the molecule ID and</span>
<span class="sd">    the highest similarity ratio (say, 1), ONLY the first one is taken,</span>
<span class="sd">    irrespective of other factors.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mols</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">bam_file_name</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">body</span><span class="p">:</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">molecule_id</span><span class="p">)</span>
        <span class="n">mol</span> <span class="o">=</span> <span class="n">Molecule</span><span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="n">bam_file_name</span><span class="p">,</span> <span class="n">line</span><span class="p">)</span>
        <span class="n">old_mol</span> <span class="o">=</span> <span class="n">mols</span><span class="o">.</span><span class="n">setdefault</span><span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="n">mol</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mol</span> <span class="o">!=</span> <span class="n">old_mol</span><span class="p">:</span>
            <span class="n">oldc</span> <span class="o">=</span> <span class="n">old_mol</span><span class="o">.</span><span class="n">cigar</span>
            <span class="n">newc</span> <span class="o">=</span> <span class="n">mol</span><span class="o">.</span><span class="n">cigar</span>
            <span class="k">if</span> <span class="n">newc</span><span class="o">.</span><span class="n">sim_ratio</span> <span class="o">&gt;</span> <span class="n">oldc</span><span class="o">.</span><span class="n">sim_ratio</span><span class="p">:</span>
                <span class="n">mols</span><span class="p">[</span><span class="n">mol_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mol</span>
    <span class="k">return</span> <span class="n">mols</span></div>


<div class="viewcode-block" id="generate_CCS_file"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.generate_CCS_file">[docs]</a><span class="k">def</span> <span class="nf">generate_CCS_file</span><span class="p">(</span>
        <span class="n">ccs</span><span class="p">:</span> <span class="n">CCS</span><span class="p">,</span> <span class="n">in_bam</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span> <span class="n">ccs_bam_file</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Idempotent computation of the Circular Consensus Sequence (CCS)</span>
<span class="sd">    version of the passed in ``in_bam`` file done with passed-in ``ccs``</span>
<span class="sd">    object.</span>

<span class="sd">    :return: the CCS bam file, if it is there, or ``None`` if if could</span>
<span class="sd">             not be computed (yet).</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">ccs_bam_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
            <span class="n">CCS_FILE_FOUND_AND_COMPUTATION_SKIPPED_MSG</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">ccs_filename</span><span class="o">=</span><span class="n">ccs_bam_file</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="n">ccs_bam_file</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="n">MISSING_CCS_MSG</span><span class="p">)</span>
        <span class="n">ccs_retcode</span> <span class="o">=</span> <span class="n">ccs</span><span class="p">(</span><span class="n">in_bam</span><span class="p">,</span> <span class="n">ccs_bam_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">ccs_retcode</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">elif</span> <span class="n">ccs_bam_file</span><span class="o">.</span><span class="n">exists</span><span class="p">():</span>
            <span class="n">result</span> <span class="o">=</span> <span class="n">ccs_bam_file</span>
            <span class="k">if</span> <span class="n">ccs_retcode</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;Although the file &#39;</span><span class="si">{</span><span class="n">ccs_bam_file</span><span class="si">}</span><span class="s2">&#39; has been generated, &quot;</span>
                    <span class="s2">&quot;there was an error.&quot;</span>
                <span class="p">)</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span>
                    <span class="s2">&quot;It is advisable to check the correctness of the generated&quot;</span>
                    <span class="s2">&quot; ccs file.&quot;</span>
                <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">critical</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;CCS BAM file &#39;</span><span class="si">{</span><span class="n">ccs_bam_file</span><span class="si">}</span><span class="s2">&#39; could not be produced.&quot;</span>
            <span class="p">)</span>
            <span class="k">raise</span> <span class="n">SMAPipelineError</span><span class="p">(</span>
                <span class="s2">&quot;The Single Molecule Analysis cannot proceed without a CCS BAM&quot;</span>
                <span class="s2">&quot; file. Aborting.&quot;</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="SingleMoleculeAnalysis"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis">[docs]</a><span class="k">class</span> <span class="nc">SingleMoleculeAnalysis</span><span class="p">:</span>
    <span class="n">_VARIANTS</span> <span class="o">=</span> <span class="p">(</span><span class="n">STRAIGHT_VARIANT</span><span class="p">,</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">)</span>
    <span class="n">_ALIGNED_GENERIC_NAME</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">STRAIGHT_VARIANT</span><span class="p">:</span> <span class="s2">&quot;aligned&quot;</span><span class="p">,</span>
        <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">PI_SHIFTED_PREF</span><span class="si">}</span><span class="s2"> aligned&quot;</span><span class="p">,</span>
    <span class="p">}</span>
    <span class="n">_ALIGNED_FILE_PREFIX</span> <span class="o">=</span> <span class="p">{</span>
        <span class="n">STRAIGHT_VARIANT</span><span class="p">:</span> <span class="s2">&quot;</span><span class="si">{aligner_pref}</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">PI_SHIFTED_PREF</span><span class="si">}</span><span class="s2">.</span><span class="se">{{</span><span class="s2">aligner_pref</span><span class="se">}}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">}</span>

<div class="viewcode-block" id="SingleMoleculeAnalysis.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">parameters</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_VARIANTS</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span> <span class="o">=</span> <span class="n">parameters</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">input_bam_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">CCS_bam_file</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">CCS_bam_file</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_create_references</span><span class="p">()</span>
        <span class="n">aligned_input_bam_file</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">STRAIGHT_VARIANT</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">aligned_ccs_bam_file</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">STRAIGHT_VARIANT</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
            <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span> <span class="kc">None</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;input&quot;</span><span class="p">:</span> <span class="n">aligned_input_bam_file</span><span class="p">,</span>
            <span class="s2">&quot;ccs&quot;</span><span class="p">:</span> <span class="n">aligned_ccs_bam_file</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_set_aligner</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ccs</span> <span class="o">=</span> <span class="n">CCS</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ccs_path</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_input_bam_aligned</span><span class="p">()</span>
        <span class="c1"># _init_summary requires aligned_bams:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_init_summary</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ensure_ccs_bam_aligned</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_set_tasks</span><span class="p">()</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">CCS_bam_file</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;It produces a Circular Consensus Sequence (CCS) version of</span>
<span class="sd">        the input BAM file and returns its name. It uses</span>
<span class="sd">        :py:func:`generate_CCS_file` to generate the file.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">try_computations_with_variants_until_done</span><span class="p">(</span>
            <span class="n">generate_CCS_file</span><span class="p">,</span> <span class="p">(</span><span class="kc">None</span><span class="p">,),</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">ccs</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_CCS_bam_file</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_CCS_bam_file</span>

    <span class="nd">@CCS_bam_file</span><span class="o">.</span><span class="n">setter</span>
    <span class="k">def</span> <span class="nf">CCS_bam_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">value</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">Path</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Sets the underlying attribute to ``value`` if the given value</span>
<span class="sd">        is not ``None``, or constructs a ``Path`` instance from</span>
<span class="sd">        ``self.input_bam_file`` prepending ``ccs.`` to the name if the</span>
<span class="sd">        given value is ``None``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">value</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">infilename</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">)</span>
            <span class="n">base</span> <span class="o">=</span> <span class="n">infilename</span><span class="o">.</span><span class="n">name</span>
            <span class="n">new_base</span> <span class="o">=</span> <span class="s2">&quot;ccs.&quot;</span> <span class="o">+</span> <span class="n">base</span>
            <span class="n">value</span> <span class="o">=</span> <span class="n">infilename</span><span class="o">.</span><span class="n">parent</span><span class="o">/</span><span class="n">new_base</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_CCS_bam_file</span> <span class="o">=</span> <span class="n">value</span>

<div class="viewcode-block" id="SingleMoleculeAnalysis._create_references"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._create_references">[docs]</a>    <span class="k">def</span> <span class="nf">_create_references</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        DNA reference sequences are created here. The &#39;true&#39; reference must</span>
<span class="sd">        exist as fasta beforehand, with its index. A π-shifted reference</span>
<span class="sd">        is created from the original one. Its index is also made.</span>

<span class="sd">        This method sets two attributes which are, both, mappings with</span>
<span class="sd">        two keys (&#39;straight&#39; and &#39;pi-shifted&#39;) and values as follows:</span>

<span class="sd">        - reference: the values are DNASeq objects</span>
<span class="sd">        - fasta: the values are Path objects</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">straight_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">fasta</span><span class="p">)</span>
        <span class="n">straight</span> <span class="o">=</span> <span class="n">DNASeq</span><span class="o">.</span><span class="n">from_fasta</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">straight_path</span><span class="p">))</span>
        <span class="n">pi_shifted</span> <span class="o">=</span> <span class="n">straight</span><span class="o">.</span><span class="n">pi_shifted</span><span class="p">()</span>
        <span class="n">pi_shifted_path</span> <span class="o">=</span> <span class="n">straight_path</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">PI_SHIFTED_PREF</span><span class="si">}</span><span class="s2">.&quot;</span><span class="o">+</span><span class="n">straight_path</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">pi_shifted</span><span class="o">.</span><span class="n">write_fasta</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">pi_shifted_path</span><span class="p">))</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">STRAIGHT_VARIANT</span><span class="p">:</span> <span class="n">straight</span><span class="p">,</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span> <span class="n">pi_shifted</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">STRAIGHT_VARIANT</span><span class="p">:</span> <span class="n">straight_path</span><span class="p">,</span>
            <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span> <span class="n">pi_shifted_path</span>
        <span class="p">}</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._init_summary"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._init_summary">[docs]</a>    <span class="k">def</span> <span class="nf">_init_summary</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method creates an instance of ``SummaryReport`` and sets</span>
<span class="sd">        an attribute with it.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">prefix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">is_proper</span><span class="p">:</span>
            <span class="n">prefix</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="p">)</span><span class="o">+</span><span class="s2">&quot;.&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span> <span class="o">=</span> <span class="n">SummaryReport</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">STRAIGHT_VARIANT</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">STRAIGHT_VARIANT</span><span class="p">],</span>
            <span class="n">figures_prefix</span><span class="o">=</span><span class="n">prefix</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._set_aligner"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._set_aligner">[docs]</a>    <span class="k">def</span> <span class="nf">_set_aligner</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method decides what aligner to use, sets an attribute with</span>
<span class="sd">        it and sets the prefixes (used in log messages) accordingly.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Refactor hint: move next two to DEFAULT_ values somewhere</span>
        <span class="n">Aligner</span> <span class="o">=</span> <span class="n">Pbmm2</span>
        <span class="n">pref</span> <span class="o">=</span> <span class="n">PBMM2_PREF</span>
        <span class="n">aligner_path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">aligner_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">common_aligner_options</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">use_blasr_aligner</span><span class="p">:</span>
            <span class="n">Aligner</span> <span class="o">=</span> <span class="n">Blasr</span>
            <span class="n">pref</span> <span class="o">=</span> <span class="n">BLASR_PREF</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">common_aligner_options</span><span class="p">[</span><span class="s2">&quot;nprocs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">nprocs_blasr</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aligner</span> <span class="o">=</span> <span class="n">Aligner</span><span class="p">(</span><span class="n">aligner_path</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
            <span class="n">raw_pref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_FILE_PREFIX</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
            <span class="n">fpref</span> <span class="o">=</span> <span class="n">raw_pref</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">aligner_pref</span><span class="o">=</span><span class="n">pref</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_FILE_PREFIX</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span> <span class="o">=</span> <span class="n">fpref</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._ensure_input_bam_aligned"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._ensure_input_bam_aligned">[docs]</a>    <span class="k">def</span> <span class="nf">_ensure_input_bam_aligned</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Main check point for aligned input bam files: this method calls</span>
<span class="sd">        whatever is necessary to ensure that the input bam is aligned,</span>
<span class="sd">        which means: normal (straight) alignment and π-shifted alignment.</span>

<span class="sd">        Warning! The method tries to find a pi-shifted aligned BAM if the</span>
<span class="sd">        input is aligned based on whether</span>

<span class="sd">        1. a file with suitable filename is found, and</span>
<span class="sd">        2. it is aligned.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">inbam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">inbam</span><span class="o">.</span><span class="n">is_aligned</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;The input BAM is aligned&quot;</span><span class="p">)</span>
            <span class="n">input_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">)</span>
            <span class="n">input_name</span> <span class="o">=</span> <span class="n">input_path</span><span class="o">.</span><span class="n">name</span>
            <span class="n">pi_shifted_name</span> <span class="o">=</span> <span class="n">input_name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_FILE_PREFIX</span><span class="p">[</span><span class="n">STRAIGHT_VARIANT</span><span class="p">],</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">pi_shifted_name</span> <span class="o">=</span> <span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_FILE_PREFIX</span><span class="p">[</span><span class="n">PI_SHIFTED_VARIANT</span><span class="p">]</span> <span class="o">+</span> <span class="n">pi_shifted_name</span>
            <span class="p">)</span>
            <span class="n">candidate</span> <span class="o">=</span> <span class="n">input_path</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="n">pi_shifted_name</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">PI_SHIFTED_VARIANT</span><span class="p">]</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">candidate</span><span class="p">)</span>
            <span class="n">aligned_generic_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_GENERIC_NAME</span><span class="p">[</span>
                <span class="n">PI_SHIFTED_VARIANT</span><span class="p">]</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_exists_pi_shifted_variant_from_aligned_input</span><span class="p">():</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;...a possible </span><span class="si">{</span><span class="n">aligned_generic_name</span><span class="si">}</span><span class="s2"> version of the &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;input BAM was found: &#39;</span><span class="si">{</span><span class="n">candidate</span><span class="si">}</span><span class="s2">&#39;. It will &quot;</span>
                    <span class="s2">&quot;be used.&quot;</span>
                <span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">PI_SHIFTED_VARIANT</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;...but no </span><span class="si">{</span><span class="n">aligned_generic_name</span><span class="si">}</span><span class="s2"> version of the &quot;</span>
                    <span class="s2">&quot;input BAM was found.&quot;</span>
                <span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_disable_pi_shifted_analysis</span><span class="p">()</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">STRAIGHT_VARIANT</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;The input BAM is NOT aligned&quot;</span><span class="p">)</span>
            <span class="n">bam_type</span> <span class="o">=</span> <span class="s2">&quot;input&quot;</span>
            <span class="n">try_computations_with_variants_until_done</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_align_bam_if_no_candidate_found</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">,</span> <span class="n">inbam</span><span class="p">,</span> <span class="n">bam_type</span>
            <span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._exists_pi_shifted_variant_from_aligned_input"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._exists_pi_shifted_variant_from_aligned_input">[docs]</a>    <span class="k">def</span> <span class="nf">_exists_pi_shifted_variant_from_aligned_input</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        It checks that the expected pi-shifted aligned file exists</span>
<span class="sd">        and is an aligned BAM file.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># pi_shifted_name = input_name.replace(</span>
        <span class="c1">#     self._ALIGNED_FILE_PREFIX[&quot;straight&quot;],</span>
        <span class="c1">#     self._ALIGNED_FILE_PREFIX[&quot;pi-shifted&quot;], 1)</span>
        <span class="c1"># pi_shifted_candidate = input_path.with_name(pi_shifted_name)</span>
        <span class="c1"># pi_shifted_candidate.exists()</span>
        <span class="c1"># BamFile(pi_shifted_candidate)</span>
        <span class="n">pi_shifted_candidate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">PI_SHIFTED_VARIANT</span><span class="p">]</span>
        <span class="n">pi_shifted_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">pi_shifted_candidate</span><span class="p">)</span>
        <span class="n">pi_shifted_bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">pi_shifted_candidate</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">pi_shifted_path</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span> <span class="n">pi_shifted_bam</span><span class="o">.</span><span class="n">is_aligned</span><span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._disable_pi_shifted_analysis"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._disable_pi_shifted_analysis">[docs]</a>    <span class="k">def</span> <span class="nf">_disable_pi_shifted_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        If the pi-shifted analysis cannot be carried out, it is disabled</span>
<span class="sd">        with this method.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s2">&quot;...therefore the pi-shifted analysis is disabled&quot;</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">variants</span> <span class="o">=</span> <span class="p">(</span><span class="n">STRAIGHT_VARIANT</span><span class="p">,)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._align_bam_if_no_candidate_found"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._align_bam_if_no_candidate_found">[docs]</a>    <span class="k">def</span> <span class="nf">_align_bam_if_no_candidate_found</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">inbam</span><span class="p">:</span> <span class="n">BamFile</span><span class="p">,</span> <span class="n">bam_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
            <span class="n">variant</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="n">STRAIGHT_VARIANT</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">str</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Auxiliary method used by ``_ensure_input_bam_aligned`` and by</span>
<span class="sd">        ``_ensure_ccs_bam_aligned``. Given a ``bam_type`` (among ``input``</span>
<span class="sd">        and ``ccs``) and a ``variant``, an initial BAM file is selected</span>
<span class="sd">        and a target aligned BAM filename is constructed.</span>
<span class="sd">        The method checks first whether the aligned file is there. If a</span>
<span class="sd">        plausible candidate is not found, the initial BAM is aligned</span>
<span class="sd">        (``straight`` or ``π-shifted``, depending on the ``variant`` and</span>
<span class="sd">        using the proper reference). IF, on the other hand, a candidate is</span>
<span class="sd">        found, its computation is skipped.</span>

<span class="sd">        If the aligner cannot be run (i.e. calling the aligner returns</span>
<span class="sd">        ``None``), ``None`` is returned, meaning that the aligner was not</span>
<span class="sd">        called. This can happen when the aligner finds a *sentinel file*</span>
<span class="sd">        indicating that the computation is *work in progress*.</span>
<span class="sd">        (See :py:meth:`pacbio_data_processing.external.Blasr.__call__` for</span>
<span class="sd">        more details on the implementation.)</span>
<span class="sd">        This mechanism allows reentrancy.</span>

<span class="sd">        :return: the aligned input bam file, if it is there, or None if</span>
<span class="sd">                 it could not be computed (yet).</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># If we want to accept a user provided bam, we need to check it</span>
        <span class="c1"># before the next block: in that case the name doesn&#39;t need to be</span>
        <span class="c1"># constructed! If the name set in self.aligned_bams is None, a</span>
        <span class="c1"># proper name must be created; else the name found must be used</span>
        <span class="c1"># and it is not necessary to make one.</span>
        <span class="n">aligned_generic_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_GENERIC_NAME</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
        <span class="n">alignment_prefix</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_FILE_PREFIX</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
        <span class="n">inbam_path</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="n">inbam</span><span class="o">.</span><span class="n">bam_file_name</span><span class="p">)</span>
        <span class="n">aligned_bam_file</span> <span class="o">=</span> <span class="n">inbam_path</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
            <span class="n">alignment_prefix</span><span class="o">+</span><span class="n">inbam_path</span><span class="o">.</span><span class="n">name</span>
        <span class="p">)</span>
        <span class="n">result</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">aligned_bam_file</span><span class="p">)</span>
        <span class="n">aligned_bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>
        <span class="n">aligner_options</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">common_aligner_options</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
        <span class="k">if</span> <span class="n">bam_type</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="s2">&quot;ccs&quot;</span> <span class="ow">and</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aligner</span><span class="o">.</span><span class="vm">__class__</span> <span class="o">==</span> <span class="n">Pbmm2</span><span class="p">):</span>
            <span class="n">aligner_options</span><span class="p">[</span><span class="s2">&quot;preset&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;CCS&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="n">aligned_bam_file</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="ow">and</span>
                <span class="n">aligned_bam</span><span class="o">.</span><span class="n">is_plausible_aligned_version_of</span><span class="p">(</span><span class="n">inbam</span><span class="p">)):</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;...but a possible </span><span class="si">{</span><span class="n">aligned_generic_name</span><span class="si">}</span><span class="s2"> version of the &quot;</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bam_type</span><span class="si">}</span><span class="s2"> BAM was found: &#39;</span><span class="si">{</span><span class="n">aligned_bam_file</span><span class="si">}</span><span class="s2">&#39;. It &quot;</span>
                <span class="s2">&quot;will be used.&quot;</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">aligner_res</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aligner</span><span class="p">(</span>
                <span class="n">in_bamfile</span><span class="o">=</span><span class="n">inbam_path</span><span class="p">,</span>
                <span class="n">fasta</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">fasta</span><span class="p">[</span><span class="n">variant</span><span class="p">],</span>
                <span class="n">out_bamfile</span><span class="o">=</span><span class="n">aligned_bam_file</span><span class="p">,</span>
                <span class="o">**</span><span class="n">aligner_options</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="k">if</span> <span class="n">aligner_res</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">result</span> <span class="o">=</span> <span class="kc">None</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="sa">f</span><span class="s2">&quot;...since no </span><span class="si">{</span><span class="n">aligned_generic_name</span><span class="si">}</span><span class="s2"> version of the &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bam_type</span><span class="si">}</span><span class="s2"> BAM was found, one has been produced and it &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;will be used: &#39;</span><span class="si">{</span><span class="n">aligned_bam_file</span><span class="si">}</span><span class="s2">&#39;&quot;</span>
                <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="n">bam_type</span><span class="p">][</span><span class="n">variant</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span>
        <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._ensure_ccs_bam_aligned"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._ensure_ccs_bam_aligned">[docs]</a>    <span class="k">def</span> <span class="nf">_ensure_ccs_bam_aligned</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        As its name suggests, it is ensured that the aligned variants of</span>
<span class="sd">        the CCS file exist.</span>
<span class="sd">        The summary report is informed about the aligned CCS files.</span>

<span class="sd">        .. note::</span>

<span class="sd">           The CCS BAM file is created *before* checking if its aligned</span>
<span class="sd">           variants are present. It might seem a logic error to proceed</span>
<span class="sd">           this way instead of checking *first* for the existence of</span>
<span class="sd">           the aligned variants of the CCS BAM before deciding if the</span>
<span class="sd">           computation of the CCS BAM file is needed, but it is not an</span>
<span class="sd">           error: in order to decide if a given file can be an aligned</span>
<span class="sd">           version of the CCS BAM, we need the CCS BAM itself.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">MISSING_ALIGNED_CCS_MSG</span><span class="p">)</span>
        <span class="n">ccs_bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">CCS_bam_file</span><span class="p">)</span>
        <span class="n">try_computations_with_variants_until_done</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_align_bam_if_no_candidate_found</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">,</span>
            <span class="n">ccs_bam</span><span class="p">,</span> <span class="s2">&quot;ccs&quot;</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">aligned_ccs_bam_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;ccs&quot;</span><span class="p">]</span></div>

    <span class="nd">@cached_property</span>
    <span class="k">def</span> <span class="nf">partition</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Partition</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;The target ``Partition`` of the input BAM file that must be</span>
<span class="sd">        processed by the current analysis, according to the input</span>
<span class="sd">        provided by the user.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">Partition</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">partition</span><span class="p">,</span>
            <span class="n">BamFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">)</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">_set_tasks</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">opmr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">only_produce_methylation_report</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_split_bam</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_filter_molecules</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_collect_statistics</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_generate_indices</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">opmr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_ipd_analysis</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">opmr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_create_raw_detections_file</span> <span class="o">=</span> <span class="ow">not</span> <span class="n">opmr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_do_produce_methylation_report</span> <span class="o">=</span> <span class="kc">True</span>

<div class="viewcode-block" id="SingleMoleculeAnalysis._select_molecules"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._select_molecules">[docs]</a>    <span class="k">def</span> <span class="nf">_select_molecules</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method is part of the main sequence irrespective of whether</span>
<span class="sd">        the user selects to only produce the methylation report, or the</span>
<span class="sd">        full analysis.</span>
<span class="sd">        After this method the mapping ``_molecules_todo`` is created</span>
<span class="sd">        of type dict[int, Molecule], with molecules that:</span>

<span class="sd">        1. Belong to the partition,</span>
<span class="sd">        2. Are correctly mapped in the aligned CCS file, and</span>
<span class="sd">        3. If they belong to the ``pi-shifted`` variant (molecules</span>
<span class="sd">           obtained after aligning with a pi-shifted reference) then</span>
<span class="sd">           they cross the origin.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_molecule_ids_todo</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
            <span class="n">bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">variant</span><span class="p">])</span>
            <span class="n">mols</span> <span class="o">=</span> <span class="p">{</span><span class="n">m</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">bam</span><span class="o">.</span><span class="n">all_molecules</span> <span class="k">if</span> <span class="n">m</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="p">}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_molecule_ids_todo</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span> <span class="o">=</span> <span class="n">mols</span>
        <span class="n">molecules_from_ccs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_collect_suitable_molecules_from_ccs</span><span class="p">()</span>
        <span class="n">molecules_from_ccs</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_keep_only_pishifted_molecules_crossing_origin</span><span class="p">(</span>
                <span class="n">molecules_from_ccs</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_crosscheck_molecules_in_partition_with_ccs</span><span class="p">(</span><span class="n">molecules_from_ccs</span><span class="p">)</span></div>

    <span class="c1"># Smell: too many responsabilities!</span>
<div class="viewcode-block" id="SingleMoleculeAnalysis._collect_suitable_molecules_from_ccs"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._collect_suitable_molecules_from_ccs">[docs]</a>    <span class="k">def</span> <span class="nf">_collect_suitable_molecules_from_ccs</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Auxiliary routine of _select_molecules in charge of choosing</span>
<span class="sd">        *suitable molecules* from the aligned CCS bam files.</span>
<span class="sd">        The resulting mapping contains all *suitable molecules* in the</span>
<span class="sd">        &#39;straight&#39; variant and the *suitable molecules* in the</span>
<span class="sd">        &#39;π-shifted&#39; variant that are *not in* the &#39;straight&#39; variant.</span>
<span class="sd">        The molecules corresponding to both variants will be joined.</span>
<span class="sd">        Among all the possible subreads of each molecule in the aligned</span>
<span class="sd">        CCS, one is chosen by ``map_molecules_with_highest_sim_ratio``.</span>
<span class="sd">        The choice of *suitable molecules* is done by the method</span>
<span class="sd">        ``_discard_molecules_with_seq_mismatch``.</span>
<span class="sd">        Moreover the molecules are labeled with the variant they belong</span>
<span class="sd">        to. It is necessary to do this labeling, so that</span>
<span class="sd">        we can later trace what reference each molecule is attached to.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">_ccs_mols</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
            <span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ALIGNED_GENERIC_NAME</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Generating molecules mapping from </span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2"> CCS file&quot;</span><span class="p">)</span>
            <span class="n">_ccs_mols</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span> <span class="o">=</span> <span class="n">map_molecules_with_highest_sim_ratio</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;ccs&quot;</span><span class="p">][</span><span class="n">variant</span><span class="p">])</span>
            <span class="n">num_mols</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">_ccs_mols</span><span class="p">[</span><span class="n">variant</span><span class="p">])</span>
            <span class="k">for</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">_ccs_mols</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
                <span class="n">mol</span><span class="o">.</span><span class="n">variant</span> <span class="o">=</span> <span class="n">variant</span>
                <span class="n">mol</span><span class="o">.</span><span class="n">reference</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;ccs lines (</span><span class="si">{</span><span class="n">name</span><span class="si">}</span><span class="s2">): </span><span class="si">{</span><span class="n">num_mols</span><span class="si">}</span><span class="s2"> molecules found&quot;</span><span class="p">)</span>
        <span class="n">molecules_from_ccs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">[::</span><span class="o">-</span><span class="mi">1</span><span class="p">]:</span>
            <span class="n">molecules_from_ccs</span> <span class="o">|=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_discard_molecules_with_seq_mismatch</span><span class="p">(</span>
                <span class="n">_ccs_mols</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_report_discarded_molecules_with_seq_mismatch</span><span class="p">(</span>
            <span class="n">_ccs_mols</span><span class="p">,</span> <span class="n">molecules_from_ccs</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">molecules_from_ccs</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._report_discarded_molecules_with_seq_mismatch"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._report_discarded_molecules_with_seq_mismatch">[docs]</a>    <span class="k">def</span> <span class="nf">_report_discarded_molecules_with_seq_mismatch</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">mols_in_raw_ccs_files</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]],</span>
            <span class="n">molecules_from_ccs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]</span>
            <span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method simply logs the ids of discarded molecules and passes</span>
<span class="sd">        the infos to the ``SummaryReport`` instance.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">before_mols</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
            <span class="n">before_mols</span> <span class="o">|=</span> <span class="nb">set</span><span class="p">(</span><span class="n">mols_in_raw_ccs_files</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">discarded_mols</span> <span class="o">=</span> <span class="p">(</span><span class="n">before_mols</span> <span class="o">-</span> <span class="nb">set</span><span class="p">(</span><span class="n">molecules_from_ccs</span><span class="o">.</span><span class="n">keys</span><span class="p">()))</span>
        <span class="k">for</span> <span class="n">mol_id</span> <span class="ow">in</span> <span class="n">discarded_mols</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Molecule </span><span class="si">{</span><span class="n">mol_id</span><span class="si">}</span><span class="s2"> discarded due to DNA &quot;</span>
                <span class="s2">&quot;sequence mismatch with reference&quot;</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">mols_used_in_aligned_ccs</span> <span class="o">=</span> <span class="n">before_mols</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">mols_dna_mismatches</span> <span class="o">=</span> <span class="n">discarded_mols</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._discard_molecules_with_seq_mismatch"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._discard_molecules_with_seq_mismatch">[docs]</a>    <span class="k">def</span> <span class="nf">_discard_molecules_with_seq_mismatch</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">molecules_from_ccs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]</span>
            <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        The aligned CCS molecules are filtered in this method to keep</span>
<span class="sd">        only molecules that match perfectly the corresponding reference</span>
<span class="sd">        (ie, taking into account variants).</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">filtered_mols_in_ccs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">molecules_from_ccs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">start</span> <span class="o">=</span> <span class="n">mol</span><span class="o">.</span><span class="n">start</span>
            <span class="n">dna_in_ref</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">mol</span><span class="o">.</span><span class="n">variant</span><span class="p">][</span><span class="n">start</span><span class="p">:</span><span class="n">start</span><span class="o">+</span><span class="nb">len</span><span class="p">(</span><span class="n">mol</span><span class="p">)]</span>
            <span class="k">if</span> <span class="n">dna_in_ref</span> <span class="o">==</span> <span class="n">mol</span><span class="o">.</span><span class="n">dna</span><span class="p">:</span>
                <span class="n">filtered_mols_in_ccs</span><span class="p">[</span><span class="n">mol_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mol</span>
        <span class="k">return</span> <span class="n">filtered_mols_in_ccs</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._keep_only_pishifted_molecules_crossing_origin"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._keep_only_pishifted_molecules_crossing_origin">[docs]</a>    <span class="k">def</span> <span class="nf">_keep_only_pishifted_molecules_crossing_origin</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">molecules_from_ccs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]</span>
            <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method filters out molecules from the CCS aligned list that</span>
<span class="sd">        1. Belong to the π-shifted variant, and</span>
<span class="sd">        2. Do not cross the origin</span>
<span class="sd">        These molecules are unwanted because the point of including</span>
<span class="sd">        π-shifting in the analysis is to catch molecules crossing the</span>
<span class="sd">        origin.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">result</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">mol</span> <span class="ow">in</span> <span class="n">molecules_from_ccs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">mol</span><span class="o">.</span><span class="n">variant</span> <span class="o">==</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">mol</span><span class="o">.</span><span class="n">is_crossing_origin</span><span class="p">(</span><span class="n">ori_pi_shifted</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
                    <span class="k">continue</span>
            <span class="n">result</span><span class="p">[</span><span class="n">mol_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mol</span>
        <span class="k">return</span> <span class="n">result</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._crosscheck_molecules_in_partition_with_ccs"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._crosscheck_molecules_in_partition_with_ccs">[docs]</a>    <span class="k">def</span> <span class="nf">_crosscheck_molecules_in_partition_with_ccs</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">molecules_from_ccs</span><span class="p">:</span> <span class="nb">dict</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="n">Molecule</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method ensures that only the molecules in the current</span>
<span class="sd">        partition are processed. It does it by crosschecking the</span>
<span class="sd">        sets corresponding to the partition (for all variants) with</span>
<span class="sd">        the set of valid molecules in the ccs file.</span>
<span class="sd">        The attribute ``_molecules_todo`` is set, and its type is:</span>

<span class="sd">        .. code-block::</span>

<span class="sd">            dict[int, Molecule]</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mol_ids_in_ccs</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">molecules_from_ccs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="n">todo</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
            <span class="n">todo</span> <span class="o">=</span> <span class="n">todo</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
                <span class="p">{</span><span class="nb">int</span><span class="p">(</span><span class="n">_</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecule_ids_todo</span><span class="p">[</span><span class="n">variant</span><span class="p">]}</span>
            <span class="p">)</span>
        <span class="n">crosschecked_molecules</span> <span class="o">=</span> <span class="n">mol_ids_in_ccs</span> <span class="o">&amp;</span> <span class="n">todo</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span> <span class="o">=</span> <span class="p">{</span>
            <span class="n">k</span><span class="p">:</span> <span class="n">molecules_from_ccs</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">crosschecked_molecules</span>
        <span class="p">}</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._split_bam"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._split_bam">[docs]</a>    <span class="k">def</span> <span class="nf">_split_bam</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Produces a generator with 2-tuples of the type</span>
<span class="sd">        (mol_id[int], Molecule)</span>
<span class="sd">        where the Molecule is related to a single molecule BAM file that</span>
<span class="sd">        has been generated by ``split_bam_file_in_molecules``.</span>
<span class="sd">        It sets an attribute called ``_per_molecule_bam_generator`` that</span>
<span class="sd">        refers to that generator.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_split_bam</span><span class="p">:</span>
            <span class="n">variants_gens</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">variants</span><span class="p">:</span>
                <span class="n">splitted_mols_gen</span> <span class="o">=</span> <span class="n">split_bam_file_in_molecules</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">variant</span><span class="p">],</span> <span class="bp">self</span><span class="o">.</span><span class="n">workdir</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
                    <span class="p">{</span><span class="n">id_</span><span class="p">:</span> <span class="n">mol</span> <span class="k">for</span> <span class="n">id_</span><span class="p">,</span> <span class="n">mol</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">mol</span><span class="o">.</span><span class="n">variant</span> <span class="o">==</span> <span class="n">variant</span><span class="p">}</span>
                <span class="p">)</span>
                <span class="n">variants_gens</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">splitted_mols_gen</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_per_molecule_bam_generator</span> <span class="o">=</span> <span class="n">chain</span><span class="p">(</span><span class="o">*</span><span class="n">variants_gens</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_per_molecule_bam_generator</span> <span class="o">=</span> <span class="p">()</span></div>

    <span class="nd">@cached_property</span>
    <span class="k">def</span> <span class="nf">_minimum_mapping_quality</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal property]</span>
<span class="sd">        This attribute (cached, but not intented to be manually</span>
<span class="sd">        overwritten) returns the minimum value of the mapping</span>
<span class="sd">        quality that is acceptable for the current analysis.</span>
<span class="sd">        If a mapping quality threshold is provided in the command line,</span>
<span class="sd">        it is used and returned as the attribute value. Otherwise a value</span>
<span class="sd">        is computed using the</span>
<span class="sd">        py:func:`bam_utils.estimate_max_mapping_quality` function.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">mapping_quality_threshold</span><span class="p">:</span>
            <span class="n">min_mapq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">mapping_quality_threshold</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">estimated_max_mapq</span> <span class="o">=</span> <span class="n">estimate_max_mapping_quality</span><span class="p">(</span>
                <span class="n">BamFile</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">STRAIGHT_VARIANT</span><span class="p">]),</span>
                <span class="n">min_lines</span><span class="o">=</span><span class="n">MAPQ_MIN_LINES</span><span class="p">,</span> <span class="n">max_lines</span><span class="o">=</span><span class="n">MAPQ_MAX_LINES</span>
            <span class="p">)</span>
            <span class="n">min_mapq</span> <span class="o">=</span> <span class="n">estimated_max_mapq</span><span class="o">//</span><span class="mi">2</span>
        <span class="k">return</span> <span class="n">min_mapq</span>

<div class="viewcode-block" id="SingleMoleculeAnalysis._filter_molecules"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._filter_molecules">[docs]</a>    <span class="k">def</span> <span class="nf">_filter_molecules</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        The ``_molecules_todo`` mapping is here reduced by removing</span>
<span class="sd">        molecules that do not fulfill a minimum requirement of quality.</span>
<span class="sd">        The summary report is updated accordingly.</span>
<span class="sd">        See the ``cleanup_molecules`` auxiliary function for details</span>
<span class="sd">        on the filtering process.</span>
<span class="sd">        An attribute called ``_filtered_molecules_generator`` is set</span>
<span class="sd">        which produces ``MoleculeWorkUnit`` s.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">all_molecules_generator</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_per_molecule_bam_generator</span>
        <span class="n">initial_mols</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_filter_molecules</span><span class="p">:</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;[filter] Sieving molecules from input BAM before the &quot;</span>
                <span class="s2">&quot;IPD analysis&quot;</span>
            <span class="p">)</span>
            <span class="n">min_mapq</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_minimum_mapping_quality</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;[filter] minimum mapping quality: </span><span class="si">{</span><span class="n">min_mapq</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">clean_molecules</span> <span class="o">=</span> <span class="n">cleanup_molecules</span><span class="p">(</span>
                <span class="n">all_molecules_generator</span><span class="p">,</span> <span class="n">min_mapq_cutoff</span><span class="o">=</span><span class="n">min_mapq</span>
            <span class="p">)</span>
            <span class="n">clean_molecules</span><span class="p">,</span> <span class="n">clean_molecules_cp</span> <span class="o">=</span> <span class="n">tee</span><span class="p">(</span><span class="n">clean_molecules</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
            <span class="n">clean_molecules_redux</span> <span class="o">=</span> <span class="p">{</span><span class="n">_</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">clean_molecules</span><span class="p">}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span> <span class="o">=</span> <span class="p">{</span>
                <span class="n">k</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="p">[</span><span class="n">k</span><span class="p">]</span> <span class="k">for</span> <span class="n">k</span> <span class="ow">in</span> <span class="n">clean_molecules_redux</span>
            <span class="p">}</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_filtered_molecules_generator</span> <span class="o">=</span> <span class="n">clean_molecules_cp</span>
            <span class="n">filtered_out_mols</span> <span class="o">=</span> <span class="n">initial_mols</span><span class="o">-</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_filtered_molecules_generator</span> <span class="o">=</span> <span class="n">all_molecules_generator</span>
            <span class="n">filtered_out_mols</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">filtered_out_mols</span> <span class="o">=</span> <span class="n">filtered_out_mols</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._collect_statistics"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._collect_statistics">[docs]</a>    <span class="k">def</span> <span class="nf">_collect_statistics</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        It sets an attribute: &#39;filtered_bam_statistics&#39; that contains</span>
<span class="sd">        some data to be consumed by the MethylationReport.</span>
<span class="sd">        For now the only data is the number of subreads per molecule</span>
<span class="sd">        and per strand.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">subreads</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="n">Counter</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_collect_statistics</span><span class="p">:</span>
            <span class="n">filtered_bams</span><span class="p">,</span> <span class="n">backup</span> <span class="o">=</span> <span class="n">tee</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_filtered_molecules_generator</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
            <span class="k">for</span> <span class="p">(</span><span class="n">mol_id</span><span class="p">,</span> <span class="n">molecule</span><span class="p">)</span> <span class="ow">in</span> <span class="n">filtered_bams</span><span class="p">:</span>
                <span class="n">bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">molecule</span><span class="o">.</span><span class="n">src_bam_path</span><span class="p">)</span>
                <span class="n">new</span> <span class="o">=</span> <span class="n">count_subreads_per_molecule</span><span class="p">(</span><span class="n">bam</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">mol</span><span class="p">,</span> <span class="n">counter</span> <span class="ow">in</span> <span class="n">new</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                    <span class="n">subreads</span><span class="p">[</span><span class="n">mol</span><span class="p">]</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">counter</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_filtered_molecules_generator</span> <span class="o">=</span> <span class="n">backup</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">filtered_bam_statistics</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;subreads&quot;</span><span class="p">:</span> <span class="n">subreads</span><span class="p">}</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._generate_indices"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._generate_indices">[docs]</a>    <span class="k">def</span> <span class="nf">_generate_indices</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Indices are generated for all files that need to be analyzed by</span>
<span class="sd">        ipdSummary.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_generate_indices</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_indexed_molecules_generator</span> <span class="o">=</span> <span class="n">gen_index_single_molecule_bams</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_filtered_molecules_generator</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">pbindex_path</span>
            <span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._ipd_analysis"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._ipd_analysis">[docs]</a>    <span class="k">def</span> <span class="nf">_ipd_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Performs the IPD analysis of the single molecule files.</span>
<span class="sd">        Sets a generator with Paths to produced GFF files.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_ipd_analysis</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_ipd_processed_molecules</span> <span class="o">=</span> <span class="n">multi_ipd_summary</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_indexed_molecules_generator</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">fasta</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ipdsummary_path</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">num_simultaneous_ipdsummarys</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">num_workers_per_ipdsummary</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">modification_types</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">ipd_model</span>
            <span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis.produce_methylation_report"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.produce_methylation_report">[docs]</a>    <span class="k">def</span> <span class="nf">produce_methylation_report</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">params</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span>
        <span class="n">mr</span> <span class="o">=</span> <span class="n">MethylationReport</span><span class="p">(</span>
            <span class="n">detections_csv</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">raw_detections_filename</span><span class="p">,</span>
            <span class="n">molecules</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="p">,</span>
            <span class="n">modification_types</span><span class="o">=</span><span class="n">params</span><span class="o">.</span><span class="n">modification_types</span><span class="p">,</span>
            <span class="n">filtered_bam_statistics</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">filtered_bam_statistics</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">mr</span><span class="o">.</span><span class="n">save</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">methylation_report</span> <span class="o">=</span> <span class="n">mr</span><span class="o">.</span><span class="n">csv_name</span>
        <span class="c1">#  Move the next line to MethylationReport.save</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">mr</span><span class="o">.</span><span class="n">PRELOG</span><span class="si">}</span><span class="s2"> Results saved to file &#39;</span><span class="si">{</span><span class="n">mr</span><span class="o">.</span><span class="n">csv_name</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._fix_positions"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._fix_positions">[docs]</a>    <span class="k">def</span> <span class="nf">_fix_positions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        The purpose is to shift back the shifted positions in the</span>
<span class="sd">        π-shifted molecules.</span>
<span class="sd">        Two operations are required to complete that task:</span>

<span class="sd">        1. fixing positions in the gff files, and</span>
<span class="sd">        2. fixing positions in the molecules themselves.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_positions_in_gffs</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_positions_in_molecules</span><span class="p">()</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._fix_positions_in_gffs"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._fix_positions_in_gffs">[docs]</a>    <span class="k">def</span> <span class="nf">_fix_positions_in_gffs</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        In the case that some molecules have been processed, the</span>
<span class="sd">        positions in the gff files corresponding to molecules that have</span>
<span class="sd">        been *π-shifted* are shifted back.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">ipd_processed_mols</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_ipd_processed_molecules</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="k">return</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ipd_processed_molecules</span><span class="p">,</span> <span class="n">branch</span> <span class="o">=</span> <span class="n">tee</span><span class="p">(</span><span class="n">ipd_processed_mols</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">workunit</span> <span class="ow">in</span> <span class="n">branch</span><span class="p">:</span>
            <span class="n">mol_id</span><span class="p">,</span> <span class="n">molecule</span> <span class="o">=</span> <span class="n">workunit</span>
            <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">variant</span> <span class="o">==</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span>
                <span class="n">gff_path</span> <span class="o">=</span> <span class="n">molecule</span><span class="o">.</span><span class="n">gff_path</span>
                <span class="k">if</span> <span class="n">gff_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">pishift_back_positions_in_gff</span><span class="p">(</span><span class="n">gff_path</span><span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._fix_positions_in_molecules"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._fix_positions_in_molecules">[docs]</a>    <span class="k">def</span> <span class="nf">_fix_positions_in_molecules</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        All positions of *π-shifted* molecules are shifted back in the</span>
<span class="sd">        ``_molecules_todo`` dictionary (which will be used to generate</span>
<span class="sd">        the methylation report).</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">molecule</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">variant</span> <span class="o">==</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span>
                <span class="n">molecule</span><span class="o">.</span><span class="n">pi_shift_back</span><span class="p">()</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._report_faulty_molecules"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._report_faulty_molecules">[docs]</a>    <span class="k">def</span> <span class="nf">_report_faulty_molecules</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        The molecules that had any problem in their processing are</span>
<span class="sd">        passed to the ``SummaryReport`` as a set.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">faulty_molecules</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">mol_id</span><span class="p">,</span> <span class="n">molecule</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecules_todo</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">molecule</span><span class="o">.</span><span class="n">had_processing_problems</span><span class="p">:</span>
                <span class="n">faulty_molecules</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mol_id</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">faulty_mols</span> <span class="o">=</span> <span class="n">faulty_molecules</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._dump_results"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._dump_results">[docs]</a>    <span class="k">def</span> <span class="nf">_dump_results</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        All the output generated is driven by this method:</span>

<span class="sd">        * a joint gff file</span>
<span class="sd">        * a *per detection* csv file</span>
<span class="sd">        * a methylation report</span>
<span class="sd">        * a summary report</span>
<span class="sd">        * the molecules sets (see</span>
<span class="sd">          :py:class:pacbio_data_processing.summary.SummaryReport)</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_create_raw_detections_file</span><span class="p">:</span>
            <span class="n">joint_gffs</span> <span class="o">=</span> <span class="n">join_gffs</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_ipd_processed_molecules</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">joint_gff_filename</span>
            <span class="p">)</span>
            <span class="n">create_raw_detections_file</span><span class="p">(</span>
                <span class="n">joint_gffs</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">raw_detections_filename</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">modification_types</span><span class="p">,</span>
            <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">gff_result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">joint_gff_filename</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">raw_detections</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">raw_detections_filename</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">mapping_quality_threshold</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_minimum_mapping_quality</span><span class="p">)</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_do_produce_methylation_report</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">produce_methylation_report</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">save</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">summary_report_html_filename</span>
        <span class="p">)</span>
        <span class="c1"># ##</span>
        <span class="c1"># dumping molecules in all cases can seem sloppy, since</span>
        <span class="c1"># it is only used to merge partitions (for now). But serves</span>
        <span class="c1"># as a preparation step for future optimizations (runs where</span>
        <span class="c1"># not the full analysis must be produced):</span>
        <span class="c1"># ##                                         (DPalao, 29sep2022)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">molecule_sets_filename</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">summary_report_html_filename</span><span class="o">.</span><span class="n">with_suffix</span><span class="p">(</span>
                <span class="n">MOLECULES_FILE_SUFFIX</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">summary_report</span><span class="o">.</span><span class="n">dump_molecule_sets</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">molecule_sets_filename</span><span class="p">)</span></div>

    <span class="k">def</span> <span class="nf">_merge_joint_gffs</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">num_parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">num_partitions</span>
        <span class="n">current_prefix</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
        <span class="n">current</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">joint_gff_filename</span>
        <span class="n">all_</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">ipart</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">pref</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span><span class="n">ipart</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="p">,</span> <span class="n">pref</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">all_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">joint</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
            <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">merge_files</span><span class="p">(</span><span class="n">all_</span><span class="p">,</span> <span class="n">joint</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">joint_gffs_filename</span> <span class="o">=</span> <span class="n">joint</span>

    <span class="k">def</span> <span class="nf">_merge_raw_detections</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">num_parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">num_partitions</span>
        <span class="n">current_prefix</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
        <span class="n">current</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">raw_detections_filename</span>
        <span class="n">all_</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">ipart</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">pref</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span><span class="n">ipart</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="p">,</span> <span class="n">pref</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">all_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">joint</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
            <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">merge_files</span><span class="p">(</span><span class="n">all_</span><span class="p">,</span> <span class="n">joint</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">joint_raw_detections_filename</span> <span class="o">=</span> <span class="n">joint</span>

    <span class="k">def</span> <span class="nf">_merge_methylation_reports</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">num_parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">num_partitions</span>
        <span class="n">current_prefix</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
        <span class="n">_current</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">raw_detections_filename</span>
        <span class="n">current</span> <span class="o">=</span> <span class="n">_current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="s2">&quot;methylation.&quot;</span><span class="o">+</span><span class="n">_current</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="n">all_</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">ipart</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">pref</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span><span class="n">ipart</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="p">,</span> <span class="n">pref</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">all_</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">joint</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
            <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">merge_files</span><span class="p">(</span><span class="n">all_</span><span class="p">,</span> <span class="n">joint</span><span class="p">,</span> <span class="n">keep_only_first_header</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">joint_methylation_report_filename</span> <span class="o">=</span> <span class="n">joint</span>

    <span class="k">def</span> <span class="nf">_merge_summary_reports</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">sr</span> <span class="o">=</span> <span class="n">SummaryReport</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">input_bam_file</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;input&quot;</span><span class="p">][</span><span class="n">STRAIGHT_VARIANT</span><span class="p">],</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">reference</span><span class="p">[</span><span class="n">STRAIGHT_VARIANT</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="n">sr</span><span class="o">.</span><span class="n">methylation_report</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">joint_methylation_report_filename</span>
        <span class="n">sr</span><span class="o">.</span><span class="n">raw_detections</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">joint_raw_detections_filename</span>
        <span class="n">sr</span><span class="o">.</span><span class="n">gff_result</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">joint_gffs_filename</span>
        <span class="n">sr</span><span class="o">.</span><span class="n">aligned_ccs_bam_files</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bams</span><span class="p">[</span><span class="s2">&quot;ccs&quot;</span><span class="p">]</span>
        <span class="n">sr</span><span class="o">.</span><span class="n">mapping_quality_threshold</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_minimum_mapping_quality</span>
        <span class="n">num_parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">num_partitions</span>
        <span class="n">current_prefix</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
        <span class="n">current</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">molecule_sets_filename</span>
        <span class="k">for</span> <span class="n">ipart</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">pref</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span><span class="n">ipart</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="p">,</span> <span class="n">pref</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">current</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">sr</span><span class="o">.</span><span class="n">load_molecule_sets</span><span class="p">(</span><span class="n">filename</span><span class="p">)</span>
        <span class="n">current_html</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">summary_report_html_filename</span>
        <span class="n">joint</span> <span class="o">=</span> <span class="n">current_html</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span>
            <span class="n">current_html</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="o">+</span><span class="s2">&quot;.&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="n">sr</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">joint</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">all_partition_done_filenames</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">[</span><span class="n">Path</span><span class="p">]:</span>
        <span class="sd">&quot;&quot;&quot;Attribute that return a list of ``Path``s corresponding</span>
<span class="sd">        to the files expected to be found when all the partitions</span>
<span class="sd">        are processed (in case of partitioning the input BAM).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">num_parts</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">num_partitions</span>
        <span class="n">current_prefix</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
        <span class="n">current_done</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">partition_done_filename</span>
        <span class="n">ready</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">ipart</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">num_parts</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">pref</span> <span class="o">=</span> <span class="n">make_partition_prefix</span><span class="p">(</span><span class="n">ipart</span><span class="p">,</span> <span class="n">num_parts</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">current_done</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">current_prefix</span><span class="p">,</span> <span class="n">pref</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
            <span class="n">part_done_filename</span> <span class="o">=</span> <span class="n">current_done</span><span class="o">.</span><span class="n">with_name</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="n">ready</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">part_done_filename</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">ready</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">all_partitions_ready</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Attribute that answers the question: are all the partitions</span>
<span class="sd">        ready?</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">all</span><span class="p">([</span><span class="n">_</span><span class="o">.</span><span class="n">exists</span><span class="p">()</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_partition_done_filenames</span><span class="p">])</span>

<div class="viewcode-block" id="SingleMoleculeAnalysis._remove_partition_done_files"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._remove_partition_done_files">[docs]</a>    <span class="k">def</span> <span class="nf">_remove_partition_done_files</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Remove the partition done marker files after the partitions have</span>
<span class="sd">        been successfully merged.</span>

<span class="sd">        .. warning::</span>

<span class="sd">           It is assumed that this is called within the</span>
<span class="sd">           :py:meth:_post_process_partition phase.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">done_file</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_partition_done_filenames</span><span class="p">:</span>
            <span class="n">done_file</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="n">missing_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._merge_partitions_if_needed"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._merge_partitions_if_needed">[docs]</a>    <span class="k">def</span> <span class="nf">_merge_partitions_if_needed</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method merges properly the output files produced during</span>
<span class="sd">        the processing of all the partitions, *if* they are ready.</span>

<span class="sd">        .. warning::</span>

<span class="sd">           It is assumed that this is called within the</span>
<span class="sd">           :py:meth:_post_process_partition phase.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">all_partitions_ready</span><span class="p">:</span>
            <span class="n">errors</span> <span class="o">=</span> <span class="p">[]</span>
            <span class="k">for</span> <span class="n">meth</span> <span class="ow">in</span> <span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_merge_joint_gffs</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_merge_raw_detections</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_merge_methylation_reports</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">_merge_summary_reports</span><span class="p">,</span>
            <span class="p">):</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">meth</span><span class="p">()</span>
                <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                    <span class="n">errors</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">errors</span><span class="p">:</span>
                <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">+</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">   &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;[error #</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">] </span><span class="si">{</span><span class="n">_</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">errors</span><span class="p">)]</span>
                <span class="p">)</span>
                <span class="k">raise</span> <span class="n">SMAMergeError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_remove_partition_done_files</span><span class="p">()</span></div>

<div class="viewcode-block" id="SingleMoleculeAnalysis._post_process_partition"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._post_process_partition">[docs]</a>    <span class="k">def</span> <span class="nf">_post_process_partition</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        After the analysis is done, if only a fraction (aka ``partition``)</span>
<span class="sd">        was processed, this method declares that the analysis of the</span>
<span class="sd">        current``partition`` is complete and tries to merge the partitions</span>
<span class="sd">        (which will only occur if the proper conditions are met).</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">is_proper</span><span class="p">:</span>
            <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">partition_done_filename</span><span class="p">)</span><span class="o">.</span><span class="n">touch</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_merge_partitions_if_needed</span><span class="p">()</span>
            <span class="k">except</span> <span class="n">SMAMergeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="n">logging</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">))</span></div>

    <span class="k">def</span> <span class="nf">_backup_temp_dir_if_needed</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">parameters</span><span class="o">.</span><span class="n">keep_temp_dir</span><span class="p">:</span>
            <span class="n">suffix</span> <span class="o">=</span> <span class="s2">&quot;.backup&quot;</span>
            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="o">.</span><span class="n">is_proper</span><span class="p">:</span>
                <span class="n">suffix</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">partition</span><span class="p">)</span><span class="o">+</span><span class="n">suffix</span>
            <span class="n">backup</span> <span class="o">=</span> <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">workdir</span><span class="o">.</span><span class="n">name</span><span class="o">+</span><span class="n">suffix</span><span class="p">)</span>
            <span class="n">shutil</span><span class="o">.</span><span class="n">copytree</span><span class="p">(</span><span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">workdir</span><span class="o">.</span><span class="n">name</span><span class="p">),</span> <span class="n">backup</span><span class="p">)</span>
            <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Copied temporary dir to: &#39;</span><span class="si">{</span><span class="n">backup</span><span class="si">}</span><span class="s2">&#39;&quot;</span><span class="p">)</span>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">workdir</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">TemporaryDirectory</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;This attribute returns the necessary temporary working</span>
<span class="sd">        directory on demand and it ensures that only one temporary</span>
<span class="sd">        dir is created by caching.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">wdir</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_workdir</span>
        <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
            <span class="n">wdir</span> <span class="o">=</span> <span class="n">TemporaryDirectory</span><span class="p">(</span><span class="nb">dir</span><span class="o">=</span><span class="s2">&quot;.&quot;</span><span class="p">)</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_workdir</span> <span class="o">=</span> <span class="n">wdir</span>
        <span class="k">return</span> <span class="n">wdir</span>

    <span class="c1"># [G30]: Functions Should Do One Thing</span>
    <span class="c1"># [G31]: Hidden Temporal Couplings</span>
<div class="viewcode-block" id="SingleMoleculeAnalysis.__call__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.__call__">[docs]</a>    <span class="k">def</span> <span class="fm">__call__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Main entry point to perform a single molecule analysis:</span>
<span class="sd">        this method triggers the analysis.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">start</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_select_molecules</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_split_bam</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_filter_molecules</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_collect_statistics</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_generate_indices</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_ipd_analysis</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_fix_positions</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_report_faulty_molecules</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_dump_results</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_post_process_partition</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_backup_temp_dir_if_needed</span><span class="p">()</span>
        <span class="n">end</span> <span class="o">=</span> <span class="n">time</span><span class="o">.</span><span class="n">time</span><span class="p">()</span>
        <span class="n">t</span> <span class="o">=</span> <span class="n">end</span><span class="o">-</span><span class="n">start</span>
        <span class="n">t_h</span> <span class="o">=</span> <span class="n">t</span><span class="o">/</span><span class="mi">3600</span>
        <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Execution time (wall clock time): </span><span class="si">{</span><span class="n">t</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> s = </span><span class="si">{</span><span class="n">t_h</span><span class="si">:</span><span class="s2">.2f</span><span class="si">}</span><span class="s2"> h&quot;</span><span class="p">)</span></div></div>


<div class="viewcode-block" id="_main"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis._main">[docs]</a><span class="k">def</span> <span class="nf">_main</span><span class="p">(</span><span class="n">config</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;This function drives the Single Molecule Analysis once the</span>
<span class="sd">    input has been parsed.</span>

<span class="sd">    :meta public:</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">config_logging</span><span class="p">(</span><span class="n">config</span><span class="o">.</span><span class="n">verbose</span><span class="p">)</span>
    <span class="n">params</span> <span class="o">=</span> <span class="n">SingleMoleculeAnalysisParameters</span><span class="p">(</span><span class="n">config</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">params</span><span class="p">))</span>
    <span class="n">sma</span> <span class="o">=</span> <span class="n">SingleMoleculeAnalysis</span><span class="p">(</span><span class="n">params</span><span class="p">)</span>
    <span class="n">sma</span><span class="p">()</span></div>


<div class="viewcode-block" id="main_cl"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.main_cl">[docs]</a><span class="nd">@high_level_handler</span>
<span class="k">def</span> <span class="nf">main_cl</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Entry point for ``sm-analysis`` executable.&quot;&quot;&quot;</span>
    <span class="n">config</span> <span class="o">=</span> <span class="n">parse_cl</span><span class="p">()</span>
    <span class="n">_main</span><span class="p">(</span><span class="n">config</span><span class="p">)</span></div>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../indepth/index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>