
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pacbio_data_processing.summary &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css" />
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for pacbio_data_processing.summary</h1><div class="highlight"><pre>
<span></span><span class="c1">#######################################################################</span>
<span class="c1">#</span>
<span class="c1"># Copyright (C) 2022 David Palao</span>
<span class="c1">#</span>
<span class="c1"># This file is part of PacBioDataProcessing.</span>
<span class="c1">#</span>
<span class="c1">#  PacBioDataProcessing is free software: you can redistribute it and/or modify</span>
<span class="c1">#  it under the terms of the GNU General Public License as published by</span>
<span class="c1">#  the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1">#  (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1">#  PacBio data processing is distributed in the hope that it will be useful,</span>
<span class="c1">#  but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1">#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1">#  GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1">#  You should have received a copy of the GNU General Public License</span>
<span class="c1">#  along with PacBioDataProcessing. If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>
<span class="c1">#</span>
<span class="c1">#######################################################################</span>

<span class="kn">from</span> <span class="nn">datetime</span> <span class="kn">import</span> <span class="n">datetime</span>
<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">socket</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">defaultdict</span>
<span class="kn">from</span> <span class="nn">collections.abc</span> <span class="kn">import</span> <span class="n">Mapping</span>
<span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">pickle</span>

<span class="kn">import</span> <span class="nn">pandas</span>

<span class="kn">from</span> <span class="nn">.</span> <span class="kn">import</span> <span class="n">__version__</span> <span class="k">as</span> <span class="n">VERSION</span>
<span class="kn">from</span> <span class="nn">.constants</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">SM_ANALYSIS_EXE</span><span class="p">,</span> <span class="n">DNA_SEQ_COLUMN</span><span class="p">,</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">,</span> <span class="n">QUALITY_COLUMN</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">.templates</span> <span class="kn">import</span> <span class="n">SUMMARY_REPORT_HTML_TEMPLATE</span>
<span class="kn">from</span> <span class="nn">.bam</span> <span class="kn">import</span> <span class="n">BamFile</span>
<span class="kn">from</span> <span class="nn">.utils</span> <span class="kn">import</span> <span class="n">find_gatc_positions</span><span class="p">,</span> <span class="n">shift_me_back</span>
<span class="kn">from</span> <span class="nn">.cigar</span> <span class="kn">import</span> <span class="n">Cigar</span>
<span class="kn">from</span> <span class="nn">.plots</span> <span class="kn">import</span> <span class="p">(</span>
    <span class="n">make_barsplot</span><span class="p">,</span> <span class="n">make_rolling_history</span><span class="p">,</span> <span class="n">make_multi_histogram</span><span class="p">,</span> <span class="n">make_histogram</span><span class="p">,</span>
<span class="p">)</span>


<span class="n">SET_RO_ATTRIBUTE_ERR_MSG</span> <span class="o">=</span> <span class="s2">&quot;attribute &#39;</span><span class="si">{}</span><span class="s2">&#39; cannot be set directly&quot;</span>

<span class="n">DEFAULT_STYLE</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;    &lt;style&gt;</span>
<span class="s2">      table {</span>
<span class="s2">	    font-family: arial, sans-serif;</span>
<span class="s2">	    border-collapse: collapse;</span>
<span class="s2">	    width: 50%;</span>
<span class="s2">      }</span>

<span class="s2">      img {</span>
<span class="s2">	    width: 50%;</span>
<span class="s2">      }</span>

<span class="s2">      td, th {</span>
<span class="s2">	    border: 1px solid #dddddd;</span>
<span class="s2">	    text-align: left;</span>
<span class="s2">	    padding: 8px;</span>
<span class="s2">      }</span>

<span class="s2">      tr:nth-child(even) {</span>
<span class="s2">	    background-color: #dddddd;</span>
<span class="s2">      }</span>
<span class="s2">      tr:hover{</span>
<span class="s2">	    background-color: #D6EEEE;</span>
<span class="s2">      }</span>
<span class="s2">      .bottom-large {</span>
<span class="s2">	    margin-bottom: 1cm;</span>
<span class="s2">      }</span>
<span class="s2">      .top-large {</span>
<span class="s2">	    margin-top: 1cm;</span>
<span class="s2">      }</span>
<span class="s2">      .text-center {</span>
<span class="s2">	    text-align: center;</span>
<span class="s2">      }</span>

<span class="s2">    &lt;/style&gt;</span>

<span class="s2">&quot;&quot;&quot;</span>


<div class="viewcode-block" id="SimpleAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SimpleAttribute">[docs]</a><span class="k">class</span> <span class="nc">SimpleAttribute</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;The base class of all other descriptor managed attributes</span>
<span class="sd">    of ``SummaryReport``.</span>
<span class="sd">    It is a wrapper around the ``_data`` dictionary of the instance</span>
<span class="sd">    owning this attribute.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="SimpleAttribute.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SimpleAttribute.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span></div>

    <span class="k">def</span> <span class="nf">__set_name__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">owner</span><span class="p">,</span> <span class="n">name</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>

    <span class="k">def</span> <span class="fm">__get__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">owner</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">instance</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">return</span> <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>

    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span></div>


<div class="viewcode-block" id="MethylationReport"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethylationReport">[docs]</a><span class="k">class</span> <span class="nc">MethylationReport</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__set__</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="c1"># refactor idea: use Counter</span>
        <span class="c1"># refactor idea: use properties or methods</span>
        <span class="n">mols_in_meth_report</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">subreads_in_meth_report</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">mols_in_meth_report_with_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">subreads_in_meth_report_with_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">mols_in_meth_report_without_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">subreads_in_meth_report_without_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">max_possible_methylations</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">fully_methylated_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">fully_unmethylated_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">hemi_plus_methylated_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">hemi_minus_methylated_gatcs</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">positions_in_meth</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">gatc_positions_in_meth</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">value</span><span class="p">,</span> <span class="n">newline</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">csvfile</span><span class="p">:</span>
            <span class="n">reader</span> <span class="o">=</span> <span class="n">csv</span><span class="o">.</span><span class="n">reader</span><span class="p">(</span><span class="n">csvfile</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
            <span class="nb">next</span><span class="p">(</span><span class="n">reader</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">reader</span><span class="p">:</span>
                <span class="n">mols_in_meth_report</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">subreads</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">5</span><span class="p">])</span><span class="o">+</span><span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">6</span><span class="p">])</span>
                <span class="n">subreads_in_meth_report</span> <span class="o">+=</span> <span class="n">subreads</span>
                <span class="n">num_gatcs</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">10</span><span class="p">])</span>
                <span class="n">max_possible_methylations</span> <span class="o">+=</span> <span class="n">num_gatcs</span>
                <span class="n">states</span> <span class="o">=</span> <span class="n">line</span><span class="p">[</span><span class="mi">13</span><span class="p">]</span>
                <span class="n">start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span>
                <span class="n">end</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">line</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span>
                <span class="c1"># if end &lt; start (molecule crossing origin) the gatcs are</span>
                <span class="c1"># not counted. FIX IT.</span>
                <span class="n">positions_in_meth</span> <span class="o">|=</span> <span class="nb">set</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="n">start</span><span class="p">,</span> <span class="n">end</span><span class="p">))</span>
                <span class="k">if</span> <span class="n">num_gatcs</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="n">mols_in_meth_report_without_gatcs</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">subreads_in_meth_report_without_gatcs</span> <span class="o">+=</span> <span class="n">subreads</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">mols_in_meth_report_with_gatcs</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="n">subreads_in_meth_report_with_gatcs</span> <span class="o">+=</span> <span class="n">subreads</span>
                    <span class="n">gatc_positions_in_meth</span> <span class="o">|=</span> <span class="p">{</span>
                        <span class="nb">int</span><span class="p">(</span><span class="n">_</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">line</span><span class="p">[</span><span class="mi">11</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="p">)}</span>
                <span class="n">fully_methylated_gatcs</span> <span class="o">+=</span> <span class="n">states</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;f&quot;</span><span class="p">)</span>
                <span class="n">fully_unmethylated_gatcs</span> <span class="o">+=</span> <span class="n">states</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;0&quot;</span><span class="p">)</span>
                <span class="n">hemi_plus_methylated_gatcs</span> <span class="o">+=</span> <span class="n">states</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;+&quot;</span><span class="p">)</span>
                <span class="n">hemi_minus_methylated_gatcs</span> <span class="o">+=</span> <span class="n">states</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)</span>

        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;mols_in_meth_report&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mols_in_meth_report</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_in_meth_report&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">subreads_in_meth_report</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;mols_in_meth_report_with_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">mols_in_meth_report_with_gatcs</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_in_meth_report_with_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">subreads_in_meth_report_with_gatcs</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;mols_in_meth_report_without_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">mols_in_meth_report_without_gatcs</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_in_meth_report_without_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">subreads_in_meth_report_without_gatcs</span><span class="p">)</span>
        <span class="n">all_gatcs_in_meth</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gatc_positions_in_meth</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_gatcs_in_meth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_gatcs_in_meth</span>
        <span class="n">all_gatcs_not_in_meth</span> <span class="o">=</span> <span class="n">instance</span><span class="o">.</span><span class="n">total_gatcs_in_ref</span><span class="o">-</span><span class="n">all_gatcs_in_meth</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_gatcs_not_in_meth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_gatcs_not_in_meth</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;max_possible_methylations&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">max_possible_methylations</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;fully_methylated_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fully_methylated_gatcs</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;fully_unmethylated_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">fully_unmethylated_gatcs</span>
        <span class="n">hemi_methylated_gatcs</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">hemi_plus_methylated_gatcs</span><span class="o">+</span><span class="n">hemi_minus_methylated_gatcs</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;hemi_methylated_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">hemi_methylated_gatcs</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;hemi_plus_methylated_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">hemi_plus_methylated_gatcs</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;hemi_minus_methylated_gatcs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">hemi_minus_methylated_gatcs</span><span class="p">)</span>
        <span class="n">all_positions_in_meth</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions_in_meth</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_positions_in_meth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_positions_in_meth</span>
        <span class="n">all_positions_not_in_meth</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">instance</span><span class="o">.</span><span class="n">reference_base_pairs</span> <span class="o">-</span> <span class="n">all_positions_in_meth</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_positions_not_in_meth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_positions_not_in_meth</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="ROAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.ROAttribute">[docs]</a><span class="k">class</span> <span class="nc">ROAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="n">msg</span> <span class="o">=</span> <span class="n">SET_RO_ATTRIBUTE_ERR_MSG</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span>
        <span class="k">raise</span> <span class="ne">AttributeError</span><span class="p">(</span><span class="n">msg</span><span class="p">)</span></div>


<div class="viewcode-block" id="MolsSetAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MolsSetAttribute">[docs]</a><span class="k">class</span> <span class="nc">MolsSetAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">ids</span><span class="p">:</span> <span class="nb">set</span><span class="p">[</span><span class="nb">int</span><span class="p">]):</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_molecule_sets</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">ids</span>
        <span class="n">result</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">ids</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span>
        <span class="n">subreads_name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;mols&quot;</span><span class="p">,</span> <span class="s2">&quot;subreads&quot;</span><span class="p">)</span>
        <span class="n">num_subreads</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">subread</span> <span class="ow">in</span> <span class="n">instance</span><span class="o">.</span><span class="n">bam</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="n">subread</span><span class="o">.</span><span class="n">molecule_id</span><span class="p">)</span> <span class="ow">in</span> <span class="n">ids</span><span class="p">:</span>
                <span class="n">num_subreads</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">subreads_name</span><span class="p">]</span> <span class="o">=</span> <span class="n">num_subreads</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="PercAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.PercAttribute">[docs]</a><span class="k">class</span> <span class="nc">PercAttribute</span><span class="p">(</span><span class="n">ROAttribute</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;From a given attribute in a SummaryReport instance, the</span>
<span class="sd">    percentage is computed (wrt the value in ``s.total_attr``) and</span>
<span class="sd">    returned as str.</span>
<span class="sd">    &quot;&quot;&quot;</span>
<div class="viewcode-block" id="PercAttribute.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.PercAttribute.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">total_attr</span><span class="p">,</span> <span class="n">pref</span><span class="o">=</span><span class="s2">&quot;perc_&quot;</span><span class="p">,</span> <span class="n">suf</span><span class="o">=</span><span class="s2">&quot;_wrt_meth&quot;</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__init__</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">total_attr</span> <span class="o">=</span> <span class="n">total_attr</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">pref</span> <span class="o">=</span> <span class="n">pref</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">suf</span> <span class="o">=</span> <span class="n">suf</span></div>

    <span class="k">def</span> <span class="fm">__get__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">owner</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">instance</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="n">ref_attr</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">removeprefix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">pref</span><span class="p">)</span><span class="o">.</span><span class="n">removesuffix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">suf</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">perc</span> <span class="o">=</span> <span class="mi">100</span><span class="o">*</span><span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">ref_attr</span><span class="p">]</span><span class="o">/</span><span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">total_attr</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">ZeroDivisionError</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="s2">&quot;N/A&quot;</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">result</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">perc</span><span class="si">:</span><span class="s2">4.2f</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="AlignedBamAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.AlignedBamAttribute">[docs]</a><span class="k">class</span> <span class="nc">AlignedBamAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="n">bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__set__</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">bam</span><span class="p">)</span>
        <span class="n">mapqs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
            <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">_</span><span class="p">[</span><span class="n">QUALITY_COLUMN</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">bam</span><span class="p">],</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span>
        <span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;mapping_qualities&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mapqs</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_aligned_ini&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mapqs</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="MappingQualityThresholdAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityThresholdAttribute">[docs]</a><span class="k">class</span> <span class="nc">MappingQualityThresholdAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">cut</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__set__</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">cut</span><span class="p">)</span>
        <span class="n">mapqs</span> <span class="o">=</span> <span class="n">instance</span><span class="o">.</span><span class="n">mapping_qualities</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_with_low_mapq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mapqs</span><span class="p">[</span><span class="n">mapqs</span> <span class="o">&lt;</span> <span class="n">cut</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_with_high_mapq&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mapqs</span><span class="p">[</span><span class="n">mapqs</span> <span class="o">&gt;=</span> <span class="n">cut</span><span class="p">]</span><span class="o">.</span><span class="n">count</span><span class="p">()</span></div>


<div class="viewcode-block" id="InputBamAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.InputBamAttribute">[docs]</a><span class="k">class</span> <span class="nc">InputBamAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__set__</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="n">bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;input_bam_size&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">size_in_bytes</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;full_md5sum&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">full_md5sum</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;body_md5sum&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">md5sum_body</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">num_molecules</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">bam</span><span class="o">.</span><span class="n">num_subreads</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">bam</span> <span class="o">=</span> <span class="n">bam</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="InputReferenceAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.InputReferenceAttribute">[docs]</a><span class="k">class</span> <span class="nc">InputReferenceAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__set__</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="o">.</span><span class="n">fasta_name</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;reference_name&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">description</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;reference_base_pairs&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;reference_md5sum&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">md5sum</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;total_gatcs_in_ref&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">count</span><span class="p">(</span><span class="s2">&quot;GATC&quot;</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="AlignedCCSBamsAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.AlignedCCSBamsAttribute">[docs]</a><span class="k">class</span> <span class="nc">AlignedCCSBamsAttribute</span><span class="p">(</span><span class="n">SimpleAttribute</span><span class="p">):</span>
    <span class="c1"># This attribute could be probably replaced by one attr containing</span>
    <span class="c1"># Molecule&#39;s computed is SingleMoleculeAnalysis...</span>
    <span class="k">def</span> <span class="fm">__set__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">):</span>
        <span class="nb">super</span><span class="p">()</span><span class="o">.</span><span class="fm">__set__</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span>
        <span class="n">positions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">mols</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">gatc_positions</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="n">len_ref</span> <span class="o">=</span> <span class="n">instance</span><span class="o">.</span><span class="n">reference_base_pairs</span>
        <span class="k">for</span> <span class="n">variant</span><span class="p">,</span> <span class="n">aligned_ccs</span> <span class="ow">in</span> <span class="n">value</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">aligned_ccs</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">bam</span> <span class="o">=</span> <span class="n">BamFile</span><span class="p">(</span><span class="n">aligned_ccs</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">subread</span> <span class="ow">in</span> <span class="n">bam</span><span class="p">:</span>
                <span class="n">seq</span> <span class="o">=</span> <span class="n">subread</span><span class="p">[</span><span class="mi">9</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">()</span>
                <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)</span>
                <span class="n">cigar</span> <span class="o">=</span> <span class="n">Cigar</span><span class="p">(</span><span class="n">subread</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">decode</span><span class="p">())</span>
                <span class="n">mol_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">subread</span><span class="o">.</span><span class="n">molecule_id</span><span class="p">)</span>
                <span class="n">pos</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">subread</span><span class="p">[</span><span class="mi">3</span><span class="p">])</span><span class="o">-</span><span class="mi">1</span>
                <span class="k">if</span> <span class="n">variant</span> <span class="o">==</span> <span class="n">PI_SHIFTED_VARIANT</span><span class="p">:</span>
                    <span class="n">pos</span> <span class="o">=</span> <span class="n">shift_me_back</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">len_ref</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">cigar</span><span class="o">.</span><span class="n">number_pb_diffs</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">mol_id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mols</span><span class="p">:</span>
                        <span class="n">mols</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mol_id</span><span class="p">)</span>
                        <span class="n">positions</span> <span class="o">|=</span> <span class="nb">set</span><span class="p">(</span>
                            <span class="p">[</span><span class="n">_</span> <span class="o">%</span> <span class="n">len_ref</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">pos</span><span class="p">,</span> <span class="n">pos</span><span class="o">+</span><span class="n">N</span><span class="p">)])</span>
                        <span class="n">gatcs</span> <span class="o">=</span> <span class="n">find_gatc_positions</span><span class="p">(</span><span class="n">seq</span><span class="p">,</span> <span class="n">offset</span><span class="o">=</span><span class="n">pos</span><span class="p">)</span>
                        <span class="n">gatc_positions</span> <span class="o">|=</span> <span class="p">{</span><span class="n">_</span> <span class="o">%</span> <span class="n">len_ref</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">gatcs</span><span class="p">}</span>
        <span class="n">all_positions_in_bam</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">positions</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_positions_in_bam&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_positions_in_bam</span>
        <span class="n">all_positions_not_in_bam</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">instance</span><span class="o">.</span><span class="n">reference_base_pairs</span><span class="o">-</span><span class="n">all_positions_in_bam</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_positions_not_in_bam&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_positions_not_in_bam</span>
        <span class="n">all_gatcs_identified_in_bam</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">gatc_positions</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_gatcs_identified_in_bam&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">all_gatcs_identified_in_bam</span><span class="p">)</span>
        <span class="n">all_gatcs_not_identified_in_bam</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">instance</span><span class="o">.</span><span class="n">total_gatcs_in_ref</span><span class="o">-</span><span class="nb">len</span><span class="p">(</span><span class="n">gatc_positions</span><span class="p">))</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="s2">&quot;all_gatcs_not_identified_in_bam&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">all_gatcs_not_identified_in_bam</span><span class="p">)</span>
        <span class="n">instance</span><span class="o">.</span><span class="n">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">)</span></div>


<div class="viewcode-block" id="BarsPlotAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.BarsPlotAttribute">[docs]</a><span class="k">class</span> <span class="nc">BarsPlotAttribute</span><span class="p">(</span><span class="n">ROAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__get__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">owner</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">instance</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">if</span> <span class="n">instance</span><span class="o">.</span><span class="n">ready_to_go</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">dependency_names</span><span class="p">):</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
            <span class="n">data_dict</span> <span class="o">=</span> <span class="p">{}</span>
            <span class="k">for</span> <span class="n">main_key</span><span class="p">,</span> <span class="n">subkeys</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_definition</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                <span class="n">data_dict</span><span class="p">[</span><span class="n">main_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>
                <span class="k">for</span> <span class="n">subkey</span> <span class="ow">in</span> <span class="n">subkeys</span><span class="p">:</span>
                    <span class="k">try</span><span class="p">:</span>
                        <span class="n">value</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">getattr</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="n">subkey</span><span class="p">))</span>
                    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                        <span class="n">value</span> <span class="o">=</span> <span class="n">math</span><span class="o">.</span><span class="n">nan</span>
                    <span class="n">data_dict</span><span class="p">[</span><span class="n">main_key</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">value</span><span class="p">)</span>
                <span class="c1"># data_dict[main_key] = [</span>
                <span class="c1">#     float(getattr(instance, _)) for _ in subkeys]</span>
            <span class="n">data</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">data_dict</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">index_labels</span><span class="p">)</span>
            <span class="k">return</span> <span class="n">data</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">title</span><span class="p">,</span> <span class="n">filename</span></div>


<div class="viewcode-block" id="HistoryPlotAttribute"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.HistoryPlotAttribute">[docs]</a><span class="k">class</span> <span class="nc">HistoryPlotAttribute</span><span class="p">(</span><span class="n">ROAttribute</span><span class="p">):</span>
    <span class="k">def</span> <span class="fm">__get__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">,</span> <span class="n">owner</span><span class="p">):</span>
        <span class="k">if</span> <span class="n">instance</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>
        <span class="k">if</span> <span class="n">instance</span><span class="o">.</span><span class="n">ready_to_go</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">dependency_name</span><span class="p">):</span>
            <span class="n">filename</span> <span class="o">=</span> <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>
            <span class="n">series</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_data_for_plot</span><span class="p">(</span><span class="n">instance</span><span class="p">)</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">other_data</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">make_up_extra_args</span><span class="p">(</span><span class="n">instance</span><span class="p">)</span>
                <span class="n">bins</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">series</span><span class="p">)</span><span class="o">+</span><span class="mi">1</span>
                <span class="n">other_data</span> <span class="o">=</span> <span class="p">(</span><span class="n">bins</span><span class="p">,)</span> <span class="o">+</span> <span class="n">other_data</span>
            <span class="k">except</span> <span class="ne">AttributeError</span><span class="p">:</span>
                <span class="n">other_data</span> <span class="o">=</span> <span class="p">()</span>
            <span class="k">return</span> <span class="p">(</span><span class="n">series</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">title</span><span class="p">,</span> <span class="n">filename</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">legend</span><span class="p">)</span> <span class="o">+</span> <span class="n">other_data</span></div>


<div class="viewcode-block" id="MoleculeTypeBarsPlot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeTypeBarsPlot">[docs]</a><span class="k">class</span> <span class="nc">MoleculeTypeBarsPlot</span><span class="p">(</span><span class="n">BarsPlotAttribute</span><span class="p">):</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Processed molecules and subreads&quot;</span>
    <span class="n">dependency_names</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;mols_used_in_aligned_ccs&quot;</span><span class="p">,</span>
        <span class="s2">&quot;methylation_report&quot;</span>
    <span class="p">)</span>
    <span class="n">data_definition</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Used in aligned CCS&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_mols_used_in_aligned_ccs&quot;</span><span class="p">,</span>
            <span class="s2">&quot;perc_subreads_used_in_aligned_ccs&quot;</span>
        <span class="p">),</span>
        <span class="s1">&#39;Mismatch discards&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_mols_dna_mismatches&quot;</span><span class="p">,</span>
            <span class="s2">&quot;perc_subreads_dna_mismatches&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s1">&#39;Filtered out&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_filtered_out_mols&quot;</span><span class="p">,</span>
            <span class="s2">&quot;perc_filtered_out_subreads&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s1">&#39;Faulty (with processing error)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_faulty_mols&quot;</span><span class="p">,</span>
            <span class="s2">&quot;perc_faulty_subreads&quot;</span>
        <span class="p">),</span>
        <span class="s1">&#39;In Methylation report with GATC&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_mols_in_meth_report_with_gatcs&quot;</span><span class="p">,</span>
            <span class="s2">&quot;perc_subreads_in_meth_report_with_gatcs&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s1">&#39;In Methylation report without GATC&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_mols_in_meth_report_without_gatcs&quot;</span><span class="p">,</span>
            <span class="s2">&quot;perc_subreads_in_meth_report_without_gatcs&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">}</span>
    <span class="n">index_labels</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;Number of molecules (%)&#39;</span><span class="p">,</span> <span class="s1">&#39;Number of subreads (%)&#39;</span><span class="p">)</span></div>


<div class="viewcode-block" id="PositionCoverageBarsPlot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageBarsPlot">[docs]</a><span class="k">class</span> <span class="nc">PositionCoverageBarsPlot</span><span class="p">(</span><span class="n">BarsPlotAttribute</span><span class="p">):</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Position coverage in BAM file and Methylation report&quot;</span>
    <span class="n">dependency_names</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;aligned_ccs_bam_files&quot;</span><span class="p">,</span>
        <span class="s2">&quot;methylation_report&quot;</span>
    <span class="p">)</span>
    <span class="n">data_definition</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Positions covered by molecules in BAM file (%)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_all_positions_in_bam&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s1">&#39;Positions NOT covered by molecules in BAM file (%)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_all_positions_not_in_bam&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s1">&#39;Positions covered by molecules in methylation report (%)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_all_positions_in_meth&quot;</span><span class="p">,</span>
        <span class="p">),</span>
        <span class="s1">&#39;Positions NOT covered by molecules in methylation report (%)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;perc_all_positions_not_in_meth&quot;</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="p">}</span>
    <span class="n">index_labels</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Percentage&quot;</span><span class="p">,)</span></div>


<div class="viewcode-block" id="GATCCoverageBarsPlot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.GATCCoverageBarsPlot">[docs]</a><span class="k">class</span> <span class="nc">GATCCoverageBarsPlot</span><span class="p">(</span><span class="n">BarsPlotAttribute</span><span class="p">):</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;GATCs in BAM file and Methylation report&quot;</span>
    <span class="n">dependency_names</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;aligned_ccs_bam_files&quot;</span><span class="p">,</span>
        <span class="s2">&quot;methylation_report&quot;</span>
    <span class="p">)</span>
    <span class="n">data_definition</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;GATCs in BAM file (%)&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;perc_all_gatcs_identified_in_bam&quot;</span><span class="p">,),</span>
        <span class="s1">&#39;GATCs NOT in BAM file (%)&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;perc_all_gatcs_not_identified_in_bam&quot;</span><span class="p">,),</span>
        <span class="s1">&#39;GATCs in methylation report (%)&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;perc_all_gatcs_in_meth&quot;</span><span class="p">,),</span>
        <span class="s1">&#39;GATCs NOT in methylation report (%)&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;perc_all_gatcs_not_in_meth&quot;</span><span class="p">,)</span>
    <span class="p">}</span>
    <span class="n">index_labels</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Percentage&quot;</span><span class="p">,)</span></div>


<div class="viewcode-block" id="MethTypeBarsPlot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethTypeBarsPlot">[docs]</a><span class="k">class</span> <span class="nc">MethTypeBarsPlot</span><span class="p">(</span><span class="n">BarsPlotAttribute</span><span class="p">):</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Methylation types in methylation report&quot;</span>
    <span class="n">dependency_names</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;methylation_report&quot;</span><span class="p">,)</span>
    <span class="n">data_definition</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;Fully methylated (%)&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;fully_methylated_gatcs_wrt_meth&quot;</span><span class="p">,),</span>
        <span class="s1">&#39;Fully unmethylated (%)&#39;</span><span class="p">:</span> <span class="p">(</span><span class="s2">&quot;fully_unmethylated_gatcs_wrt_meth&quot;</span><span class="p">,),</span>
        <span class="s1">&#39;Hemi-methylated in + strand (%)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;hemi_plus_methylated_gatcs_wrt_meth&quot;</span><span class="p">,),</span>
        <span class="s1">&#39;Hemi-methylated in - strand (%)&#39;</span><span class="p">:</span> <span class="p">(</span>
            <span class="s2">&quot;hemi_minus_methylated_gatcs_wrt_meth&quot;</span><span class="p">,)</span>
    <span class="p">}</span>
    <span class="n">index_labels</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Percentage&quot;</span><span class="p">,)</span></div>


<div class="viewcode-block" id="MoleculeLenHistogram"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram">[docs]</a><span class="k">class</span> <span class="nc">MoleculeLenHistogram</span><span class="p">(</span><span class="n">HistoryPlotAttribute</span><span class="p">):</span>
    <span class="n">dependency_name</span> <span class="o">=</span> <span class="s2">&quot;methylation_report&quot;</span>
    <span class="n">column_name</span> <span class="o">=</span> <span class="s2">&quot;len(molecule)&quot;</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Initial subreads and analyzed molecule length histogram&quot;</span>
    <span class="n">data_name</span> <span class="o">=</span> <span class="s2">&quot;length&quot;</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Initial subreads&quot;</span><span class="p">,</span> <span class="s2">&quot;Analyzed molecules&quot;</span><span class="p">)</span>
    <span class="n">legend</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="c1"># hue = &quot;source&quot;</span>

<div class="viewcode-block" id="MoleculeLenHistogram.make_data_for_plot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.make_data_for_plot">[docs]</a>    <span class="k">def</span> <span class="nf">make_data_for_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">):</span>
        <span class="n">series</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="n">subreads</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
            <span class="p">[</span><span class="nb">len</span><span class="p">(</span><span class="n">_</span><span class="p">[</span><span class="n">DNA_SEQ_COLUMN</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">instance</span><span class="o">.</span><span class="n">bam</span><span class="p">],</span> <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data_name</span>
        <span class="p">)</span>
        <span class="n">series</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">subreads</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="nb">getattr</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dependency_name</span><span class="p">),</span> <span class="n">delimiter</span><span class="o">=</span><span class="s2">&quot;;&quot;</span><span class="p">)</span>
        <span class="n">mols</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">column_name</span><span class="p">]</span>
        <span class="n">mols</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">data_name</span>
        <span class="n">series</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">mols</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span><span class="n">k</span><span class="p">:</span> <span class="n">v</span> <span class="k">for</span> <span class="n">k</span><span class="p">,</span> <span class="n">v</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">labels</span><span class="p">,</span> <span class="n">series</span><span class="p">)}</span></div></div>


<div class="viewcode-block" id="MappingQualityHistogram"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram">[docs]</a><span class="k">class</span> <span class="nc">MappingQualityHistogram</span><span class="p">(</span><span class="n">HistoryPlotAttribute</span><span class="p">):</span>
    <span class="n">dependency_name</span> <span class="o">=</span> <span class="s2">&quot;aligned_bam&quot;</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Mapping quality histogram of subreads in the aligned input BAM&quot;</span>
    <span class="n">data_name</span> <span class="o">=</span> <span class="s2">&quot;mapping quality&quot;</span>
    <span class="n">legend</span> <span class="o">=</span> <span class="kc">True</span>

<div class="viewcode-block" id="MappingQualityHistogram.make_up_extra_args"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.make_up_extra_args">[docs]</a>    <span class="k">def</span> <span class="nf">make_up_extra_args</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">):</span>
        <span class="n">log_scale</span> <span class="o">=</span> <span class="p">(</span><span class="kc">False</span><span class="p">,</span> <span class="kc">True</span><span class="p">)</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">vertical_line_at</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">instance</span><span class="o">.</span><span class="n">mapping_quality_threshold</span><span class="p">)</span>
        <span class="k">except</span> <span class="p">(</span><span class="ne">AttributeError</span><span class="p">,</span> <span class="ne">ValueError</span><span class="p">):</span>
            <span class="n">vertical_line_at</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="n">vertical_line_label</span> <span class="o">=</span> <span class="s2">&quot;mapping quality threshold&quot;</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">log_scale</span><span class="p">,</span> <span class="n">vertical_line_at</span><span class="p">,</span> <span class="n">vertical_line_label</span><span class="p">)</span></div>

<div class="viewcode-block" id="MappingQualityHistogram.make_data_for_plot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.make_data_for_plot">[docs]</a>    <span class="k">def</span> <span class="nf">make_data_for_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">):</span>
        <span class="n">bam</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="n">instance</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">dependency_name</span><span class="p">)</span>
        <span class="n">mapqs</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span>
            <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">_</span><span class="p">[</span><span class="n">QUALITY_COLUMN</span><span class="p">])</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">bam</span><span class="p">],</span> <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">data_name</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">mapqs</span></div></div>


<div class="viewcode-block" id="PositionCoverageHistory"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory">[docs]</a><span class="k">class</span> <span class="nc">PositionCoverageHistory</span><span class="p">(</span><span class="n">HistoryPlotAttribute</span><span class="p">):</span>
    <span class="n">dependency_name</span> <span class="o">=</span> <span class="s2">&quot;methylation_report&quot;</span>
    <span class="n">title</span> <span class="o">=</span> <span class="s2">&quot;Sequencing positions covered by analyzed molecules&quot;</span>
    <span class="n">start_column_name</span> <span class="o">=</span> <span class="s2">&quot;start of molecule&quot;</span>
    <span class="n">len_column_name</span> <span class="o">=</span> <span class="s2">&quot;len(molecule)&quot;</span>
    <span class="n">labels</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;Positions&quot;</span><span class="p">,)</span>
    <span class="n">legend</span> <span class="o">=</span> <span class="kc">False</span>

<div class="viewcode-block" id="PositionCoverageHistory.make_data_for_plot"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.make_data_for_plot">[docs]</a>    <span class="k">def</span> <span class="nf">make_data_for_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">instance</span><span class="p">):</span>
        <span class="n">pre_df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
            <span class="n">instance</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">dependency_name</span><span class="p">],</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
        <span class="n">starts</span> <span class="o">=</span> <span class="n">pre_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">start_column_name</span><span class="p">]</span>
        <span class="n">lengths</span> <span class="o">=</span> <span class="n">pre_df</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">len_column_name</span><span class="p">]</span>
        <span class="n">N</span> <span class="o">=</span> <span class="n">instance</span><span class="o">.</span><span class="n">reference_base_pairs</span>
        <span class="n">coverage</span> <span class="o">=</span> <span class="p">{</span><span class="n">i</span><span class="p">:</span> <span class="mi">0</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">N</span><span class="o">+</span><span class="mi">1</span><span class="p">)}</span>
        <span class="k">for</span> <span class="n">s</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">starts</span><span class="p">,</span> <span class="n">lengths</span><span class="p">):</span>
            <span class="k">for</span> <span class="n">j</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">s</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="n">s</span><span class="o">+</span><span class="n">l</span><span class="o">-</span><span class="mi">1</span><span class="p">):</span>
                <span class="n">position</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">+</span> <span class="n">j</span> <span class="o">%</span> <span class="n">N</span>
                <span class="n">coverage</span><span class="p">[</span><span class="n">position</span><span class="p">]</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">return</span> <span class="n">coverage</span></div></div>


<div class="viewcode-block" id="SummaryReport"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport">[docs]</a><span class="k">class</span> <span class="nc">SummaryReport</span><span class="p">(</span><span class="n">Mapping</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Final summary report generated by ``sm-analysis`` initially</span>
<span class="sd">    intended for humans.</span>

<span class="sd">    This class has been crafted to carefully control most of its</span>
<span class="sd">    attributes. Data can be fed into the class by setting some</span>
<span class="sd">    attributes. That process can trigger the generation of other</span>
<span class="sd">    attributes, that are typically *read-only*. In some cases</span>
<span class="sd">    the attributes are *simple* attributes, without side effects.</span>

<span class="sd">    After instantiating the class with the path to the input BAM and</span>
<span class="sd">    the dna sequence of the reference (instance of ``DNASeq``), one</span>
<span class="sd">    must set some attributes to be able to save the summary report::</span>

<span class="sd">      s = SummaryReport(bam_path, aligned_bam_path, dnaseq)</span>
<span class="sd">      s.methylation_report = path_to_meth_report</span>
<span class="sd">      s.raw_detections = path_to_raw_detections_file</span>
<span class="sd">      s.gff_result = path_to_gff_result_file</span>
<span class="sd">      s.aligned_ccs_bam_files = {</span>
<span class="sd">          &#39;straight&#39;: aligned_ccs_path,</span>
<span class="sd">          &#39;pi-shifted&#39;: pi_shifted_aligned_ccs_path</span>
<span class="sd">      }</span>
<span class="sd">      # The next is optional: it will add a vertical line to the</span>
<span class="sd">      # mapping quality histogram:</span>
<span class="sd">      s.mapping_quality_threshold = 30</span>

<span class="sd">      # Some information about what happened with some molecules must</span>
<span class="sd">      # be given as well. There are two options for that. First, in the</span>
<span class="sd">      # *normal flow* the following would be done:</span>
<span class="sd">      s.mols_used_in_aligned_ccs = {3, 67, ...}  # set of ints</span>
<span class="sd">      # Optionally you can provide:</span>
<span class="sd">      s.mols_dna_mismatches = {20, 49, ...}  # set of ints</span>
<span class="sd">      # or/and:</span>
<span class="sd">      s.filtered_out_mols = {22, 493, ...}  # set of ints</span>
<span class="sd">      # or/and:</span>
<span class="sd">      s.faulty_mols = {332, 389, ...}  # set of ints</span>

<span class="sd">      # The second possibility is to load the data about the molecules</span>
<span class="sd">      # from file(s). That is an option if a partitioned</span>
<span class="sd">      # ``SingleMoleculeAnalysis`` has been carried out and the results</span>
<span class="sd">      # must be merged. In that case, you would do:</span>
<span class="sd">      s.load_molecule_sets(&quot;file1.pickle&quot;)</span>
<span class="sd">      s.load_molecule_sets(&quot;file2.pickle&quot;)</span>
<span class="sd">      ...</span>
<span class="sd">      # and so many files as necessary can be loaded. Their information</span>
<span class="sd">      # will be added together.</span>
<span class="sd">      # The names of the files can be also ``Path`` instances (which is</span>
<span class="sd">      # the usual case).</span>

<span class="sd">    At this point all the necessary data is there and the report</span>
<span class="sd">    can be created::</span>

<span class="sd">      s.save(&#39;summary_whatever.html&#39;)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">methylation_report</span> <span class="o">=</span> <span class="n">MethylationReport</span><span class="p">()</span>
    <span class="n">raw_detections</span> <span class="o">=</span> <span class="n">SimpleAttribute</span><span class="p">()</span>
    <span class="n">gff_result</span> <span class="o">=</span> <span class="n">SimpleAttribute</span><span class="p">()</span>
    <span class="n">input_bam</span> <span class="o">=</span> <span class="n">InputBamAttribute</span><span class="p">()</span>
    <span class="n">input_bam_size</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">full_md5sum</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">body_md5sum</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">aligned_bam</span> <span class="o">=</span> <span class="n">AlignedBamAttribute</span><span class="p">()</span>
    <span class="n">input_reference</span> <span class="o">=</span> <span class="n">InputReferenceAttribute</span><span class="p">()</span>
    <span class="n">reference_name</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">reference_base_pairs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">reference_md5sum</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">mols_ini</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">subreads_ini</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">subreads_aligned_ini</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">subreads_with_low_mapq</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_with_low_mapq</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_aligned_ini&quot;</span><span class="p">)</span>
    <span class="n">subreads_with_high_mapq</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_with_high_mapq</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_aligned_ini&quot;</span><span class="p">)</span>
    <span class="n">mols_dna_mismatches</span> <span class="o">=</span> <span class="n">MolsSetAttribute</span><span class="p">()</span>
    <span class="n">perc_mols_dna_mismatches</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">subreads_dna_mismatches</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_dna_mismatches</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">filtered_out_mols</span> <span class="o">=</span> <span class="n">MolsSetAttribute</span><span class="p">()</span>
    <span class="n">perc_filtered_out_mols</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">filtered_out_subreads</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_filtered_out_subreads</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">faulty_mols</span> <span class="o">=</span> <span class="n">MolsSetAttribute</span><span class="p">()</span>
    <span class="n">perc_faulty_mols</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">faulty_subreads</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_faulty_subreads</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">mols_in_meth_report</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_mols_in_meth_report</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">subreads_in_meth_report</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_in_meth_report</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">mols_in_meth_report_with_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_mols_in_meth_report_with_gatcs</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">subreads_in_meth_report_with_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_in_meth_report_with_gatcs</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">mols_in_meth_report_without_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_mols_in_meth_report_without_gatcs</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">subreads_in_meth_report_without_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_in_meth_report_without_gatcs</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">mols_used_in_aligned_ccs</span> <span class="o">=</span> <span class="n">MolsSetAttribute</span><span class="p">()</span>
    <span class="n">perc_mols_used_in_aligned_ccs</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;mols_ini&quot;</span><span class="p">)</span>
    <span class="n">subreads_used_in_aligned_ccs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_subreads_used_in_aligned_ccs</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;subreads_ini&quot;</span><span class="p">)</span>
    <span class="n">aligned_ccs_bam_files</span> <span class="o">=</span> <span class="n">AlignedCCSBamsAttribute</span><span class="p">()</span>
    <span class="n">all_positions_in_bam</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_positions_in_bam</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;reference_base_pairs&quot;</span><span class="p">)</span>
    <span class="n">all_positions_not_in_bam</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_positions_not_in_bam</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;reference_base_pairs&quot;</span><span class="p">)</span>
    <span class="n">all_positions_in_meth</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_positions_in_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;reference_base_pairs&quot;</span><span class="p">)</span>
    <span class="n">all_positions_not_in_meth</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_positions_not_in_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;reference_base_pairs&quot;</span><span class="p">)</span>
    <span class="n">total_gatcs_in_ref</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">all_gatcs_identified_in_bam</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_gatcs_identified_in_bam</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;total_gatcs_in_ref&quot;</span><span class="p">)</span>
    <span class="n">all_gatcs_not_identified_in_bam</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_gatcs_not_identified_in_bam</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;total_gatcs_in_ref&quot;</span><span class="p">)</span>
    <span class="n">all_gatcs_in_meth</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_gatcs_in_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;total_gatcs_in_ref&quot;</span><span class="p">)</span>
    <span class="n">all_gatcs_not_in_meth</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">perc_all_gatcs_not_in_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span><span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;total_gatcs_in_ref&quot;</span><span class="p">)</span>
    <span class="n">max_possible_methylations</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">fully_methylated_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">fully_methylated_gatcs_wrt_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;max_possible_methylations&quot;</span><span class="p">)</span>
    <span class="n">fully_unmethylated_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">fully_unmethylated_gatcs_wrt_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;max_possible_methylations&quot;</span><span class="p">)</span>
    <span class="n">hemi_methylated_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">hemi_methylated_gatcs_wrt_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;max_possible_methylations&quot;</span><span class="p">)</span>
    <span class="n">hemi_plus_methylated_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">hemi_plus_methylated_gatcs_wrt_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;max_possible_methylations&quot;</span><span class="p">)</span>
    <span class="n">hemi_minus_methylated_gatcs</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">hemi_minus_methylated_gatcs_wrt_meth</span> <span class="o">=</span> <span class="n">PercAttribute</span><span class="p">(</span>
        <span class="n">total_attr</span><span class="o">=</span><span class="s2">&quot;max_possible_methylations&quot;</span><span class="p">)</span>

    <span class="n">molecule_type_bars</span> <span class="o">=</span> <span class="n">MoleculeTypeBarsPlot</span><span class="p">()</span>
    <span class="n">molecule_len_histogram</span> <span class="o">=</span> <span class="n">MoleculeLenHistogram</span><span class="p">()</span>
    <span class="n">mapping_quality_histogram</span> <span class="o">=</span> <span class="n">MappingQualityHistogram</span><span class="p">()</span>
    <span class="n">position_coverage_bars</span> <span class="o">=</span> <span class="n">PositionCoverageBarsPlot</span><span class="p">()</span>
    <span class="n">position_coverage_history</span> <span class="o">=</span> <span class="n">PositionCoverageHistory</span><span class="p">()</span>
    <span class="n">gatc_coverage_bars</span> <span class="o">=</span> <span class="n">GATCCoverageBarsPlot</span><span class="p">()</span>
    <span class="n">meth_type_bars</span> <span class="o">=</span> <span class="n">MethTypeBarsPlot</span><span class="p">()</span>
    <span class="n">mapping_qualities</span> <span class="o">=</span> <span class="n">ROAttribute</span><span class="p">()</span>
    <span class="n">mapping_quality_threshold</span> <span class="o">=</span> <span class="n">MappingQualityThresholdAttribute</span><span class="p">()</span>

<div class="viewcode-block" id="SummaryReport.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">bam_path</span><span class="p">,</span> <span class="n">aligned_bam_path</span><span class="p">,</span> <span class="n">dnaseq</span><span class="p">,</span> <span class="n">figures_prefix</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_primary_attributes</span> <span class="o">=</span> <span class="n">defaultdict</span><span class="p">(</span><span class="k">lambda</span><span class="p">:</span> <span class="kc">False</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span> <span class="o">=</span> <span class="s2">&quot;figures&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_data</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;style&quot;</span><span class="p">:</span> <span class="n">DEFAULT_STYLE</span><span class="p">,</span>
            <span class="s2">&quot;version&quot;</span><span class="p">:</span> <span class="n">VERSION</span><span class="p">,</span>
            <span class="s2">&quot;when&quot;</span><span class="p">:</span> <span class="n">datetime</span><span class="o">.</span><span class="n">now</span><span class="p">()</span><span class="o">.</span><span class="n">isoformat</span><span class="p">(</span><span class="n">timespec</span><span class="o">=</span><span class="s2">&quot;minutes&quot;</span><span class="p">),</span>
            <span class="s2">&quot;program&quot;</span><span class="p">:</span> <span class="n">SM_ANALYSIS_EXE</span><span class="p">,</span>
            <span class="s2">&quot;clos&quot;</span><span class="p">:</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">[</span><span class="mi">1</span><span class="p">:]),</span>
            <span class="s2">&quot;hostname&quot;</span><span class="p">:</span> <span class="n">socket</span><span class="o">.</span><span class="n">gethostname</span><span class="p">(),</span>
            <span class="s2">&quot;molecule_type_bars&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;molecule_type_bars.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;molecule_len_histogram&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;molecule_length_histogram.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;mapping_quality_histogram&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;mapping_quality_histogram.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;position_coverage_bars&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;position_coverage_bars.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;position_coverage_history&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;position_coverage_history.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;gatc_coverage_bars&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;gatc_coverage_bars.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;meth_type_bars&quot;</span><span class="p">:</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">figures_prefix</span><span class="si">}</span><span class="s2">&quot;</span>
                <span class="s2">&quot;meth_type_bars.png&quot;</span>
            <span class="p">),</span>
            <span class="s2">&quot;filtered_out_mols&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;filtered_out_subreads&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;mols_dna_mismatches&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;subreads_dna_mismatches&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;faulty_mols&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
            <span class="s2">&quot;faulty_subreads&quot;</span><span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="c1"># _molecule_sets is used as low level storage of the molecule ids</span>
        <span class="c1"># passed to:</span>
        <span class="c1"># mols_dna_mismatches,</span>
        <span class="c1"># filtered_out_mols,</span>
        <span class="c1"># faulty_mols, and</span>
        <span class="c1"># mols_used_in_aligned_ccs</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_molecule_sets</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;mols_dna_mismatches&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">(),</span>
            <span class="s2">&quot;filtered_out_mols&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">(),</span>
            <span class="s2">&quot;faulty_mols&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">(),</span>
            <span class="s2">&quot;mols_used_in_aligned_ccs&quot;</span><span class="p">:</span> <span class="nb">set</span><span class="p">(),</span>
        <span class="p">}</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_loaded_molecule_sets</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_bam</span> <span class="o">=</span> <span class="n">bam_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">aligned_bam</span> <span class="o">=</span> <span class="n">aligned_bam_path</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">input_reference</span> <span class="o">=</span> <span class="n">dnaseq</span>
        <span class="n">Path</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">figures_dir_name</span><span class="p">)</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">exist_ok</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">as_html</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">SUMMARY_REPORT_HTML_TEMPLATE</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="o">**</span><span class="bp">self</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">_pre_save</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;In case of having some molecule sets that have been loaded from</span>
<span class="sd">        pickled files, the attribute must be explicitly set to trigger the</span>
<span class="sd">        computation of the data needed for the plots.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">for</span> <span class="n">name</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_loaded_molecule_sets</span><span class="p">:</span>
            <span class="nb">setattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecule_sets</span><span class="p">[</span><span class="n">name</span><span class="p">])</span>

<div class="viewcode-block" id="SummaryReport.save"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.save">[docs]</a>    <span class="k">def</span> <span class="nf">save</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_pre_save</span><span class="p">()</span>
        <span class="n">make_barsplot</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">molecule_type_bars</span><span class="p">)</span>
        <span class="n">make_multi_histogram</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">molecule_len_histogram</span><span class="p">)</span>
        <span class="n">make_histogram</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">mapping_quality_histogram</span><span class="p">)</span>
        <span class="n">make_barsplot</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">position_coverage_bars</span><span class="p">)</span>
        <span class="n">make_barsplot</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">gatc_coverage_bars</span><span class="p">)</span>
        <span class="n">make_rolling_history</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">position_coverage_history</span><span class="p">)</span>
        <span class="n">make_barsplot</span><span class="p">(</span><span class="o">*</span><span class="bp">self</span><span class="o">.</span><span class="n">meth_type_bars</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">as_html</span><span class="p">)</span></div>

<div class="viewcode-block" id="SummaryReport.switch_on"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.switch_on">[docs]</a>    <span class="k">def</span> <span class="nf">switch_on</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attribute</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Method used by descriptors to inform the instance of</span>
<span class="sd">        ``SummaryReport`` that some computed attributes needed by</span>
<span class="sd">        the plots are already computed and usable.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_primary_attributes</span><span class="p">[</span><span class="n">attribute</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span></div>

<div class="viewcode-block" id="SummaryReport.ready_to_go"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.ready_to_go">[docs]</a>    <span class="k">def</span> <span class="nf">ready_to_go</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="o">*</span><span class="n">attrs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Method used to check if some attributes are already usable</span>
<span class="sd">        or not (in other words if they have been already set or not).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">all</span><span class="p">([</span><span class="bp">self</span><span class="o">.</span><span class="n">_primary_attributes</span><span class="p">[</span><span class="n">_</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">attrs</span><span class="p">])</span></div>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">):</span>
        <span class="sd">&quot;&quot;&quot;The items are fetched from either ``self._data``, or directly</span>
<span class="sd">        as attributes of ``self``. The priority is ``self._data``. For</span>
<span class="sd">        example::</span>

<span class="sd">          s = SummaryReport(...)</span>
<span class="sd">          s[&#39;temperature&#39;]</span>

<span class="sd">        will try to return ``s._data[&#39;temperature&#39;]``, and only if</span>
<span class="sd">        missing, ``s.temperature``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">value</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="p">[</span><span class="n">item</span><span class="p">]</span>
        <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">value</span> <span class="o">=</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">item</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">AttributeError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
                <span class="k">raise</span><span class="p">(</span><span class="ne">KeyError</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)))</span>
        <span class="k">return</span> <span class="n">value</span>

    <span class="k">def</span> <span class="fm">__iter__</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="k">return</span> <span class="p">(</span><span class="bp">self</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

<div class="viewcode-block" id="SummaryReport.keys"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.keys">[docs]</a>    <span class="k">def</span> <span class="nf">keys</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
        <span class="n">forbidden</span> <span class="o">=</span> <span class="p">{</span><span class="n">_</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="n">_</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)}</span>
        <span class="n">forbidden</span> <span class="o">|=</span> <span class="p">{</span>
            <span class="s2">&quot;as_html&quot;</span><span class="p">,</span> <span class="s2">&quot;save&quot;</span><span class="p">,</span> <span class="s2">&quot;switch_on&quot;</span><span class="p">,</span> <span class="s2">&quot;ready_to_go&quot;</span><span class="p">,</span>
            <span class="s2">&quot;keys&quot;</span><span class="p">,</span> <span class="s2">&quot;values&quot;</span><span class="p">,</span> <span class="s2">&quot;items&quot;</span>
        <span class="p">}</span>
        <span class="n">methods</span> <span class="o">=</span> <span class="p">{</span><span class="n">_</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="s2">&quot;__call__&quot;</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="n">_</span><span class="p">)}</span>
        <span class="n">descriptors</span> <span class="o">=</span> <span class="p">{</span><span class="n">_</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">dir</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="k">if</span> <span class="n">_</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">forbidden</span><span class="p">}</span>
        <span class="n">from_data</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_data</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>
        <span class="k">return</span> <span class="p">(</span><span class="n">descriptors</span> <span class="o">|</span> <span class="n">from_data</span><span class="p">)</span><span class="o">-</span><span class="n">methods</span></div>

<div class="viewcode-block" id="SummaryReport.dump_molecule_sets"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.dump_molecule_sets">[docs]</a>    <span class="k">def</span> <span class="nf">dump_molecule_sets</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;This method stores in a file the ``_molecule_sets`` attribute.</span>
<span class="sd">        It is done using ``pickle``. The motivation for that is to be</span>
<span class="sd">        able to easily combine several ``SummaryReport`` instances coming</span>
<span class="sd">        from different partitioned analysis. To be able to do that without</span>
<span class="sd">        repeating the filtering process, etc, it is necessary to have</span>
<span class="sd">        the information about what molecules have been discarded for</span>
<span class="sd">        different reasons and what molecules are used from the aligned</span>
<span class="sd">        files.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;wb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">pickle</span><span class="o">.</span><span class="n">dump</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_molecule_sets</span><span class="p">,</span> <span class="n">f</span><span class="p">,</span> <span class="n">protocol</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span></div>

<div class="viewcode-block" id="SummaryReport.load_molecule_sets"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.load_molecule_sets">[docs]</a>    <span class="k">def</span> <span class="nf">load_molecule_sets</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filename</span><span class="p">:</span> <span class="n">Path</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;This method reads data from the file ``filename` (using</span>
<span class="sd">        ``pickle``), it assumes that a dictionary is obtained with</span>
<span class="sd">        the sets of molecule ids (``int``) that are important to</span>
<span class="sd">        re-create the state of the ``SummaryReport`` without going through</span>
<span class="sd">        the ``SingleMoleculeAnalysis`` process all over again.</span>

<span class="sd">        If can be used multiple times and the sets obtained each time</span>
<span class="sd">        will update the current ones (a mathematical union of sets).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filename</span><span class="p">,</span> <span class="s2">&quot;rb&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">pickled_mol_sets</span> <span class="o">=</span> <span class="n">pickle</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="n">f</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">mol_set</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_molecule_sets</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">pickled_mol_set</span> <span class="o">=</span> <span class="n">pickled_mol_sets</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">name</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">pickled_mol_set</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">mol_set</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">pickled_mol_set</span><span class="p">)</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">_loaded_molecule_sets</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">name</span><span class="p">)</span></div></div>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../indepth/index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>