
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>pacbio_data_processing.utils &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/alabaster.css" />
    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
   
  <link rel="stylesheet" href="../../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <h1>Source code for pacbio_data_processing.utils</h1><div class="highlight"><pre>
<span></span><span class="c1">#######################################################################</span>
<span class="c1">#</span>
<span class="c1"># Copyright (C) 2021, 2022 David Palao</span>
<span class="c1">#</span>
<span class="c1"># This file is part of PacBioDataProcessing.</span>
<span class="c1">#</span>
<span class="c1">#  PacBioDataProcessing is free software: you can redistribute it and/or modify</span>
<span class="c1">#  it under the terms of the GNU General Public License as published by</span>
<span class="c1">#  the Free Software Foundation, either version 3 of the License, or</span>
<span class="c1">#  (at your option) any later version.</span>
<span class="c1">#</span>
<span class="c1">#  PacBio data processing is distributed in the hope that it will be useful,</span>
<span class="c1">#  but WITHOUT ANY WARRANTY; without even the implied warranty of</span>
<span class="c1">#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the</span>
<span class="c1">#  GNU General Public License for more details.</span>
<span class="c1">#</span>
<span class="c1">#  You should have received a copy of the GNU General Public License</span>
<span class="c1">#  along with PacBioDataProcessing. If not, see &lt;http://www.gnu.org/licenses/&gt;.</span>
<span class="c1">#</span>
<span class="c1">#######################################################################</span>

<span class="kn">from</span> <span class="nn">math</span> <span class="kn">import</span> <span class="n">log10</span>
<span class="kn">from</span> <span class="nn">typing</span> <span class="kn">import</span> <span class="n">Optional</span><span class="p">,</span> <span class="n">Union</span><span class="p">,</span> <span class="n">TypeVar</span><span class="p">,</span> <span class="n">Any</span><span class="p">,</span> <span class="n">Generic</span><span class="p">,</span> <span class="n">Callable</span>
<span class="kn">from</span> <span class="nn">pathlib</span> <span class="kn">import</span> <span class="n">Path</span>
<span class="kn">from</span> <span class="nn">functools</span> <span class="kn">import</span> <span class="n">cached_property</span>
<span class="kn">from</span> <span class="nn">hashlib</span> <span class="kn">import</span> <span class="n">md5</span>
<span class="kn">from</span> <span class="nn">collections.abc</span> <span class="kn">import</span> <span class="n">Sequence</span>
<span class="kn">import</span> <span class="nn">time</span>
<span class="kn">from</span> <span class="nn">uuid</span> <span class="kn">import</span> <span class="n">uuid1</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">fileinput</span>

<span class="kn">import</span> <span class="nn">Bio.Seq</span>
<span class="kn">import</span> <span class="nn">Bio.SeqIO</span>
<span class="kn">from</span> <span class="nn">Bio.SeqIO</span> <span class="kn">import</span> <span class="n">SeqRecord</span>
<span class="kn">from</span> <span class="nn">pyfaidx</span> <span class="kn">import</span> <span class="n">Faidx</span>

<span class="kn">from</span> <span class="nn">.bam</span> <span class="kn">import</span> <span class="n">BamFile</span>
<span class="kn">from</span> <span class="nn">.types</span> <span class="kn">import</span> <span class="n">PathOrStr</span>
<span class="kn">from</span> <span class="nn">.constants</span> <span class="kn">import</span> <span class="n">WAIT_TIME_SECONDS</span><span class="p">,</span> <span class="n">NEXT_WAIT_TIME_FACTOR</span>


<span class="n">DNASeqType</span> <span class="o">=</span> <span class="n">TypeVar</span><span class="p">(</span><span class="s2">&quot;DNASeqType&quot;</span><span class="p">,</span> <span class="n">bound</span><span class="o">=</span><span class="s2">&quot;DNASeq&quot;</span><span class="p">)</span>
<span class="n">DNASeqLike</span> <span class="o">=</span> <span class="n">TypeVar</span><span class="p">(</span><span class="s2">&quot;DNASeqLike&quot;</span><span class="p">)</span>

<span class="n">_MERGED_FILE_MSG_TEMPLATE</span> <span class="o">=</span> <span class="s2">&quot;Merged file &#39;</span><span class="si">{filename}</span><span class="s2">&#39; created&quot;</span>


<div class="viewcode-block" id="merge_files"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.merge_files">[docs]</a><span class="k">def</span> <span class="nf">merge_files</span><span class="p">(</span>
        <span class="n">infiles</span><span class="p">:</span> <span class="nb">list</span><span class="p">[</span><span class="n">Path</span><span class="p">],</span>
        <span class="n">outfile</span><span class="p">:</span> <span class="n">Path</span><span class="p">,</span>
        <span class="n">keep_only_first_header</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Utility function that concatenates files optionally handling one-line</span>
<span class="sd">    headers correctly: if the files have (one-line) header, it must</span>
<span class="sd">    be declared at call time and then the function will only keep the</span>
<span class="sd">    header found in the first file. All other headers (first line of the</span>
<span class="sd">    remaining files) will be discarded.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">outfile</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outf</span><span class="p">:</span>
        <span class="k">with</span> <span class="n">fileinput</span><span class="o">.</span><span class="n">input</span><span class="p">(</span><span class="n">files</span><span class="o">=</span><span class="n">infiles</span><span class="p">)</span> <span class="k">as</span> <span class="n">inf</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inf</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">keep_only_first_header</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">fileinput</span><span class="o">.</span><span class="n">lineno</span><span class="p">()</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
                        <span class="k">pass</span>
                    <span class="k">elif</span> <span class="n">fileinput</span><span class="o">.</span><span class="n">isfirstline</span><span class="p">():</span>
                        <span class="k">continue</span>
                <span class="n">outf</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
    <span class="n">logging</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="n">_MERGED_FILE_MSG_TEMPLATE</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">filename</span><span class="o">=</span><span class="n">outfile</span><span class="p">))</span></div>


<div class="viewcode-block" id="make_partition_prefix"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.make_partition_prefix">[docs]</a><span class="k">def</span> <span class="nf">make_partition_prefix</span><span class="p">(</span><span class="n">partition</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">partitions</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Simple function to act as a Single Source of Truth for the</span>
<span class="sd">    partition prefix used elsewhere in the project.</span>
<span class="sd">    No validation is done. It just blindly returns a string constructed</span>
<span class="sd">    with the arguments.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="sa">f</span><span class="s2">&quot;partition_</span><span class="si">{</span><span class="n">partition</span><span class="si">}</span><span class="s2">of</span><span class="si">{</span><span class="n">partitions</span><span class="si">}</span><span class="s2">&quot;</span></div>


<div class="viewcode-block" id="combine_scores"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.combine_scores">[docs]</a><span class="k">def</span> <span class="nf">combine_scores</span><span class="p">(</span><span class="n">scores</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">float</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;It computes the combined phred transformed score of the ``scores``</span>
<span class="sd">    provided. Some examples:</span>

<span class="sd">    &gt;&gt;&gt; combine_scores([10])</span>
<span class="sd">    10.0</span>
<span class="sd">    &gt;&gt;&gt; q = combine_scores([10, 12, 14])</span>
<span class="sd">    &gt;&gt;&gt; print(round(q, 6))</span>
<span class="sd">    7.204355</span>
<span class="sd">    &gt;&gt;&gt; q = combine_scores([30, 20, 100, 92])</span>
<span class="sd">    &gt;&gt;&gt; print(round(q, 6))</span>
<span class="sd">    19.590023</span>
<span class="sd">    &gt;&gt;&gt; q_500 = combine_scores([30, 20, 500])</span>
<span class="sd">    &gt;&gt;&gt; q_no_500 = combine_scores([30, 20])</span>
<span class="sd">    &gt;&gt;&gt; q_500 == q_no_500</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; combine_scores([200, 300, 500])</span>
<span class="sd">    200.0</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">p_ok</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="n">scores</span><span class="p">:</span>
        <span class="n">p_ok</span> <span class="o">*=</span> <span class="mi">1</span><span class="o">-</span><span class="mi">10</span><span class="o">**</span><span class="p">(</span><span class="o">-</span><span class="n">q</span><span class="o">/</span><span class="mi">10</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">combined_score</span> <span class="o">=</span> <span class="o">-</span><span class="mi">10</span><span class="o">*</span><span class="n">log10</span><span class="p">(</span><span class="mi">1</span><span class="o">-</span><span class="n">p_ok</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="n">combined_score</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="nb">min</span><span class="p">(</span><span class="n">scores</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">combined_score</span></div>


<div class="viewcode-block" id="DNASeq"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq">[docs]</a><span class="k">class</span> <span class="nc">DNASeq</span><span class="p">(</span><span class="n">Generic</span><span class="p">[</span><span class="n">DNASeqLike</span><span class="p">]):</span>
    <span class="sd">&quot;&quot;&quot;Wrapper around &#39;Bio.Seq.Seq&#39;.&quot;&quot;&quot;</span>
<div class="viewcode-block" id="DNASeq.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span> <span class="n">raw_seq</span><span class="p">:</span> <span class="n">DNASeqLike</span><span class="p">,</span> <span class="n">name</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">,</span> <span class="n">description</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span><span class="p">):</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_seq</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">(</span><span class="n">raw_seq</span><span class="p">)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">description</span> <span class="o">=</span> <span class="n">description</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">fasta_name</span> <span class="o">=</span> <span class="kc">None</span></div>

    <span class="k">def</span> <span class="fm">__eq__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">other</span><span class="p">:</span> <span class="n">DNASeqLike</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="o">==</span> <span class="n">other</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>

    <span class="k">def</span> <span class="fm">__getattr__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">attr</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Any</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">getattr</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="p">,</span> <span class="n">attr</span><span class="p">)</span>

<div class="viewcode-block" id="DNASeq.upper"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.upper">[docs]</a>    <span class="k">def</span> <span class="nf">upper</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Bio</span><span class="o">.</span><span class="n">Seq</span><span class="o">.</span><span class="n">Seq</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span></div>

    <span class="k">def</span> <span class="fm">__getitem__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">idx</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">slice</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">DNASeqType</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">DNASeq</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="p">[</span><span class="n">idx</span><span class="p">])</span>

    <span class="k">def</span> <span class="fm">__len__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
        <span class="k">return</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="p">)</span>

<div class="viewcode-block" id="DNASeq.from_fasta"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.from_fasta">[docs]</a>    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">from_fasta</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">DNASeqType</span><span class="p">],</span> <span class="n">fasta_name</span><span class="p">:</span> <span class="nb">str</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DNASeqType</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Returns a DNASeq from the first DNA sequence stored in</span>
<span class="sd">        the fasta named &#39;fasta_name&#39; after ensuring that the fasta index</span>
<span class="sd">        is there.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">Faidx</span><span class="p">(</span><span class="n">fasta_name</span><span class="p">)</span>
        <span class="n">seq_record_iter</span> <span class="o">=</span> <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fasta_name</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
        <span class="n">record</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="n">seq_record_iter</span><span class="p">)</span>
        <span class="n">seq</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">(</span>
            <span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="n">record</span><span class="o">.</span><span class="n">description</span>
        <span class="p">)</span>
        <span class="n">seq</span><span class="o">.</span><span class="n">fasta_name</span> <span class="o">=</span> <span class="n">fasta_name</span>
        <span class="k">return</span> <span class="n">seq</span></div>

<div class="viewcode-block" id="DNASeq.pi_shifted"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.pi_shifted">[docs]</a>    <span class="k">def</span> <span class="nf">pi_shifted</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DNASeqType</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Method to return a pi-shifted DNASeq from the original one.</span>
<span class="sd">        pi-shifted means that a circular topology is assumed in the</span>
<span class="sd">        DNA sequence and a shift in the origin is done by π radians, ie</span>
<span class="sd">        the sequence is splitted in two parts and both parts are permuted.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span>
        <span class="n">original</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_seq</span>
        <span class="k">return</span> <span class="n">DNASeq</span><span class="p">(</span>
            <span class="n">original</span><span class="p">[</span><span class="n">N</span><span class="o">//</span><span class="mi">2</span><span class="p">:]</span><span class="o">+</span><span class="n">original</span><span class="p">[:</span><span class="n">N</span><span class="o">//</span><span class="mi">2</span><span class="p">],</span> <span class="n">name</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span>
            <span class="n">description</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">description</span><span class="o">+</span><span class="s2">&quot; (pi-shifted)&quot;</span>
        <span class="p">)</span></div>

<div class="viewcode-block" id="DNASeq.write_fasta"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.write_fasta">[docs]</a>    <span class="k">def</span> <span class="nf">write_fasta</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">output_file_name</span><span class="p">:</span> <span class="n">PathOrStr</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="c1"># The next would be useful to have the metadata:</span>
        <span class="c1">#     orig = Bio.SeqIO.parse(self.source_fasta, &quot;fasta&quot;)</span>
        <span class="c1">#     rec = next(orig)</span>
        <span class="c1">#     rec.seq = self._seq</span>
        <span class="n">rec</span> <span class="o">=</span> <span class="n">SeqRecord</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="p">,</span> <span class="nb">id</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">name</span><span class="p">,</span> <span class="n">description</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">description</span><span class="p">)</span>
        <span class="n">Bio</span><span class="o">.</span><span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">([</span><span class="n">rec</span><span class="p">],</span> <span class="n">output_file_name</span><span class="p">,</span> <span class="s2">&quot;fasta&quot;</span><span class="p">)</span>
        <span class="n">Faidx</span><span class="p">(</span><span class="n">output_file_name</span><span class="p">)</span></div>

    <span class="nd">@cached_property</span>
    <span class="k">def</span> <span class="nf">md5sum</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;It returns the MD5 checksum&#39;s hexdigest of the *upper*</span>
<span class="sd">        version of the sequence as a string.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">md5</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_seq</span><span class="p">)</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">encode</span><span class="p">())</span><span class="o">.</span><span class="n">hexdigest</span><span class="p">()</span></div>


<div class="viewcode-block" id="Partition"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition">[docs]</a><span class="k">class</span> <span class="nc">Partition</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;A ``Partition`` is a class that helps answering the following</span>
<span class="sd">    question: assuming that we are interested in processing a fraction</span>
<span class="sd">    of a ``BamFile``, does the molecule ID ``mol_id`` belong to that</span>
<span class="sd">    fraction, or not?</span>
<span class="sd">    A prior implementation consisted in storing all the molecule IDs</span>
<span class="sd">    in the ``BamFile`` corresponding to a given partition in a set, and</span>
<span class="sd">    the answer is just obtained by querying if a molecule ID belongs to</span>
<span class="sd">    the set or not.</span>
<span class="sd">    That former implementation is not enough for the case of multiple</span>
<span class="sd">    alignment processes for the same raw ``BamFile`` (eg, when a</span>
<span class="sd">    combined analysis of the so-called &#39;straight&#39; and &#39;pi-shifted&#39;</span>
<span class="sd">    variants is performed). In that case the partition is decided with</span>
<span class="sd">    one file. And all molecule IDs belonging to the non-empty</span>
<span class="sd">    intersection with the other file must be unambiguously accomodated</span>
<span class="sd">    in a certain partition. This class has been designed to solve</span>
<span class="sd">    that problem.</span>
<span class="sd">    &quot;&quot;&quot;</span>

<div class="viewcode-block" id="Partition.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>
            <span class="bp">self</span><span class="p">,</span>
            <span class="n">partition_specification</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">tuple</span><span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">int</span><span class="p">]],</span>
            <span class="n">bamfile</span><span class="p">:</span> <span class="n">BamFile</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Creates a ``Partition`` object *without* validating the</span>
<span class="sd">        ``partition_specification``, which is done at the time of</span>
<span class="sd">        reading the input given by the user. See</span>
<span class="sd">        :py:class:pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">current</span><span class="p">,</span> <span class="n">num_partitions</span> <span class="o">=</span> <span class="n">partition_specification</span>
        <span class="k">except</span> <span class="ne">TypeError</span><span class="p">:</span>
            <span class="n">current</span><span class="p">,</span> <span class="n">num_partitions</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">1</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">current</span> <span class="o">=</span> <span class="n">current</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">num_partitions</span> <span class="o">=</span> <span class="n">num_partitions</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_bamfile</span> <span class="o">=</span> <span class="n">bamfile</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_delimit_partitions</span><span class="p">()</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_set_current_limits</span><span class="p">()</span></div>

    <span class="nd">@property</span>
    <span class="k">def</span> <span class="nf">is_proper</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;A *proper* partition is one that refers to a proper subset</span>
<span class="sd">        of the given ``BamFile``. Since an empty set is not permitted</span>
<span class="sd">        by the :py:class:SingleMoleculeAnalysisParameters class, an</span>
<span class="sd">        improper partition can only be a partition that refers to the</span>
<span class="sd">        whole ``BamFile``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_partitions</span><span class="p">)</span> <span class="o">==</span> <span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">True</span>

<div class="viewcode-block" id="Partition._delimit_partitions"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition._delimit_partitions">[docs]</a>    <span class="k">def</span> <span class="nf">_delimit_partitions</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        This method decides what are the limits of all partitions</span>
<span class="sd">        given the number of partitions.</span>
<span class="sd">        The method sets an internal mapping, ``self._lower_limits``, of</span>
<span class="sd">        the type ``{partition number [int]: lower limit [int]}``</span>
<span class="sd">        with that information.</span>
<span class="sd">        This mapping is populated with *all* the partition numbers and</span>
<span class="sd">        corresponding values.</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">nmols</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bamfile</span><span class="o">.</span><span class="n">num_molecules</span>
        <span class="n">mols_per_part</span> <span class="o">=</span> <span class="n">nmols</span><span class="o">//</span><span class="bp">self</span><span class="o">.</span><span class="n">num_partitions</span>
        <span class="n">lower_limits</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">all_mols</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">_</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_bamfile</span><span class="o">.</span><span class="n">all_molecules</span><span class="p">]</span>
        <span class="n">all_mols</span><span class="o">.</span><span class="n">sort</span><span class="p">()</span>
        <span class="n">lower_limits</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">for</span> <span class="n">ipart</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_partitions</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="n">lower_limits</span><span class="p">[</span><span class="n">ipart</span><span class="p">]</span> <span class="o">=</span> <span class="n">all_mols</span><span class="p">[</span><span class="n">mols_per_part</span><span class="o">*</span><span class="p">(</span><span class="n">ipart</span><span class="o">-</span><span class="mi">1</span><span class="p">)]</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_lower_limits</span> <span class="o">=</span> <span class="n">lower_limits</span></div>

<div class="viewcode-block" id="Partition._set_current_limits"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition._set_current_limits">[docs]</a>    <span class="k">def</span> <span class="nf">_set_current_limits</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;[Internal method]</span>
<span class="sd">        Auxiliary method for __contains__</span>
<span class="sd">        Here it is determined what is the range of molecule IDs, as</span>
<span class="sd">        ints, that belong to the partition.</span>
<span class="sd">        The method sets two integer attributes, namely:</span>

<span class="sd">        - ``_lower_limit_current``: the minimum molecule ID of the</span>
<span class="sd">          current partition, and</span>
<span class="sd">        - ``_higher_limit_current``: the maximum molecule ID of the</span>
<span class="sd">          current partition; it can be ``None``, meaning that there</span>
<span class="sd">          is no maximum (last partition).</span>

<span class="sd">        :meta public:</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_lower_limit_current</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_lower_limits</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">current</span><span class="p">]</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">current</span> <span class="o">==</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_partitions</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_higher_limit_current</span> <span class="o">=</span> <span class="kc">None</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">_higher_limit_current</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_lower_limits</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">current</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span></div>

    <span class="k">def</span> <span class="fm">__contains__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mol_id</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">bytes</span><span class="p">,</span> <span class="nb">int</span><span class="p">,</span> <span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="nb">bool</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;Main mathod of the Partition class. It decides whether a</span>
<span class="sd">        given molecule ID, ``mol_id`` is within the limits of the</span>
<span class="sd">        partition.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mol_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">mol_id</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">mol_id</span> <span class="o">&gt;=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_lower_limit_current</span><span class="p">:</span>
            <span class="n">lower_check</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">lower_check</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">if</span> <span class="p">((</span><span class="bp">self</span><span class="o">.</span><span class="n">_higher_limit_current</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">)</span> <span class="ow">or</span>
                <span class="p">(</span><span class="n">mol_id</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">_higher_limit_current</span><span class="p">)):</span>
            <span class="n">higher_check</span> <span class="o">=</span> <span class="kc">True</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">higher_check</span> <span class="o">=</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="n">lower_check</span> <span class="ow">and</span> <span class="n">higher_check</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="sd">&quot;&quot;&quot;A string representing the partition that can be used as</span>
<span class="sd">        prefix/suffix in file names created by ``SingleMoleculeAnalysis``.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">make_partition_prefix</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">current</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">num_partitions</span><span class="p">)</span></div>


<div class="viewcode-block" id="find_gatc_positions"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.find_gatc_positions">[docs]</a><span class="k">def</span> <span class="nf">find_gatc_positions</span><span class="p">(</span><span class="n">seq</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">offset</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">set</span><span class="p">[</span><span class="nb">int</span><span class="p">]:</span>
    <span class="sd">&quot;&quot;&quot;Convenience function that computes the positions of all GATCs</span>
<span class="sd">    found in the given sequence.</span>
<span class="sd">    The values are relative to the offset.</span>

<span class="sd">    &gt;&gt;&gt; find_gatc_positions(&#39;AAAGAGAGATCGCGCGATC&#39;) == {7, 15}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; find_gatc_positions(&#39;AAAGAGAGTCGCGCCATC&#39;)</span>
<span class="sd">    set()</span>
<span class="sd">    &gt;&gt;&gt; find_gatc_positions(&#39;AAAGAGAGATCGgaTcCGCGATC&#39;) == {7, 12, 19}</span>
<span class="sd">    True</span>
<span class="sd">    &gt;&gt;&gt; s = find_gatc_positions(&#39;AAAGAGAGATCGgaTcCGCGATC&#39;, offset=23)</span>
<span class="sd">    &gt;&gt;&gt; s == {30, 35, 42}</span>
<span class="sd">    True</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">result</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="n">prev</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">seq</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="k">while</span> <span class="p">(</span><span class="n">pos</span> <span class="o">:=</span> <span class="n">seq</span><span class="o">.</span><span class="n">find</span><span class="p">(</span><span class="s2">&quot;GATC&quot;</span><span class="p">,</span> <span class="n">prev</span><span class="p">))</span> <span class="o">!=</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
        <span class="n">result</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">pos</span><span class="o">+</span><span class="n">offset</span><span class="p">)</span>
        <span class="n">prev</span> <span class="o">=</span> <span class="n">pos</span><span class="o">+</span><span class="mi">1</span>
    <span class="k">return</span> <span class="n">result</span></div>


<div class="viewcode-block" id="shift_me_back"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.shift_me_back">[docs]</a><span class="k">def</span> <span class="nf">shift_me_back</span><span class="p">(</span><span class="n">pos</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span> <span class="n">nbp</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">int</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;Unshifts a given position taking into account that it has been</span>
<span class="sd">    previously shifted by half of the number of base pairs. It takes</span>
<span class="sd">    into account the possibility of having a sequence with an odd</span>
<span class="sd">    length.</span>

<span class="sd">    @params:</span>

<span class="sd">       * pos - 1-based position of a base pair to unshift</span>
<span class="sd">       * nbp - number of base pairs in the reference</span>

<span class="sd">    @returns:</span>

<span class="sd">       * unshifted position</span>

<span class="sd">    Some examples:</span>

<span class="sd">    &gt;&gt;&gt; shift_me_back(3, 10)</span>
<span class="sd">    8</span>
<span class="sd">    &gt;&gt;&gt; shift_me_back(1, 20)</span>
<span class="sd">    11</span>
<span class="sd">    &gt;&gt;&gt; shift_me_back(3, 7)</span>
<span class="sd">    6</span>
<span class="sd">    &gt;&gt;&gt; shift_me_back(4, 7)</span>
<span class="sd">    7</span>
<span class="sd">    &gt;&gt;&gt; shift_me_back(5, 7)</span>
<span class="sd">    1</span>
<span class="sd">    &gt;&gt;&gt; shift_me_back(7, 7)</span>
<span class="sd">    3</span>
<span class="sd">    &gt;&gt;&gt; shift_me_back(1, 7)</span>
<span class="sd">    4</span>

<span class="sd">    To understand the operation of this function consider the following</span>
<span class="sd">    example. Given a sequence of 7 base pairs with the following indices</span>
<span class="sd">    found in the reference in the natural order, ie</span>

<span class="sd">    1 2 3 4 5 6 7</span>

<span class="sd">    then, after being *pi-shifted* the base pairs in the sequence are</span>
<span class="sd">    reordered, and the indices become (in parenthesis the former</span>
<span class="sd">    indices):</span>

<span class="sd">    1&#39;(=4) 2&#39;(=5) 3&#39;(=6) 4&#39;(=7) 5&#39;(=1) 6&#39;(=2) 7&#39;(=3)</span>

<span class="sd">    The current function accepts *primed* indices and transforms them to</span>
<span class="sd">    the *unprimed* indices, ie, the positions returned refer to the</span>
<span class="sd">    original reference.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">shift</span> <span class="o">=</span> <span class="n">nbp</span><span class="o">//</span><span class="mi">2</span>
    <span class="k">if</span> <span class="n">pos</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="n">nbp</span><span class="o">-</span><span class="n">shift</span><span class="p">):</span>
        <span class="n">pos</span> <span class="o">+=</span> <span class="n">shift</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">pos</span> <span class="o">-=</span> <span class="p">(</span><span class="n">nbp</span><span class="o">-</span><span class="n">shift</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">pos</span></div>


<div class="viewcode-block" id="pishift_back_positions_in_gff"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.pishift_back_positions_in_gff">[docs]</a><span class="k">def</span> <span class="nf">pishift_back_positions_in_gff</span><span class="p">(</span><span class="n">gff_path</span><span class="p">:</span> <span class="n">Union</span><span class="p">[</span><span class="nb">str</span><span class="p">,</span> <span class="n">Path</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;A function that parses the input GFF file (assumed to be a valid</span>
<span class="sd">    :term:`GFF` file) and *shifts back* the positions found in it (columns</span>
<span class="sd">    4th and 5th of lines not starting by ``#``).</span>
<span class="sd">    It is assumed that the positions in the input file (``gff_path``)</span>
<span class="sd">    are referring to a *pi-shifted* origin.</span>
<span class="sd">    To undo the shift, the length of the sequence(s) is (are) read</span>
<span class="sd">    from the *GFF3 directives* (lines starting by ``##``), in</span>
<span class="sd">    particular from the ``##sequence-region`` *pragmas*.</span>
<span class="sd">    This function can handle the case of multiple sequences.</span>

<span class="sd">    .. warning::</span>

<span class="sd">        The function overwrites the input ``gff_path``.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gff_path</span><span class="p">)</span> <span class="k">as</span> <span class="n">ingff</span><span class="p">:</span>
        <span class="n">inlines</span> <span class="o">=</span> <span class="n">ingff</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="n">outlines</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">seq_lens</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">inlines</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;##sequence-region&quot;</span><span class="p">):</span>
            <span class="n">_</span><span class="p">,</span> <span class="n">name</span><span class="p">,</span> <span class="n">seq_num</span><span class="p">,</span> <span class="n">seq_len</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="n">N</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">seq_len</span><span class="p">)</span>
            <span class="n">seq_lens</span><span class="p">[</span><span class="n">name</span><span class="p">]</span> <span class="o">=</span> <span class="n">N</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">pieces</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">name</span> <span class="o">=</span> <span class="n">pieces</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">name</span> <span class="ow">in</span> <span class="n">seq_lens</span><span class="p">:</span>
                <span class="n">N</span> <span class="o">=</span> <span class="n">seq_lens</span><span class="p">[</span><span class="n">name</span><span class="p">]</span>
                <span class="n">positions</span> <span class="o">=</span> <span class="p">[</span><span class="nb">int</span><span class="p">(</span><span class="n">_</span><span class="p">)</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">pieces</span><span class="p">[</span><span class="mi">3</span><span class="p">:</span><span class="mi">5</span><span class="p">]]</span>
                <span class="n">new_positions</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">shift_me_back</span><span class="p">(</span><span class="n">_</span><span class="p">,</span> <span class="n">N</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="n">positions</span><span class="p">]</span>
                <span class="n">line</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                    <span class="n">pieces</span><span class="p">[:</span><span class="mi">3</span><span class="p">]</span><span class="o">+</span><span class="n">new_positions</span><span class="o">+</span><span class="n">pieces</span><span class="p">[</span><span class="mi">5</span><span class="p">:]</span>
                <span class="p">)</span>
        <span class="n">outlines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">line</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gff_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">outgff</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">outlines</span><span class="p">:</span>
            <span class="n">outgff</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">line</span><span class="p">)</span></div>


<div class="viewcode-block" id="try_computations_with_variants_until_done"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.try_computations_with_variants_until_done">[docs]</a><span class="k">def</span> <span class="nf">try_computations_with_variants_until_done</span><span class="p">(</span>
        <span class="n">func</span><span class="p">:</span> <span class="n">Callable</span><span class="p">,</span> <span class="n">variants</span><span class="p">:</span> <span class="n">Sequence</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span> <span class="o">*</span><span class="n">args</span><span class="p">:</span> <span class="n">Any</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;This function runs the passed in function ``func`` with the</span>
<span class="sd">    arguments``*args`` and for each ``variant`` in ``variants``,eg.</span>
<span class="sd">    something like this:</span>
<span class="sd">    .. code-block::</span>

<span class="sd">        for v in variants:</span>
<span class="sd">            result = func(*args, variant=v)</span>

<span class="sd">    but it keeps doing so until each result returned by ``func`` is not</span>
<span class="sd">    ``None``. When a ``None`` is returned by ``func``, a call to ``sleep``</span>
<span class="sd">    is warranted before continuing. The time slept depends on how many</span>
<span class="sd">    times it was sleeping before; the sleep time grows exponentially</span>
<span class="sd">    with every iteration:</span>

<span class="sd">    .. code-block::</span>

<span class="sd">        t -&gt; 2*t</span>

<span class="sd">    until all the computations (results of ``func`` for each variant)</span>
<span class="sd">    are completed, ie all are not ``None``.</span>
<span class="sd">    The main application of this function is to ensure that some common</span>
<span class="sd">    operations of the ``SingleMoleculeAnalysis`` are done once and only</span>
<span class="sd">    once irrespective of how many parallel instances of the analysis</span>
<span class="sd">    (with different partitions each) are carried out. For example, this</span>
<span class="sd">    function can be used to avoid collisions in the generation of aligned</span>
<span class="sd">    BAM files since :py:class:`pacbio_data_processing.external.Blasr` has a</span>
<span class="sd">    mechanism that allows concurrent computations.</span>
<span class="sd">    This function delegates the decision on whether the computation is</span>
<span class="sd">    done or not to ``func``.</span>

<span class="sd">    .. note::</span>

<span class="sd">       A special case is when a ``variant`` is ``None``, in that case</span>
<span class="sd">       the function ``func`` is called without the ``variant`` argument:</span>

<span class="sd">       .. code-block::</span>

<span class="sd">          result = func(*args)</span>

<span class="sd">      Therefore, if ``variants`` is, e.g. ``(None,)``, then ``func`` is</span>
<span class="sd">      only called once in each iteration WITHOUT ``variant`` keyword</span>
<span class="sd">      argument. That is useful if the function ``func`` must be called</span>
<span class="sd">      until is done, but it takes no variant argument.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">computations</span> <span class="o">=</span> <span class="p">{</span><span class="n">variant</span><span class="p">:</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">}</span>
    <span class="n">wait_time_seconds</span> <span class="o">=</span> <span class="n">WAIT_TIME_SECONDS</span>
    <span class="n">already_run</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">while</span> <span class="kc">None</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">computations</span><span class="o">.</span><span class="n">values</span><span class="p">()):</span>
        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">variants</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">computations</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">variant</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
                    <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{}</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">kwargs</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;variant&quot;</span><span class="p">:</span> <span class="n">variant</span><span class="p">}</span>
                <span class="n">result</span> <span class="o">=</span> <span class="n">func</span><span class="p">(</span><span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>
                <span class="n">computations</span><span class="p">[</span><span class="n">variant</span><span class="p">]</span> <span class="o">=</span> <span class="n">result</span>
        <span class="k">if</span> <span class="n">already_run</span> <span class="ow">or</span> <span class="p">(</span><span class="kc">None</span> <span class="ow">in</span> <span class="nb">set</span><span class="p">(</span><span class="n">computations</span><span class="o">.</span><span class="n">values</span><span class="p">())):</span>
            <span class="n">time</span><span class="o">.</span><span class="n">sleep</span><span class="p">(</span><span class="n">wait_time_seconds</span><span class="p">)</span>
            <span class="n">wait_time_seconds</span> <span class="o">*=</span> <span class="n">NEXT_WAIT_TIME_FACTOR</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">already_run</span> <span class="o">=</span> <span class="kc">True</span></div>


<div class="viewcode-block" id="AlmostUUID"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.AlmostUUID">[docs]</a><span class="k">class</span> <span class="nc">AlmostUUID</span><span class="p">:</span>
    <span class="sd">&quot;&quot;&quot;A class that provides a 5 letters summary of a UUID.</span>
<span class="sd">    It is intended to be used as prefix in all log messages.</span>
<span class="sd">    It is not necessary that two instances are different. But it is</span>
<span class="sd">    necessary that:</span>

<span class="sd">    1. the string representation is short, and</span>
<span class="sd">    2. given two instances their string representations most</span>
<span class="sd">       *probably differ*.</span>

<span class="sd">    The underlying UUID is obtained from the stdlib using</span>
<span class="sd">    ``uuid.uuid1``.</span>
<span class="sd">    The class is implemented using the Borg pattern: all instances</span>
<span class="sd">    running in the same interpreter share a common ``_uuid`` attribute.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">_uuid</span> <span class="o">=</span> <span class="kc">None</span>

<div class="viewcode-block" id="AlmostUUID.__init__"><a class="viewcode-back" href="../../devel/pacbio_data_processing.html#pacbio_data_processing.utils.AlmostUUID.__init__">[docs]</a>    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="n">_uuid</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="vm">__class__</span><span class="o">.</span><span class="n">_uuid</span> <span class="o">=</span> <span class="n">uuid1</span><span class="p">()</span></div>

    <span class="nd">@cached_property</span>
    <span class="k">def</span> <span class="nf">_short_str</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">([</span><span class="n">_</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">_</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">_uuid</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)])</span>

    <span class="k">def</span> <span class="fm">__str__</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_short_str</span></div>
</pre></div>

          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../indepth/index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../index.html">Documentation overview</a><ul>
  <li><a href="../index.html">Module code</a><ul>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>