
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Single molecule analysis with the sm-analysis program &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/alabaster.css" />
    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Using the package as a library" href="as_library.html" />
    <link rel="prev" title="Filtering Bam files with bam-filter" href="bam_filter.html" />
   
  <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="single-molecule-analysis-with-the-sm-analysis-program">
<span id="sm-analysis"></span><h1>Single molecule analysis with the <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> program<a class="headerlink" href="#single-molecule-analysis-with-the-sm-analysis-program" title="Permalink to this headline">¶</a></h1>
<p><code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> is the main program provided by <strong>PacBio Data Processing</strong>. It analyzes a
given input PacBio BAM file evaluating the methylation status of
each <a class="reference internal" href="../indepth/glossary.html#term-molecule"><span class="xref std std-term">molecule</span></a>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>As mentioned in the <a class="reference internal" href="installation.html#installation"><span class="std std-ref">Installation</span></a> section, by default only
the <a class="reference internal" href="../indepth/glossary.html#term-Command-Line-Interface-CLI"><span class="xref std std-term">command line interface</span></a> for
the single molecule analysis, <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>, is installed. The
corresponding <a class="reference internal" href="../indepth/glossary.html#term-Graphical-User-Interface-GUI"><span class="xref std std-term">graphical user interface</span></a>
can be enabled but requires an extra library: <a class="reference internal" href="../indepth/references.html#wxpython" id="id1"><span>[wxpython]</span></a>.
Both <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> and <code class="docutils literal notranslate"><span class="pre">sm-analysis-gui</span></code> provide access to the
same pipeline, the same program is run under the hood. They only differ
in the interface.</p>
</div>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> expects a PacBio BAM file as first input and the
corresponding <a class="reference internal" href="../indepth/glossary.html#term-FASTA"><span class="xref std std-term">FASTA</span></a> file containing the reference as the second
argument:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sm-analysis myseq.bam bacterium.fasta
</pre></div>
</div>
<p>Other options can be added to that line to customize the default behaviour.
For example, to speed up the analysis we might want to use more than one
<em>CPU core</em>. <strong>PacBio Data Processing</strong> allows us to run several instances of the
<a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a> program in parallel. And we might also
let each <a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a> instance use multiple <em>worker</em>
processes. The following line adds two
<a class="reference internal" href="../indepth/glossary.html#term-Command-Line-Option"><span class="xref std std-term">command line options</span></a> to do all that:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="gp">$ </span>sm-analysis myseq.bam bacterium.fasta -N <span class="m">3</span> -n <span class="m">2</span>
</pre></div>
</div>
<p>which, upon execution, will analize the given BAM file with the given reference
using <code class="docutils literal notranslate"><span class="pre">3</span></code> <a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a> instances, and each instance
will use <code class="docutils literal notranslate"><span class="pre">2</span></code> worker processes in turn.</p>
<p>For a complete list of options available and their description, see the
section <a class="reference internal" href="#clo-sm-analysis"><span class="std std-ref">Command line options</span></a> below.</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>Once the <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> program runs successfully, some files are produced:</p>
<ul class="simple">
<li><p>a <a class="reference internal" href="../indepth/summary_reports.html#summary-report"><span class="std std-ref">summary report</span></a> with statistics about the
<em>Single Molecule Analysis</em> process,</p></li>
<li><p>a single <a class="reference internal" href="../indepth/glossary.html#term-molecule"><span class="xref std std-term">molecule</span></a> summary of all the methylations found (see
<a class="reference internal" href="../indepth/methylation_reports.html#methylation-report-format"><span class="std std-ref">Methylation Reports</span></a> for details),</p></li>
<li><p>a file with the <em>low level</em> information of the analysis, the so-called
<a class="reference internal" href="../indepth/raw_detections.html#raw-detections-format"><span class="std std-ref">raw detections file</span></a>, and</p></li>
<li><p>a concatenated file made up of all the <a class="reference internal" href="../indepth/glossary.html#term-GFF"><span class="xref std std-term">GFF</span></a> files produced
by <a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a> during the analysis.</p></li>
</ul>
</section>
<section id="how-it-works">
<h2>How it works<a class="headerlink" href="#how-it-works" title="Permalink to this headline">¶</a></h2>
<p>The following figure is a flow chart of the <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> pipeline.</p>
<figure class="align-default" id="id2">
<a class="reference internal image-reference" href="../_images/flow_chart_sm_analysis.png"><img alt="sm-analysis" src="../_images/flow_chart_sm_analysis.png" style="width: 1446.5px; height: 1431.5px;" /></a>
<figcaption>
<p><span class="caption-text">Flow chart of the <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> program.</span><a class="headerlink" href="#id2" title="Permalink to this image">¶</a></p>
</figcaption>
</figure>
<p>One of the first things that
the program will do is to ensure that the input BAM is aligned. Actually
two alignment processes will be carried out with the help of
<a class="reference internal" href="../indepth/aligners.html#aligners"><span class="std std-ref">an aligner</span></a> on the input BAM file, each one producing
what we call an <a class="reference internal" href="../indepth/glossary.html#term-alignment-variant"><span class="xref std std-term">alignment variant</span></a>. One is the <em>straight</em> alignment
variant and another is an alignment with a rotated reference, the so-called
<em>pi shifted</em> <a class="reference internal" href="../indepth/glossary.html#term-alignment-variant"><span class="xref std std-term">alignment variant</span></a>, where the starting point of the
reference is rotated by 180 degrees. The aim of this second alignment
process is to catch molecules that cross the origin. With the help of those
two files a complete set of molecules can be detected.</p>
<p>Before running <a class="reference internal" href="../indepth/aligners.html#aligners"><span class="std std-ref">the aligner</span></a> the program will try to
find the two aligned versions of the input file, <em>if it is</em> unaligned.</p>
<p>On the other hand, if the input file is actually aligned, a <em>pi shifted</em>
version of it will be sought. And if found, it will be used.</p>
<p>If only a <em>straight</em> aligned file is at hand, the circular topology of the
reference will not be considered.</p>
<p>To find the aligned versions of the input BAM, the program tries to answer
three questions:</p>
<ol class="arabic simple">
<li><p>Is there a candidate with the expected filename?</p></li>
<li><p>Is the candidate aligned?</p></li>
<li><p>Are the molecules in it a subset of the molecules in the input BAM?</p></li>
</ol>
<p>If the answer to the three questions is yes, the candidate is considered
a <em>plausible aligned version</em> of the input BAM, and it is as such used
within the rest of the analysis. If not, the alignment process is carried
out.</p>
<section id="filters">
<h3>Filters<a class="headerlink" href="#filters" title="Permalink to this headline">¶</a></h3>
<p>The <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> program applies several filters to <em>each</em> <a class="reference internal" href="../indepth/glossary.html#term-molecule"><span class="xref std std-term">molecule</span></a>
in the input BAM file. The aim is to ensure a minimum of quality in the
sequencing data for the processed molecules.
The following filters are applied, in the given order:</p>
<ol class="arabic">
<li><p>A minimum of <code class="docutils literal notranslate"><span class="pre">50</span></code> base pairs is required to each <a class="reference internal" href="../indepth/glossary.html#term-molecule"><span class="xref std std-term">molecule</span></a>.
The sequence corresponding to each molecule is taken from the
aligned CCS BAM file.</p></li>
<li><p>A minimum value is set for the mapping quality (column 5 in each
<a class="reference internal" href="../indepth/glossary.html#term-subread"><span class="xref std std-term">subread</span></a>) to half of the estimated maximum mapping quality found
in the aligned input bam file. That is done using the
<code class="xref py py-func docutils literal notranslate"><span class="pre">pabio_data_processing.bam_utils.estimate_max_mapping_quality()</span></code>
function. This filter is applied to the subreads in each <a class="reference internal" href="../indepth/glossary.html#term-molecule"><span class="xref std std-term">molecule</span></a>.</p></li>
<li><p>At least 90% of the <a class="reference internal" href="../indepth/glossary.html#term-subread"><span class="xref std std-term">subreads</span></a> in each molecule
must have a mapping of <code class="docutils literal notranslate"><span class="pre">0</span></code> or <code class="docutils literal notranslate"><span class="pre">16</span></code>.</p></li>
<li><p>Only <a class="reference internal" href="../indepth/glossary.html#term-subread"><span class="xref std std-term">subreads</span></a> with mapping in the set <code class="docutils literal notranslate"><span class="pre">{0,</span> <span class="pre">16}</span></code>
are taken.</p></li>
<li><p>Only <a class="reference internal" href="../indepth/glossary.html#term-molecule"><span class="xref std std-term">molecules</span></a> with:</p>
<ol class="loweralpha simple">
<li><p>at least 20 <a class="reference internal" href="../indepth/glossary.html#term-subread"><span class="xref std std-term">subreads</span></a>, and</p></li>
<li><p>data for both <em>strands</em> (<code class="docutils literal notranslate"><span class="pre">+</span></code> and <code class="docutils literal notranslate"><span class="pre">-</span></code>)</p></li>
</ol>
<p>are taken.</p>
</li>
</ol>
<p>For details about the technical terms used in the description of the filters
(i.e. what is a <em>mapping</em> or what is the meaning of <em>mapping quality</em>),
please consult the <a class="reference external" href="https://samtools.github.io/hts-specs/SAMv1.pdf">SAM/BAM format</a> specification.</p>
</section>
</section>
<section id="command-line-options">
<span id="clo-sm-analysis"></span><h2>Command line options<a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> program has a <a class="reference internal" href="../indepth/glossary.html#term-Command-Line-Interface-CLI"><span class="xref std std-term">Command Line Interface (CLI)</span></a>
with the following options:</p>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-arg-BAM-FILE">
<span id="cmdoption-sm-analysis-arg-bam-file"></span><span class="sig-name descname"><span class="pre">&lt;BAM-FILE&gt;</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sm-analysis-arg-BAM-FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Input file in BAM format</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-arg-ALIGNMENT-FILE">
<span id="cmdoption-sm-analysis-arg-alignment-file"></span><span class="sig-name descname"><span class="pre">&lt;ALIGNMENT-FILE&gt;</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sm-analysis-arg-ALIGNMENT-FILE" title="Permalink to this definition">¶</a></dt>
<dd><p>Input file containing the alignment in FASTA format (typically a file
ending in ‘.fa’ or ‘.fasta’). A companion ‘.fa.fai’/’.fasta.fai’ file
is also needed but it will be created if not found.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-M">
<span id="cmdoption-sm-analysis-m"></span><span id="cmdoption-sm-analysis-ipd-model"></span><span class="sig-name descname"><span class="pre">-M</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;MODEL&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ipd-model</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;MODEL&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-M" title="Permalink to this definition">¶</a></dt>
<dd><p>Model to be used by ipdSummary to identify the type of modification. MODEL
must be either the model name or the path to the ipd model. First, the
program will make an attempt to interprete MODEL as a path to a file
defining a model; if that fails, MODEL will be understood to be the name
of a model that must be accessible in the resources directory of
<a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">kineticsTools</span></a> (e.g. <code class="docutils literal notranslate"><span class="pre">-M</span> <span class="pre">SP3-C3</span></code> would trigger a search for a
file called <code class="docutils literal notranslate"><span class="pre">SP3-C3.npz.gz</span></code> within the directory with models provided
by <a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">kineticsTools</span></a>). If this option is not given, the default model
in ipdSummary is used.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-a">
<span id="cmdoption-sm-analysis-aligner"></span><span class="sig-name descname"><span class="pre">-a</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--aligner</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-a" title="Permalink to this definition">¶</a></dt>
<dd><p>program to use as <a class="reference internal" href="../indepth/aligners.html#aligners"><span class="std std-ref">aligner</span></a>. It can be a path or an
executable in the <a class="reference internal" href="../indepth/glossary.html#term-PATH"><span class="xref std std-term">PATH</span></a> (default: <a class="reference internal" href="../indepth/aligners.html#pbmm2"><span class="std std-ref">Pbmm2</span></a>)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-p">
<span id="cmdoption-sm-analysis-pbindex"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--pbindex</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-p" title="Permalink to this definition">¶</a></dt>
<dd><p>program to generate indices of BAM files. It must have the same interface
as PacBio’s <code class="docutils literal notranslate"><span class="pre">pbindex</span></code> and it can be a path or an executable in the
<a class="reference internal" href="../indepth/glossary.html#term-PATH"><span class="xref std std-term">PATH</span></a> (default: <a class="reference internal" href="../indepth/pbindex.html#about-pbindex"><span class="std std-ref">pbindex</span></a>)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-i">
<span id="cmdoption-sm-analysis-ipdsummary"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ipdsummary</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-i" title="Permalink to this definition">¶</a></dt>
<dd><p>program to analyze the IPDs. It must have the same interface as PacBio’s
<code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code>. It can be a path or an executable in the <a class="reference internal" href="../indepth/glossary.html#term-PATH"><span class="xref std std-term">PATH</span></a>
(default: <a class="reference internal" href="../indepth/kinetics_tools.html#kineticstools"><span class="std std-ref">ipdSummary</span></a>).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-c">
<span id="cmdoption-sm-analysis-ccs"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ccs</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PROGRAM&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-c" title="Permalink to this definition">¶</a></dt>
<dd><p>program to compute the Hi-Fi version of the input BAM. It must have the
same interface as PacBio’s <code class="docutils literal notranslate"><span class="pre">CCS</span></code>. It can be a path or an executable in
the <a class="reference internal" href="../indepth/glossary.html#term-PATH"><span class="xref std std-term">PATH</span></a> (default: <a class="reference internal" href="../indepth/ccs.html#about-ccs"><span class="std std-ref">ccs</span></a>)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-N">
<span id="cmdoption-sm-analysis-n"></span><span id="cmdoption-sm-analysis-num-simultaneous-ipdsummarys"></span><span class="sig-name descname"><span class="pre">-N</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;INT&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--num-simultaneous-ipdsummarys</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;INT&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-N" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of simultaneous instances of ipdSummary that will
cooperate to process the molecules (default: 1).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-0">
<span id="cmdoption-sm-analysis-num-workers-per-ipdsummary"></span><span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;INT&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--num-workers-per-ipdsummary</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;INT&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-0" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of worker processes that each instance of ipdSummary will
spawn (default: 1).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-nprocs-blasr">
<span class="sig-name descname"><span class="pre">--nprocs-blasr</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;INT&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-nprocs-blasr" title="Permalink to this definition">¶</a></dt>
<dd><p>Number of worker processes that each instance of <a class="reference internal" href="../indepth/aligners.html#aligners"><span class="std std-ref">blasr</span></a>
will spawn (default: 1).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-P">
<span id="cmdoption-sm-analysis-partition"></span><span class="sig-name descname"><span class="pre">-P</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PARTITION:NUMBER-OF-PARTITIONS&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--partition</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PARTITION:NUMBER-OF-PARTITIONS&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-P" title="Permalink to this definition">¶</a></dt>
<dd><p>This option instructs the program to only analyze a fraction
(partition) of the molecules present in the input bam file. The
file is divided in <cite>NUMBER OF PARTITIONS</cite> (almost) equal pieces
but ONLY the PARTITION-th partition (fraction) is analyzed.
For instance, <cite>–partition 3:7</cite> means that the bam file is
divided in seven pieces but only the third piece is analyzed
by the current instance of sm-analysis. By default, all the file
is analyzed.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-C">
<span id="cmdoption-sm-analysis-CCS-bam-file"></span><span id="cmdoption-sm-analysis-ccs-bam-file"></span><span class="sig-name descname"><span class="pre">-C</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PATH&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--CCS-bam-file</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;PATH&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-C" title="Permalink to this definition">¶</a></dt>
<dd><p>The CCS file in BAM format can be optionally provided;
otherwise it is computed. It is necessary to create the
reference mapping between <em>hole numbers</em> and the DNA sequence
of the corresponding fragment, or <em>molecule</em>. After being
aligned, the file will be also used to determine the position
of each molecule in the report of methylation states. If the
CCS BAM file is provided, and any of the necessary aligned
versions of it is not found, the CCS file will be aligned to
be able to get the positions. If this option is not used, a
CCS BAM will be generated from the original BAM file using
the <code class="docutils literal notranslate"><span class="pre">ccs</span></code> program.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-keep-temp-dir">
<span class="sig-name descname"><span class="pre">--keep-temp-dir</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sm-analysis-keep-temp-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Should we keep a copy of the temporary files generated?
(default: No).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-1">
<span id="cmdoption-sm-analysis-modification-types"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;MOD-TYPE&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--modification-types</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;MOD-TYPE&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-1" title="Permalink to this definition">¶</a></dt>
<dd><p>Focus only in the requested modification types (default: m6A).
Multiple space-separated values can be given.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-only-produce-methylation-report">
<span class="sig-name descname"><span class="pre">--only-produce-methylation-report</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sm-analysis-only-produce-methylation-report" title="Permalink to this definition">¶</a></dt>
<dd><p>Use this flag to only produce the methylation report from the
per detection csv file (default: No).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-use-blasr-aligner">
<span class="sig-name descname"><span class="pre">--use-blasr-aligner</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sm-analysis-use-blasr-aligner" title="Permalink to this definition">¶</a></dt>
<dd><p>this option sets blasr as the aligner, instead of the default
aligner (pbmm2)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-mapping-quality-threshold">
<span class="sig-name descname"><span class="pre">--mapping-quality-threshold</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;INT&gt;</span></span><a class="headerlink" href="#cmdoption-sm-analysis-mapping-quality-threshold" title="Permalink to this definition">¶</a></dt>
<dd><p>minimum mapping quality that each individual subread is required
to have in order to pass the filters. The possible mapping quality
values are positive integers in the range [0, 255]
(default: half the estimated maximum value found in the aligned
BAM file).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-sm-analysis-v">
<span id="cmdoption-sm-analysis-verbose"></span><span class="sig-name descname"><span class="pre">-v</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--verbose</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-sm-analysis-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Run the program in <em>verbose mode</em> such that much more runtime
details are produced.</p>
</dd></dl>

</section>
<section id="graphical-user-interface-sm-analysis-gui">
<h2>Graphical User Interface: <code class="docutils literal notranslate"><span class="pre">sm-analysis-gui</span></code><a class="headerlink" href="#graphical-user-interface-sm-analysis-gui" title="Permalink to this headline">¶</a></h2>
<p>Despite the power, beauty and <em>vintage</em> flavor of the command line, <strong>PacBio Data Processing</strong> offers
a <a class="reference internal" href="../indepth/glossary.html#term-Graphical-User-Interface-GUI"><span class="xref std std-term">Graphical User Interface (GUI)</span></a> for its main executable <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>:
<code class="docutils literal notranslate"><span class="pre">sm-analysis-gui</span></code> which, upon execution, will open a window that will allow
you to drive the single molecule analysis.</p>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="../index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="../quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="../indepth/index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="../devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../index.html">Documentation overview</a><ul>
  <li><a href="index.html">Usage</a><ul>
      <li>Previous: <a href="bam_filter.html" title="previous chapter">Filtering Bam files with <code class="docutils literal notranslate"><span class="pre">bam-filter</span></code></a></li>
      <li>Next: <a href="as_library.html" title="next chapter">Using the package as a library</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="../_sources/usage/sm_analysis.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>