
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Index &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="#" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            

<h1 id="index">Index</h1>

<div class="genindex-jumpbox">
 <a href="#Symbols"><strong>Symbols</strong></a>
 | <a href="#_"><strong>_</strong></a>
 | <a href="#A"><strong>A</strong></a>
 | <a href="#B"><strong>B</strong></a>
 | <a href="#C"><strong>C</strong></a>
 | <a href="#D"><strong>D</strong></a>
 | <a href="#E"><strong>E</strong></a>
 | <a href="#F"><strong>F</strong></a>
 | <a href="#G"><strong>G</strong></a>
 | <a href="#H"><strong>H</strong></a>
 | <a href="#I"><strong>I</strong></a>
 | <a href="#J"><strong>J</strong></a>
 | <a href="#K"><strong>K</strong></a>
 | <a href="#L"><strong>L</strong></a>
 | <a href="#M"><strong>M</strong></a>
 | <a href="#N"><strong>N</strong></a>
 | <a href="#O"><strong>O</strong></a>
 | <a href="#P"><strong>P</strong></a>
 | <a href="#R"><strong>R</strong></a>
 | <a href="#S"><strong>S</strong></a>
 | <a href="#T"><strong>T</strong></a>
 | <a href="#U"><strong>U</strong></a>
 | <a href="#V"><strong>V</strong></a>
 | <a href="#W"><strong>W</strong></a>
 
</div>
<h2 id="Symbols">Symbols</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    --aligner &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-a">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --ccs &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-c">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --CCS-bam-file &lt;PATH&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-C">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --ipd-model &lt;MODEL&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-M">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --ipdsummary &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-i">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --keep-temp-dir

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-keep-temp-dir">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --mapping-quality-threshold &lt;INT&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-mapping-quality-threshold">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --mappings &lt;MAPPING&gt;[ &lt;MAPPING&gt; ...]

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-m">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    --min-dna-seq-length &lt;INT&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-l">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    --min-relative-mapping-ratio &lt;NUM&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-R">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    --min-subreads-per-molecule &lt;INT&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-r">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    --modification-types &lt;MOD-TYPE&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-1">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --nprocs-blasr &lt;INT&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-nprocs-blasr">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --num-simultaneous-ipdsummarys &lt;INT&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-N">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --num-workers-per-ipdsummary &lt;INT&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-0">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --only-produce-methylation-report

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-only-produce-methylation-report">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --partition &lt;PARTITION:NUMBER-OF-PARTITIONS&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-P">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --pbindex &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-p">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    --quality-threshold &lt;INT&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-q">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    --use-blasr-aligner

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-use-blasr-aligner">sm-analysis command line option</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    --verbose

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-v">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -a &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-a">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -C &lt;PATH&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-C">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -c &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-c">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -i &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-i">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -l &lt;INT&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-l">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    -m &lt;MAPPING&gt;[ &lt;MAPPING&gt; ...]

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-m">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    -m &lt;MOD-TYPE&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-1">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -M &lt;MODEL&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-M">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -N &lt;INT&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-N">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -n &lt;INT&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-0">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -P &lt;PARTITION:NUMBER-OF-PARTITIONS&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-P">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -p &lt;PROGRAM&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-p">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    -q &lt;INT&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-q">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    -r &lt;INT&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-r">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    -R &lt;NUM&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-R">bam-filter command line option</a>
</li>
      </ul></li>
      <li>
    -v

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-v">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    &lt;ALIGNMENT-FILE&gt;

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-arg-ALIGNMENT-FILE">sm-analysis command line option</a>
</li>
      </ul></li>
      <li>
    &lt;BAM-FILE&gt;

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-arg-BAM-FILE">bam-filter command line option</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-arg-BAM-FILE">sm-analysis command line option</a>
</li>
      </ul></li>
  </ul></td>
</tr></table>

<h2 id="_">_</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_file_filter.BamFilter.__call__">__call__() (pacbio_data_processing.bam_file_filter.BamFilter method)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.Blasr.__call__">(pacbio_data_processing.external.Blasr method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.CCS.__call__">(pacbio_data_processing.external.CCS method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.ExternalProgram.__call__">(pacbio_data_processing.external.ExternalProgram method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.Pbmm2.__call__">(pacbio_data_processing.external.Pbmm2 method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.__call__">(pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.BamFile.__init__">__init__() (pacbio_data_processing.bam.BamFile method)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.BamFileStrategy.__init__">(pacbio_data_processing.bam.BamFileStrategy method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_file_filter.BamFilter.__init__">(pacbio_data_processing.bam_file_filter.BamFilter method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.CircularDNAPosition.__init__">(pacbio_data_processing.bam_utils.CircularDNAPosition method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.__init__">(pacbio_data_processing.bam_utils.Molecule method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.__init__">(pacbio_data_processing.cigar.Cigar method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.ExternalProgram.__init__">(pacbio_data_processing.external.ExternalProgram method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.methylation.MethylationReport.__init__">(pacbio_data_processing.methylation.MethylationReport method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.ParametersBase.__init__">(pacbio_data_processing.parameters.ParametersBase method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sentinel.Sentinel.__init__">(pacbio_data_processing.sentinel.Sentinel method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.__init__">(pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PercAttribute.__init__">(pacbio_data_processing.summary.PercAttribute method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SimpleAttribute.__init__">(pacbio_data_processing.summary.SimpleAttribute method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.__init__">(pacbio_data_processing.summary.SummaryReport method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.AlmostUUID.__init__">(pacbio_data_processing.utils.AlmostUUID method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.__init__">(pacbio_data_processing.utils.DNASeq method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition.__init__">(pacbio_data_processing.utils.Partition method)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._align_bam_if_no_candidate_found">_align_bam_if_no_candidate_found() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sentinel.Sentinel._anti_aging">_anti_aging() (pacbio_data_processing.sentinel.Sentinel method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._collect_statistics">_collect_statistics() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._collect_suitable_molecules_from_ccs">_collect_suitable_molecules_from_ccs() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._create_references">_create_references() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._crosscheck_molecules_in_partition_with_ccs">_crosscheck_molecules_in_partition_with_ccs() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition._delimit_partitions">_delimit_partitions() (pacbio_data_processing.utils.Partition method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._disable_pi_shifted_analysis">_disable_pi_shifted_analysis() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._discard_molecules_with_seq_mismatch">_discard_molecules_with_seq_mismatch() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._dump_results">_dump_results() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._ensure_ccs_bam_aligned">_ensure_ccs_bam_aligned() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._ensure_input_bam_aligned">_ensure_input_bam_aligned() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._exists_pi_shifted_variant_from_aligned_input">_exists_pi_shifted_variant_from_aligned_input() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._filter_molecules">_filter_molecules() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._fix_positions">_fix_positions() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._fix_positions_in_gffs">_fix_positions_in_gffs() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._fix_positions_in_molecules">_fix_positions_in_molecules() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._generate_indices">_generate_indices() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._init_summary">_init_summary() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._ipd_analysis">_ipd_analysis() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._keep_only_pishifted_molecules_crossing_origin">_keep_only_pishifted_molecules_crossing_origin() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis._main">_main() (in module pacbio_data_processing.sm_analysis)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._merge_partitions_if_needed">_merge_partitions_if_needed() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._minimum_mapping_quality">_minimum_mapping_quality (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._post_process_partition">_post_process_partition() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._remove_partition_done_files">_remove_partition_done_files() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._report_discarded_molecules_with_seq_mismatch">_report_discarded_molecules_with_seq_mismatch() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._report_faulty_molecules">_report_faulty_molecules() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._select_molecules">_select_molecules() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._set_aligner">_set_aligner() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition._set_current_limits">_set_current_limits() (pacbio_data_processing.utils.Partition method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis._split_bam">_split_bam() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam._strategy_factory">_strategy_factory() (in module pacbio_data_processing.bam)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="A">A</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.aligned_bam">aligned_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.aligned_ccs_bam_files">aligned_ccs_bam_files (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.AlignedBamAttribute">AlignedBamAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.AlignedCCSBamsAttribute">AlignedCCSBamsAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.aligner_path">aligner_path (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.AlignerMixIn">AlignerMixIn (class in pacbio_data_processing.external)</a>
</li>
      <li><a href="indepth/glossary.html#term-alignment-variant"><strong>alignment variant</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_gatcs_identified_in_bam">all_gatcs_identified_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_gatcs_in_meth">all_gatcs_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_gatcs_not_identified_in_bam">all_gatcs_not_identified_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_gatcs_not_in_meth">all_gatcs_not_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.all_partition_done_filenames">all_partition_done_filenames (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.all_partitions_ready">all_partitions_ready (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_positions_in_bam">all_positions_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_positions_in_meth">all_positions_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_positions_not_in_bam">all_positions_not_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.all_positions_not_in_meth">all_positions_not_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.AlmostUUID">AlmostUUID (class in pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.CircularDNAPosition.as_1base">as_1base() (pacbio_data_processing.bam_utils.CircularDNAPosition method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.as_html">as_html (pacbio_data_processing.summary.SummaryReport property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.ascii_quals">ascii_quals (pacbio_data_processing.bam_utils.Molecule property)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="B">B</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    bam-filter command line option

      <ul>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-m">--mappings &lt;MAPPING&gt;[ &lt;MAPPING&gt; ...]</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-l">--min-dna-seq-length &lt;INT&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-R">--min-relative-mapping-ratio &lt;NUM&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-r">--min-subreads-per-molecule &lt;INT&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-q">--quality-threshold &lt;INT&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-l">-l &lt;INT&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-m">-m &lt;MAPPING&gt;[ &lt;MAPPING&gt; ...]</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-q">-q &lt;INT&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-r">-r &lt;INT&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-R">-R &lt;NUM&gt;</a>
</li>
        <li><a href="usage/bam_filter.html#cmdoption-bam-filter-arg-BAM-FILE">&lt;BAM-FILE&gt;</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.BamFile">BamFile (class in pacbio_data_processing.bam)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.BamFile.bamfile_strategy_name">bamfile_strategy_name (pacbio_data_processing.bam.BamFile attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.BamFileStrategy">BamFileStrategy (class in pacbio_data_processing.bam)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_file_filter.BamFilter">BamFilter (class in pacbio_data_processing.bam_file_filter)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.BamFilteringParameters">BamFilteringParameters (class in pacbio_data_processing.parameters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.BarsPlotAttribute">BarsPlotAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="indepth/glossary.html#term-Bash"><strong>Bash</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.Blasr">Blasr (class in pacbio_data_processing.external)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.body_md5sum">body_md5sum (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="C">C</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.CCS">CCS (class in pacbio_data_processing.external)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.CCS_bam_file">CCS_bam_file (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar">Cigar (class in pacbio_data_processing.cigar)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.cigar">cigar (pacbio_data_processing.bam_utils.Molecule property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.CircularDNAPosition">CircularDNAPosition (class in pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.cleanup_molecules">cleanup_molecules() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.column_name">column_name (pacbio_data_processing.summary.MoleculeLenHistogram attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.combine_scores">combine_scores() (in module pacbio_data_processing.utils)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="indepth/glossary.html#term-Command-Line-Interface-CLI"><strong>Command Line Interface (CLI)</strong></a>
</li>
      <li><a href="indepth/glossary.html#term-Command-Line-Option"><strong>Command Line Option</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.logs.config_logging">config_logging() (in module pacbio_data_processing.logs)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.count_subreads_per_molecule">count_subreads_per_molecule() (in module pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="indepth/glossary.html#term-CPython"><strong>CPython</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.create_raw_detections_file">create_raw_detections_file() (in module pacbio_data_processing.sm_analysis)</a>
</li>
      <li><a href="indepth/glossary.html#term-CSV-file"><strong>CSV file</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.ui.html#pacbio_data_processing.ui.gui.custom_Gooey">custom_Gooey() (in module pacbio_data_processing.ui.gui)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="D">D</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.GATCCoverageBarsPlot.data_definition">data_definition (pacbio_data_processing.summary.GATCCoverageBarsPlot attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethTypeBarsPlot.data_definition">(pacbio_data_processing.summary.MethTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeTypeBarsPlot.data_definition">(pacbio_data_processing.summary.MoleculeTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageBarsPlot.data_definition">(pacbio_data_processing.summary.PositionCoverageBarsPlot attribute)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.data_name">data_name (pacbio_data_processing.summary.MappingQualityHistogram attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.data_name">(pacbio_data_processing.summary.MoleculeLenHistogram attribute)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.dependency_name">dependency_name (pacbio_data_processing.summary.MappingQualityHistogram attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.dependency_name">(pacbio_data_processing.summary.MoleculeLenHistogram attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.dependency_name">(pacbio_data_processing.summary.PositionCoverageHistory attribute)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.GATCCoverageBarsPlot.dependency_names">dependency_names (pacbio_data_processing.summary.GATCCoverageBarsPlot attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethTypeBarsPlot.dependency_names">(pacbio_data_processing.summary.MethTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeTypeBarsPlot.dependency_names">(pacbio_data_processing.summary.MoleculeTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageBarsPlot.dependency_names">(pacbio_data_processing.summary.PositionCoverageBarsPlot attribute)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.diff_ratio">diff_ratio (pacbio_data_processing.cigar.Cigar property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.dna">dna (pacbio_data_processing.bam_utils.Molecule property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq">DNASeq (class in pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.dump_molecule_sets">dump_molecule_sets() (pacbio_data_processing.summary.SummaryReport method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="E">E</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.empty_buffer">empty_buffer() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.end">end (pacbio_data_processing.bam_utils.Molecule property)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.estimate_max_mapping_quality">estimate_max_mapping_quality() (in module pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="indepth/glossary.html#term-exit-status"><strong>exit status</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.ExternalProgram">ExternalProgram (class in pacbio_data_processing.external)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="F">F</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="indepth/glossary.html#term-FASTA"><strong>FASTA</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.faulty_mols">faulty_mols (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.faulty_subreads">faulty_subreads (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.filter_enough_data_per_molecule">filter_enough_data_per_molecule() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.BamFilteringParameters.filter_mappings">filter_mappings (pacbio_data_processing.parameters.BamFilteringParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.filter_mappings_binary">filter_mappings_binary() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.filter_mappings_ratio">filter_mappings_ratio() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.filter_quality">filter_quality() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.filters.filter_seq_len">filter_seq_len() (in module pacbio_data_processing.filters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.filtered_out_mols">filtered_out_mols (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.filtered_out_subreads">filtered_out_subreads (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.find_gatc_positions">find_gatc_positions() (in module pacbio_data_processing.utils)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.find_gatc_positions">(pacbio_data_processing.bam_utils.Molecule method)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.flag2strand">flag2strand() (in module pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.from_fasta">from_fasta() (pacbio_data_processing.utils.DNASeq class method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.full_md5sum">full_md5sum (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.fully_methylated_gatcs">fully_methylated_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.fully_methylated_gatcs_wrt_meth">fully_methylated_gatcs_wrt_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.fully_unmethylated_gatcs">fully_unmethylated_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.fully_unmethylated_gatcs_wrt_meth">fully_unmethylated_gatcs_wrt_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="G">G</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.gatc_coverage_bars">gatc_coverage_bars (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.GATCCoverageBarsPlot">GATCCoverageBarsPlot (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.gen_index_single_molecule_bams">gen_index_single_molecule_bams() (in module pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.generate_CCS_file">generate_CCS_file() (in module pacbio_data_processing.sm_analysis)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="indepth/glossary.html#term-GFF"><strong>GFF</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.gff_result">gff_result (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="indepth/glossary.html#term-Graphical-User-Interface-GUI"><strong>Graphical User Interface (GUI)</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.ui.html#pacbio_data_processing.ui.gui.guiize_options">guiize_options() (in module pacbio_data_processing.ui.gui)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="H">H</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.hemi_methylated_gatcs">hemi_methylated_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.hemi_methylated_gatcs_wrt_meth">hemi_methylated_gatcs_wrt_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.hemi_minus_methylated_gatcs">hemi_minus_methylated_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.hemi_minus_methylated_gatcs_wrt_meth">hemi_minus_methylated_gatcs_wrt_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.hemi_plus_methylated_gatcs">hemi_plus_methylated_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.hemi_plus_methylated_gatcs_wrt_meth">hemi_plus_methylated_gatcs_wrt_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.errors.high_level_handler">high_level_handler() (in module pacbio_data_processing.errors)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.HistoryPlotAttribute">HistoryPlotAttribute (class in pacbio_data_processing.summary)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="I">I</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.id">id (pacbio_data_processing.bam_utils.Molecule attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.GATCCoverageBarsPlot.index_labels">index_labels (pacbio_data_processing.summary.GATCCoverageBarsPlot attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethTypeBarsPlot.index_labels">(pacbio_data_processing.summary.MethTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeTypeBarsPlot.index_labels">(pacbio_data_processing.summary.MoleculeTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageBarsPlot.index_labels">(pacbio_data_processing.summary.PositionCoverageBarsPlot attribute)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.input_bam">input_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.input_bam_size">input_bam_size (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.input_reference">input_reference (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.InputBamAttribute">InputBamAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.InputReferenceAttribute">InputReferenceAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.ipd_model">ipd_model (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.ipd.ipd_summary">ipd_summary() (in module pacbio_data_processing.ipd)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.is_crossing_origin">is_crossing_origin() (pacbio_data_processing.bam_utils.Molecule method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sentinel.Sentinel.is_file_too_old">is_file_too_old (pacbio_data_processing.sentinel.Sentinel property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition.is_proper">is_proper (pacbio_data_processing.utils.Partition property)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="J">J</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.join_gffs">join_gffs() (in module pacbio_data_processing.bam_utils)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.joint_gff_filename">joint_gff_filename (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="K">K</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.keys">keys() (pacbio_data_processing.summary.SummaryReport method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="L">L</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.labels">labels (pacbio_data_processing.summary.MoleculeLenHistogram attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.labels">(pacbio_data_processing.summary.PositionCoverageHistory attribute)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.legend">legend (pacbio_data_processing.summary.MappingQualityHistogram attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.legend">(pacbio_data_processing.summary.MoleculeLenHistogram attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.legend">(pacbio_data_processing.summary.PositionCoverageHistory attribute)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.len_column_name">len_column_name (pacbio_data_processing.summary.PositionCoverageHistory attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.BamFilteringParameters.limit_mappings">limit_mappings (pacbio_data_processing.parameters.BamFilteringParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.load_molecule_sets">load_molecule_sets() (pacbio_data_processing.summary.SummaryReport method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="M">M</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_file_filter.main">main() (in module pacbio_data_processing.bam_file_filter)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.main_cl">main_cl() (in module pacbio_data_processing.sm_analysis)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis_gui.main_gui">main_gui() (in module pacbio_data_processing.sm_analysis_gui)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.plots.make_barsplot">make_barsplot() (in module pacbio_data_processing.plots)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.plots.make_continuous_rolled_data">make_continuous_rolled_data() (in module pacbio_data_processing.plots)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.make_data_for_plot">make_data_for_plot() (pacbio_data_processing.summary.MappingQualityHistogram method)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.make_data_for_plot">(pacbio_data_processing.summary.MoleculeLenHistogram method)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.make_data_for_plot">(pacbio_data_processing.summary.PositionCoverageHistory method)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.plots.make_histogram">make_histogram() (in module pacbio_data_processing.plots)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.plots.make_multi_histogram">make_multi_histogram() (in module pacbio_data_processing.plots)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.make_partition_prefix">make_partition_prefix() (in module pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.plots.make_rolling_history">make_rolling_history() (in module pacbio_data_processing.plots)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.make_up_extra_args">make_up_extra_args() (pacbio_data_processing.summary.MappingQualityHistogram method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.map_molecules_with_highest_sim_ratio">map_molecules_with_highest_sim_ratio() (in module pacbio_data_processing.sm_analysis)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mapping_qualities">mapping_qualities (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mapping_quality_histogram">mapping_quality_histogram (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mapping_quality_threshold">mapping_quality_threshold (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram">MappingQualityHistogram (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityThresholdAttribute">MappingQualityThresholdAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.methylation.match_methylation_states_m6A">match_methylation_states_m6A() (in module pacbio_data_processing.methylation)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.max_possible_methylations">max_possible_methylations (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="indepth/glossary.html#term-MD5-checksum"><strong>MD5 checksum</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.md5sum">md5sum (pacbio_data_processing.utils.DNASeq property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.merge_files">merge_files() (in module pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.meth_type_bars">meth_type_bars (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethTypeBarsPlot">MethTypeBarsPlot (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.methylation_report">methylation_report (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.methylation.MethylationReport">MethylationReport (class in pacbio_data_processing.methylation)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethylationReport">(class in pacbio_data_processing.summary)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.BamFilteringParameters.min_relative_mapping_ratio">min_relative_mapping_ratio (pacbio_data_processing.parameters.BamFilteringParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.MissingExternalToolError">MissingExternalToolError</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.errors.MissingGooeyError">MissingGooeyError</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.ipd.MissingIpdSummaryError">MissingIpdSummaryError</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.methylation.MethylationReport.modification_types">modification_types (pacbio_data_processing.methylation.MethylationReport property)</a>
</li>
      <li><a href="indepth/glossary.html#term-module"><strong>module</strong></a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing">pacbio_data_processing</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.bam">pacbio_data_processing.bam</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.bam_file_filter">pacbio_data_processing.bam_file_filter</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.bam_utils">pacbio_data_processing.bam_utils</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.cigar">pacbio_data_processing.cigar</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.constants">pacbio_data_processing.constants</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.errors">pacbio_data_processing.errors</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.external">pacbio_data_processing.external</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.filters">pacbio_data_processing.filters</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.ipd">pacbio_data_processing.ipd</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.logs">pacbio_data_processing.logs</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.methylation">pacbio_data_processing.methylation</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.parameters">pacbio_data_processing.parameters</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.plots">pacbio_data_processing.plots</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sam">pacbio_data_processing.sam</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sentinel">pacbio_data_processing.sentinel</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sm_analysis">pacbio_data_processing.sm_analysis</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sm_analysis_gui">pacbio_data_processing.sm_analysis_gui</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.summary">pacbio_data_processing.summary</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.templates">pacbio_data_processing.templates</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.types">pacbio_data_processing.types</a>
</li>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui">pacbio_data_processing.ui</a>
</li>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.cl">pacbio_data_processing.ui.cl</a>
</li>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.common">pacbio_data_processing.ui.common</a>
</li>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.gui">pacbio_data_processing.ui.gui</a>
</li>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.options">pacbio_data_processing.ui.options</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.utils">pacbio_data_processing.utils</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="indepth/glossary.html#term-molecule"><strong>molecule</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule">Molecule (class in pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.molecule_len_histogram">molecule_len_histogram (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.molecule_type_bars">molecule_type_bars (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram">MoleculeLenHistogram (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeTypeBarsPlot">MoleculeTypeBarsPlot (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mols_dna_mismatches">mols_dna_mismatches (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mols_in_meth_report">mols_in_meth_report (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mols_in_meth_report_with_gatcs">mols_in_meth_report_with_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mols_in_meth_report_without_gatcs">mols_in_meth_report_without_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mols_ini">mols_ini (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.mols_used_in_aligned_ccs">mols_used_in_aligned_ccs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MolsSetAttribute">MolsSetAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.ipd.multi_ipd_summary">multi_ipd_summary() (in module pacbio_data_processing.ipd)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.ipd.multi_ipd_summary_direct">multi_ipd_summary_direct() (in module pacbio_data_processing.ipd)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.ipd.multi_ipd_summary_threads">multi_ipd_summary_threads() (in module pacbio_data_processing.ipd)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="N">N</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.number_diff_items">number_diff_items (pacbio_data_processing.cigar.Cigar property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.number_diff_types">number_diff_types (pacbio_data_processing.cigar.Cigar property)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.number_pb_diffs">number_pb_diffs (pacbio_data_processing.cigar.Cigar property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.number_pbs">number_pbs (pacbio_data_processing.cigar.Cigar property)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="O">O</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.old_single_molecule_work_units_gen">old_single_molecule_work_units_gen() (in module pacbio_data_processing.bam_utils)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.BamFilteringParameters.out_bam_file">out_bam_file (pacbio_data_processing.parameters.BamFilteringParameters property)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="P">P</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    pacbio_data_processing

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.bam

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.bam">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.bam_file_filter

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.bam_file_filter">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.bam_utils

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.bam_utils">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.cigar

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.cigar">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.constants

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.constants">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.errors

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.errors">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.external

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.external">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.filters

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.filters">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.ipd

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.ipd">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.logs

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.logs">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.methylation

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.methylation">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.parameters

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.parameters">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.plots

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.plots">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.sam

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sam">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.sentinel

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sentinel">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.sm_analysis

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sm_analysis">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.sm_analysis_gui

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.sm_analysis_gui">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.summary

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.summary">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.templates

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.templates">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.types

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.types">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.ui

      <ul>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.ui.cl

      <ul>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.cl">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.ui.common

      <ul>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.common">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.ui.gui

      <ul>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.gui">module</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li>
    pacbio_data_processing.ui.options

      <ul>
        <li><a href="devel/pacbio_data_processing.ui.html#module-pacbio_data_processing.ui.options">module</a>
</li>
      </ul></li>
      <li>
    pacbio_data_processing.utils

      <ul>
        <li><a href="devel/pacbio_data_processing.html#module-pacbio_data_processing.utils">module</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.pack_lines">pack_lines() (in module pacbio_data_processing.bam)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.ParametersBase">ParametersBase (class in pacbio_data_processing.parameters)</a>
</li>
      <li><a href="devel/pacbio_data_processing.ui.html#pacbio_data_processing.ui.cl.parse_cl_bam_filter">parse_cl_bam_filter() (in module pacbio_data_processing.ui.cl)</a>
</li>
      <li><a href="devel/pacbio_data_processing.ui.html#pacbio_data_processing.ui.cl.parse_cl_sm_analysis">parse_cl_sm_analysis() (in module pacbio_data_processing.ui.cl)</a>
</li>
      <li><a href="devel/pacbio_data_processing.ui.html#pacbio_data_processing.ui.common.parse_user_input">parse_user_input() (in module pacbio_data_processing.ui.common)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.Partition">Partition (class in pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.partition">partition (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.partition">(pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis property)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.partition_done_filename">partition_done_filename (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>
</li>
      <li><a href="indepth/glossary.html#term-PATH"><strong>PATH</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.external.Pbmm2">Pbmm2 (class in pacbio_data_processing.external)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_gatcs_identified_in_bam">perc_all_gatcs_identified_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_gatcs_in_meth">perc_all_gatcs_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_gatcs_not_identified_in_bam">perc_all_gatcs_not_identified_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_gatcs_not_in_meth">perc_all_gatcs_not_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_positions_in_bam">perc_all_positions_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_positions_in_meth">perc_all_positions_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_positions_not_in_bam">perc_all_positions_not_in_bam (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_all_positions_not_in_meth">perc_all_positions_not_in_meth (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_faulty_mols">perc_faulty_mols (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_faulty_subreads">perc_faulty_subreads (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_filtered_out_mols">perc_filtered_out_mols (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_filtered_out_subreads">perc_filtered_out_subreads (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_mols_dna_mismatches">perc_mols_dna_mismatches (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_mols_in_meth_report">perc_mols_in_meth_report (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_mols_in_meth_report_with_gatcs">perc_mols_in_meth_report_with_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_mols_in_meth_report_without_gatcs">perc_mols_in_meth_report_without_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_mols_used_in_aligned_ccs">perc_mols_used_in_aligned_ccs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_dna_mismatches">perc_subreads_dna_mismatches (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_in_meth_report">perc_subreads_in_meth_report (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_in_meth_report_with_gatcs">perc_subreads_in_meth_report_with_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_in_meth_report_without_gatcs">perc_subreads_in_meth_report_without_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_used_in_aligned_ccs">perc_subreads_used_in_aligned_ccs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_with_high_mapq">perc_subreads_with_high_mapq (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.perc_subreads_with_low_mapq">perc_subreads_with_low_mapq (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PercAttribute">PercAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.pi_shift_back">pi_shift_back() (pacbio_data_processing.bam_utils.Molecule method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.pi_shifted">pi_shifted() (pacbio_data_processing.utils.DNASeq method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.pishift_back_positions_in_gff">pishift_back_positions_in_gff() (in module pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.position_coverage_bars">position_coverage_bars (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.position_coverage_history">position_coverage_history (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageBarsPlot">PositionCoverageBarsPlot (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory">PositionCoverageHistory (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.methylation.MethylationReport.PRELOG">PRELOG (pacbio_data_processing.methylation.MethylationReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.produce_methylation_report">produce_methylation_report() (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="R">R</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.raw_detections">raw_detections (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.raw_detections_filename">raw_detections_filename (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.ready_to_go">ready_to_go() (pacbio_data_processing.summary.SummaryReport method)</a>
</li>
      <li><a href="indepth/glossary.html#term-reference"><strong>reference</strong></a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.reference_base_pairs">reference_base_pairs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.reference_md5sum">reference_md5sum (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.reference_name">reference_name (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.restore_old_run">restore_old_run() (in module pacbio_data_processing.sm_analysis)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.ROAttribute">ROAttribute (class in pacbio_data_processing.summary)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="S">S</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.methylation.MethylationReport.save">save() (pacbio_data_processing.methylation.MethylationReport method)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.save">(pacbio_data_processing.summary.SummaryReport method)</a>
</li>
      </ul></li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sentinel.Sentinel">Sentinel (class in pacbio_data_processing.sentinel)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sentinel.SentinelFileFound">SentinelFileFound</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sentinel.SentinelFileNotFound">SentinelFileNotFound</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam.set_pysam_verbosity">set_pysam_verbosity() (in module pacbio_data_processing.bam)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.shift_me_back">shift_me_back() (in module pacbio_data_processing.utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.cigar.Cigar.sim_ratio">sim_ratio (pacbio_data_processing.cigar.Cigar property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SimpleAttribute">SimpleAttribute (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.single_molecule_work_units_gen">single_molecule_work_units_gen() (in module pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis">SingleMoleculeAnalysis (class in pacbio_data_processing.sm_analysis)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters">SingleMoleculeAnalysisParameters (class in pacbio_data_processing.parameters)</a>
</li>
      <li>
    sm-analysis command line option

      <ul>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-a">--aligner &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-c">--ccs &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-C">--CCS-bam-file &lt;PATH&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-M">--ipd-model &lt;MODEL&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-i">--ipdsummary &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-keep-temp-dir">--keep-temp-dir</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-mapping-quality-threshold">--mapping-quality-threshold &lt;INT&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-1">--modification-types &lt;MOD-TYPE&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-nprocs-blasr">--nprocs-blasr &lt;INT&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-N">--num-simultaneous-ipdsummarys &lt;INT&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-0">--num-workers-per-ipdsummary &lt;INT&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-only-produce-methylation-report">--only-produce-methylation-report</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-P">--partition &lt;PARTITION:NUMBER-OF-PARTITIONS&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-p">--pbindex &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-use-blasr-aligner">--use-blasr-aligner</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-v">--verbose</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-a">-a &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-C">-C &lt;PATH&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-c">-c &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-i">-i &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-1">-m &lt;MOD-TYPE&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-M">-M &lt;MODEL&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-N">-N &lt;INT&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-0">-n &lt;INT&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-P">-P &lt;PARTITION:NUMBER-OF-PARTITIONS&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-p">-p &lt;PROGRAM&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-v">-v</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-arg-ALIGNMENT-FILE">&lt;ALIGNMENT-FILE&gt;</a>
</li>
        <li><a href="usage/sm_analysis.html#cmdoption-sm-analysis-arg-BAM-FILE">&lt;BAM-FILE&gt;</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.errors.SMAMergeError">SMAMergeError</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.errors.SMAPipelineError">SMAPipelineError</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.split_bam_file_in_molecules">split_bam_file_in_molecules() (in module pacbio_data_processing.bam_utils)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.src_bam_path">src_bam_path (pacbio_data_processing.bam_utils.Molecule attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.Molecule.start">start (pacbio_data_processing.bam_utils.Molecule property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.start_column_name">start_column_name (pacbio_data_processing.summary.PositionCoverageHistory attribute)</a>
</li>
      <li><a href="indepth/glossary.html#term-subread"><strong>subread</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_aligned_ini">subreads_aligned_ini (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_dna_mismatches">subreads_dna_mismatches (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_in_meth_report">subreads_in_meth_report (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_in_meth_report_with_gatcs">subreads_in_meth_report_with_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_in_meth_report_without_gatcs">subreads_in_meth_report_without_gatcs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_ini">subreads_ini (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_used_in_aligned_ccs">subreads_used_in_aligned_ccs (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_with_high_mapq">subreads_with_high_mapq (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.subreads_with_low_mapq">subreads_with_low_mapq (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="indepth/glossary.html#term-summary-report"><strong>summary report</strong></a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters.summary_report_html_filename">summary_report_html_filename (pacbio_data_processing.parameters.SingleMoleculeAnalysisParameters property)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport">SummaryReport (class in pacbio_data_processing.summary)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.switch_on">switch_on() (pacbio_data_processing.summary.SummaryReport method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="T">T</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.GATCCoverageBarsPlot.title">title (pacbio_data_processing.summary.GATCCoverageBarsPlot attribute)</a>

      <ul>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MappingQualityHistogram.title">(pacbio_data_processing.summary.MappingQualityHistogram attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MethTypeBarsPlot.title">(pacbio_data_processing.summary.MethTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeLenHistogram.title">(pacbio_data_processing.summary.MoleculeLenHistogram attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.MoleculeTypeBarsPlot.title">(pacbio_data_processing.summary.MoleculeTypeBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageBarsPlot.title">(pacbio_data_processing.summary.PositionCoverageBarsPlot attribute)</a>
</li>
        <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.PositionCoverageHistory.title">(pacbio_data_processing.summary.PositionCoverageHistory attribute)</a>
</li>
      </ul></li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.summary.SummaryReport.total_gatcs_in_ref">total_gatcs_in_ref (pacbio_data_processing.summary.SummaryReport attribute)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.try_computations_with_variants_until_done">try_computations_with_variants_until_done() (in module pacbio_data_processing.utils)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="U">U</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.ipd.UnknownErrorIpdSummary">UnknownErrorIpdSummary</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.upper">upper() (pacbio_data_processing.utils.DNASeq method)</a>
</li>
  </ul></td>
</tr></table>

<h2 id="V">V</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="indepth/glossary.html#term-variant"><strong>variant</strong></a>
</li>
  </ul></td>
</tr></table>

<h2 id="W">W</h2>
<table style="width: 100%" class="indextable genindextable"><tr>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis.workdir">workdir (pacbio_data_processing.sm_analysis.SingleMoleculeAnalysis property)</a>
</li>
  </ul></td>
  <td style="width: 33%; vertical-align: top;"><ul>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.utils.DNASeq.write_fasta">write_fasta() (pacbio_data_processing.utils.DNASeq method)</a>
</li>
      <li><a href="devel/pacbio_data_processing.html#pacbio_data_processing.bam_utils.write_one_molecule_bam">write_one_molecule_bam() (in module pacbio_data_processing.bam_utils)</a>
</li>
  </ul></td>
</tr></table>



          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="indepth/index.html">Guides</a></li>
<li class="toctree-l1"><a class="reference internal" href="devel/index.html">Developers documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velzquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
    </div>

    

    
  </body>
</html>