
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

    <title>Changelog &#8212; PacBio Data Processing 1.2.3 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css" />
    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Credits" href="authors.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="changelog">
<span id="id1"></span><h1>Changelog<a class="headerlink" href="#changelog" title="Permalink to this headline">¶</a></h1>
<section id="version-1-2-3">
<h2>Version 1.2.3<a class="headerlink" href="#version-1-2-3" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>New mechanism to choose the minimum mapping quality in the filters
applied by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> to the input BAM.
It defaults to an estimation of half the maximum value of the mapping
quality found in the aligned file, but it can be manually set using the
<code class="docutils literal notranslate"><span class="pre">--mapping-quality-threshold</span></code> command line option.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> can use the blasr aligner if the <code class="docutils literal notranslate"><span class="pre">--use-blasr-aligner</span></code>
flag is given in the command line.</p></li>
<li><p>Issue #105 closed: the pbmm2 aligner is called with <code class="docutils literal notranslate"><span class="pre">--preset</span> <span class="pre">SUBREAD</span></code>
when a <em>normal</em> BAM file is aligned and with <code class="docutils literal notranslate"><span class="pre">--preset</span> <span class="pre">CCS</span></code> if the
output of the <code class="docutils literal notranslate"><span class="pre">ccs</span></code> program is processed.</p></li>
</ul>
</section>
<section id="version-1-2-0">
<h2>Version 1.2.0<a class="headerlink" href="#version-1-2-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Results of multiple partitions are merged after the last partition
is finished (issue #92)</p></li>
<li><p>(dev) Link to pre-compiled wxPython fixed in Pipelines (issue #109).</p></li>
</ul>
</section>
<section id="version-1-1-0">
<h2>Version 1.1.0<a class="headerlink" href="#version-1-1-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Fixed issue #51: new algorithm to split the input BAM file into
one-molecule BAM files. Now, even if the input BAM is not sorted
by molecule id (ZMW), the resulting one-molecule BAMs include
all the subreads for each molecule.</p></li>
<li><p>Fixed issue #106: reading BAM files whose line lengths
are different works now (as v1.6 of the SAM/BAM requires)</p></li>
<li><p>Fixed issue #100: Makefile is using flit now</p></li>
<li><p>Fixed issue #101: slow plots with seaborn-0.12.0</p></li>
<li><p>Several minor improvements in the docs.</p></li>
<li><p>The time to check if a <code class="docutils literal notranslate"><span class="pre">BamFile</span></code> is aligned is bounded
by setting a max. amount of lines that are checked (issue #94).</p></li>
<li><p>Fixed issue #91: reference window analyzed by <code class="docutils literal notranslate"><span class="pre">ipdSummary</span></code>
reduced to molecules’ boundaries.</p></li>
</ul>
</section>
<section id="version-1-0-0">
<h2>Version 1.0.0<a class="headerlink" href="#version-1-0-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Summary Report includes information about <em>faulty molecules</em>:
molecules for which some external tool experienced a problem.</p></li>
<li><p>Figures created in partitioned analysis have prefixes according
to partition number (issue #86).</p></li>
<li><p>If an external tool is missing, <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> provides a hint to
install it.</p></li>
<li><p>Informative error message if Gooey is not installed
and <code class="docutils literal notranslate"><span class="pre">sm-analysis-gui</span></code> is launched (issue #49).</p></li>
<li><p>Wrong input partitions trigger error message (issue #12).</p></li>
<li><p>Documentation</p></li>
</ul>
</section>
<section id="version-0-20-0">
<h2>Version 0.20.0<a class="headerlink" href="#version-0-20-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Command line option to provide path to <code class="docutils literal notranslate"><span class="pre">ccs</span></code> program.</p></li>
<li><p>Fasta index is created for the original reference if it is missing
(issue #55)</p></li>
<li><p>Removed runtime dependency on <code class="docutils literal notranslate"><span class="pre">setuptools</span></code> (issue #81).</p></li>
</ul>
</section>
<section id="version-0-19-0">
<h2>Version 0.19.0<a class="headerlink" href="#version-0-19-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> filters out molecules with subreads in only one
strand (issue #77).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">blasr</span></code> and <code class="docutils literal notranslate"><span class="pre">ccs</span></code> operations protected by sentinel to prevent
that parallel computations overwrite the aligned/ccs files (iss. #64).</p></li>
<li><p>Bug fixed in plot with rolling average of coverage</p></li>
<li><p>Issues: #58, #67</p></li>
<li><p>Added Unique ID to each line of output (issue #14)</p></li>
<li><p>Some improvements in the documentation.</p></li>
<li><p>The result of calling external tools is taken into account in the
pipeline: to inform about a possible error and to omit a faulty
molecule.</p></li>
</ul>
</section>
<section id="version-0-18-0">
<h2>Version 0.18.0<a class="headerlink" href="#version-0-18-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Summary report for humans in HTML format with</p>
<ul>
<li><p>plots, and</p></li>
<li><p>basic statistics about subreads, molecules, methylations, etc.</p></li>
</ul>
</li>
<li><p>Issues: #59</p></li>
<li><p>Bugfixes: #61, #62, #65</p></li>
</ul>
</section>
<section id="version-0-17-0">
<h2>Version 0.17.0<a class="headerlink" href="#version-0-17-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>The molecules in the methylation report are the ones found
in the aligned CCS file that pass all the filters (issue #43)</p></li>
<li><p>Circular alignment: it is assumed that the chromosome has a
circular topology, which is taken into account in the analysis
(issue #38)</p></li>
<li><p>bugfixes in _BamFilePysam: it was writing spurious newlines
after the header and between lines in the body.</p></li>
<li><p>Other issues closed: #45, #54</p></li>
</ul>
</section>
<section id="version-0-16-0">
<h2>Version 0.16.0<a class="headerlink" href="#version-0-16-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>GUI for <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>: <code class="docutils literal notranslate"><span class="pre">sm-analysis-gui</span></code></p></li>
<li><p>Automatic alignment of unaligned input BAM (issue #44)</p></li>
<li><p>Unpleasant error messages from Pysam suppressed (issue #46)</p></li>
<li><p>Consistent behavior of CCS step in <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> (
issues #47 and #3)</p></li>
</ul>
</section>
<section id="version-0-15-0">
<h2>Version 0.15.0<a class="headerlink" href="#version-0-15-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Calls to external tool <code class="docutils literal notranslate"><span class="pre">samtools</span></code> replaced by usage of the <code class="docutils literal notranslate"><span class="pre">pysam</span></code>
library</p></li>
<li><p>Issues closed:</p>
<ul>
<li><p>#10 (methylation report’s name shown on screen upon creation)</p></li>
<li><p>#27 (patitions taken into account in the methylation report)</p></li>
<li><p>#36 (a standard set of filters integrated into <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>)</p></li>
</ul>
</li>
</ul>
</section>
<section id="version-0-14-0">
<h2>Version 0.14.0<a class="headerlink" href="#version-0-14-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Issues #32 and #19 fixed</p></li>
<li><p>Added systematic documentation</p></li>
</ul>
</section>
<section id="version-0-13-0">
<h2>Version 0.13.0<a class="headerlink" href="#version-0-13-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Methylation report V3, with</p>
<ul>
<li><p>quality columns</p></li>
<li><p>length of (DNA) molecule</p></li>
<li><p>column with number of methylations</p></li>
<li><p>columns with number of subreads per strand per molecule</p></li>
</ul>
</li>
<li><p>The list of molecules in a methylation report comes from
the CCS file. Only molecules having the very same sequence
in the reference and in the CCS are included.</p></li>
<li><p>Some improvements on the developmnet side:</p>
<ul>
<li><p>code follows style guide (<code class="docutils literal notranslate"><span class="pre">flake8</span></code> tests pass)</p></li>
<li><p>pipelines ready</p></li>
</ul>
</li>
<li><p>some issues fixed</p></li>
</ul>
</section>
<section id="version-0-12-0">
<h2>Version 0.12.0<a class="headerlink" href="#version-0-12-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Added new command line options to <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-C|--aligned-CCS-bam-file</span></code> to pass an aligned ccs file (that file
is used to produce the Methylation report)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-c|--CCS-bam-file</span></code> to pass a ccs file (that file is used to produce
the Methylation report, after being aligned, if the aligned version
itself is not provided)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--keep-temp-dir</span></code> to preserve a copy of the temporary directory
with all the intermediate files used in the process.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-m|--modification-types</span></code> to select the modification types (m6A,
m4C, …)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--only-produce-methylation-report</span></code> to skip the analysis itself and
only perform the last step: production of the methylation report.</p></li>
</ul>
</li>
<li><p>Issue #2 closed</p></li>
</ul>
</section>
<section id="version-0-11-0">
<h2>Version 0.11.0<a class="headerlink" href="#version-0-11-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Added option <code class="docutils literal notranslate"><span class="pre">-P|--partition</span></code> to <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> to select what fraction
of an input file must be processed. This change allows for an easy way to
further parallelize the processing of input files within different nodes
in a cluster.</p></li>
</ul>
</section>
<section id="version-0-10-0">
<h2>Version 0.10.0<a class="headerlink" href="#version-0-10-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Bugfix in methylation report</p></li>
<li><p>New command line options for <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>:</p>
<ul>
<li><p><code class="docutils literal notranslate"><span class="pre">-N|--num-simultaneous-ipdsummarys</span></code> to launch multiple instances of
ipdSummary</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">-n|--num-workers-per-ipdsummary</span></code> to use multiple workers within each
instance of ipdSummary</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--nprocs-blasr</span></code> to use multiple workers with blasr</p></li>
</ul>
</li>
</ul>
</section>
<section id="version-0-9-0">
<h2>Version 0.9.0<a class="headerlink" href="#version-0-9-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Methylation reports (output by <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code>) conform now to V2
(see <a class="reference internal" href="indepth/methylation_reports.html#methylation-report-format"><span class="std std-ref">Methylation Reports</span></a>).</p></li>
</ul>
</section>
<section id="version-0-8-0">
<h2>Version 0.8.0<a class="headerlink" href="#version-0-8-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>New command line options for <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> to choose the path to the</p>
<ul>
<li><p>aligner (option <code class="docutils literal notranslate"><span class="pre">-b|--blasr-path</span></code>)</p></li>
<li><p>indexer (option <code class="docutils literal notranslate"><span class="pre">-b|--blasr-path</span></code>)</p></li>
<li><p>ipdSummary (option <code class="docutils literal notranslate"><span class="pre">-i|--ipdsummary-path</span></code>)</p></li>
</ul>
</li>
<li><p>High level documentation about PacBio sequencing</p></li>
</ul>
</section>
<section id="version-0-7-0">
<h2>Version 0.7.0<a class="headerlink" href="#version-0-7-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> and <code class="docutils literal notranslate"><span class="pre">bam-filter</span></code> automatically identify the
structure of BAM file:</p>
<ul>
<li><p>where the molecule id is located (column)</p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> has option to select the IPD model in ipdSummary
(option <code class="docutils literal notranslate"><span class="pre">-M|--ipd-model</span></code>).</p></li>
</ul>
</section>
<section id="version-0-6-0">
<h2>Version 0.6.0<a class="headerlink" href="#version-0-6-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>added option <code class="docutils literal notranslate"><span class="pre">--version</span></code> to <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code></p></li>
<li><p>verbosity is configurable (cl option: <code class="docutils literal notranslate"><span class="pre">-v|--verbose</span></code>)</p></li>
<li><p>no tracebacks should reach the end user, only error messages</p></li>
<li><p>More user friendly output of <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> (with relevant key infos)</p></li>
</ul>
</section>
<section id="version-0-5-0">
<h2>Version 0.5.0<a class="headerlink" href="#version-0-5-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><em>Legacy code</em> covered with tests: minimal <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code> functionality</p></li>
<li><p>New <code class="docutils literal notranslate"><span class="pre">csv</span></code> output with methylation states per GATC</p></li>
</ul>
</section>
<section id="version-0-4-0">
<h2>Version 0.4.0<a class="headerlink" href="#version-0-4-0" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>Switched to Double-loop TDD approach</p></li>
<li><p>Code for <code class="docutils literal notranslate"><span class="pre">bam-filter</span></code> re-organized and covered with tests (most of it)</p></li>
<li><p>some bugs fixed</p></li>
<li><p>spike to parallelize (in node) <code class="docutils literal notranslate"><span class="pre">sm-analysis</span></code></p></li>
</ul>
<section id="history">
<h3>History<a class="headerlink" href="#history" title="Permalink to this headline">¶</a></h3>
</section>
</section>
<section id="id2">
<h2>0.1.0 (2020-04-30)<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p>First release on PyPI.</p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">PacBio Data Processing</a></h1>








<h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">Quickstart for the impatient</a></li>
<li class="toctree-l1"><a class="reference internal" href="quickstart_9steps.html">Quickstart on a cluster</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage/index.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="indepth/index.html">Guides</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="devel/index.html">Developer’s documentation</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  <li><a href="devel/index.html">Developer’s documentation</a><ul>
      <li>Previous: <a href="authors.html" title="previous chapter">Credits</a></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2021-2022, David Palao and David Velázquez.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 4.2.0</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/history.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>